---
title: "Peripheral neuropathy clustering differentially expressed genes"
author: "Clemens Hug"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(data.table)
library(synExtra)
library(qs)
library(powerjoin)

synapser::synLogin()
syn <- synDownloader(normalizePath("~/data"), .cache = TRUE)
```

## Reading DEA results

DEA performed by Riki Kawaguchi from UCLA

```{r}
de <- syn("syn70777776") %>%
  read_csv() %>%
  mutate(
    signed_p = -sign(log2FoldChange) * log10(padj)
  )

rescue_list <- syn("syn64608137") %>%
  read_csv()

rescue_type_list <- syn("syn64608138") %>%
  read_csv()

counts_trans <- syn("syn70780244") %>%
  read_csv()

de_meta <- syn("syn70780412") %>%
  read_csv() %>%
  mutate(
    condition_fixed = if_else(
      gne_combined == 0 & jnk_combined == 0 & ptx == "no",
      "DMSO",
      paste(
        if_else(ptx == "yes", "PTX", NA_character_),
        if_else(
          gne_combined > 0,
          paste0("GNE-495"),
          NA_character_
        ),
        if_else(
          jnk_combined > 0,
          paste0("JNK-inh"),
          NA_character_
        ),
        sep = " + "
      ) %>%
        paste0(
          if_else(
            jnk_combined > 0 | gne_combined > 0,
            paste0(" (", if_else(experiment == 1, "3μM", "1μM"), ")"),
            ""
          )
        ) %>%
        str_remove_all("NA\\s+\\+\\s+|\\s+\\+\\s+NA") %>%
        str_trim() %>%
        factor(levels = str_sort(unique(.), numeric = TRUE))
    )
  )
```

## Heatmap functions

```{r}
library(seriation)
library(ComplexHeatmap)

cluster_df <- function(df, row_var, col_var, value_var, values_fill = 0) {
  # browser()
  mat <- df %>%
    select({{row_var}}, {{col_var}}, {{value_var}}) %>%
    pivot_wider(names_from = {{col_var}}, values_from = {{value_var}}, values_fill = values_fill) %>%
    column_to_rownames(rlang::as_name(rlang::enquo(row_var)))
  # browser()
  if (rlang::is_bare_numeric(pull(df, {{value_var}}))) {
    dist_rows <- dist(mat, method = "euclidian")
    dist_cols <- dist(t(mat), method = "euclidian")
  } else {
    # browser()
    dist_rows <- cluster::daisy(mat, metric = "gower")
    dist_cols <- t(mat) %>%
      as.data.frame() %>%
      mutate(across(everything(), \(x) factor(x, levels = levels(pull(df, {{value_var}}))))) %>%
      cluster::daisy(metric = "gower")
  }
  clust_rows <- hclust(dist_rows, method = "average") %>%
      reorder(dist_rows, method = "olo")
  clust_cols <- hclust(dist_cols, method = "average") %>%
      reorder(dist_cols, method = "olo")
  df %>%
    mutate(
      "{{row_var}}" := factor({{row_var}}, levels = clust_rows$labels[clust_rows$order]),
      "{{col_var}}" := factor({{col_var}}, levels = clust_cols$labels[clust_cols$order])
    )
}


cluster_fun_eucl <- function(mat, sample_in_col = TRUE) {
  # if (!sample_in_col) {
  #   mat <- t(mat)
  # }
  # mat_imp <- impute.knn(
  #   mat, rng.seed = 42
  # )[["data"]]
  # if (!sample_in_col) {
  #   mat_imp <- t(mat_imp)
  #   mat <- t(mat)
  # }
  # browser()
  dist_mat <- dist(mat)
  # dist_mat <- as.dist(mat)
  clust <- hclust(dist_mat, method = "average")
  reorder(clust, dist_mat, method = "OLO")
}

```


## LFC Heatmap

```{r}
de_conditions_hm <- de %>%
  distinct(name, batch) %>%
  filter(
    # Only one PTX condition. They're identical
    !(name == "PTX" & batch == "batch_2")
  ) %>%
  mutate(
    name_display = if_else(
      str_detect(name, "GNE|JNK"),
      paste0(name, " (", if_else(batch == "batch_1", "3", "1"), "μM)"),
      name
    )
  )

de_conditions_display_order <- c(
  "PTX",
  "GNE-495 (1μM)",
  "GNE-495 (3μM)",
  "GNE-495 + PTX vs PTX (1μM)",
  "GNE-495 + PTX vs PTX (3μM)",
  "JNK-inh (1μM)",
  "JNK-inh (3μM)",
  "JNK-inh + PTX vs PTX (1μM)",
  "JNK-inh + PTX vs PTX (3μM)",
  "GNE-495 + JNK-inh + PTX vs PTX (1μM)"
)
setdiff(de_conditions_display_order, de_conditions_hm$name_display)

hm_genes <- de %>%
  filter(
    k == 0,
    name %in% c(
      "PTX",
      "GNE-495 + PTX vs PTX (1μM)",
      "GNE-495 + PTX vs PTX (3μM)",
      "JNK-inh + PTX vs PTX (1μM)",
      "JNK-inh + PTX vs PTX (3μM)",
      "GNE-495 + JNK-inh + PTX vs PTX (1μM)"
    ),
    padj < 0.05
  ) %>%
  distinct(gene_id)
```


```{r}
hm_mat <- de %>%
  power_inner_join(
    de_conditions_hm,
    by = c("name", "batch"),
    check = check_specs(
      duplicate_keys_right = "warn"
    )
  ) %>%
  filter(
    k == 0,
    name_display %in% de_conditions_display_order
  ) %>%
  semi_join(hm_genes) %>%
  select(gene_id, name_display, signed_p) %>%
  pivot_wider(names_from = name_display, values_from = signed_p) %>%
  column_to_rownames("gene_id") %>%
  as.matrix() %>% {
    .[, de_conditions_display_order]
  }

hm_mat_scaled <- hm_mat %>%
  t() %>%
  scale(scale = TRUE, center = FALSE) %>%
  t()

row_meta <- rescue_type_list %>%
  filter(
    perturbed_threshold == 0.05,
    gene_id %in% rownames(hm_mat),
    rescue_definition == "significant_opposite",
    batch != "batch_1_riki"
  ) %>%
  transmute(
    gene_id,
    direction = perturbed,
    rescue_type,
    concentration = if_else(batch == "batch_1", "3μM", "1μM")
  ) %>%
  pivot_wider(
    names_from = concentration,
    values_from = rescue_type,
    names_prefix = "rescue_type_"
  ) %>%
  column_to_rownames("gene_id") %>%
  as.data.frame() %>% {
    .[rownames(hm_mat), ]
  }

row_clust <- cluster_fun_eucl(hm_mat_scaled, sample_in_col = TRUE)

mat_max_abs <- max(abs(quantile(hm_mat_scaled, c(.025, .975), na.rm = TRUE)))
hm <- Heatmap(
  hm_mat_scaled,
  col = circlize::colorRamp2(seq(from = -mat_max_abs, to = mat_max_abs, length.out = 51), paletteer::paletteer_c("pals::ocean.balance", n = 51)),
  show_column_names = TRUE,
  show_row_names = FALSE,
  cluster_rows = row_clust,
  cluster_columns = FALSE,
  use_raster = TRUE,
  raster_device = "agg_png",
  right_annotation = HeatmapAnnotation(
    df = row_meta,
    which = "row",
    col = list(
      direction = c("up" = "red", "down" = "blue", "not perturbed" = "white"),
      rescue_type_3μM = c(
        "both" = "purple", "no rescue" = "grey",
        "GNE-495 only" = "blue", "JNK-inh only" = "red",
        "not perturbed" = "white"
      ),
      rescue_type_1μM = c(
        "both" = "purple", "no rescue" = "grey",
        "GNE-495 only" = "blue", "JNK-inh only" = "red",
        "not perturbed" = "white"
      )
    )
  ),
  show_row_dend = FALSE
)

withr::with_cairo_pdf(
  file.path("plots", "de_heatmaps", "de_heatmap_all_combined_signed_p_scaled.pdf"),
  draw(hm),
  width = 7, height = 15
)
```



### PCA


```{r}
pca_combined <- counts_trans %>%
  filter(type %in% c("vst", "vst_batch_removed")) %>%
  group_nest(type) %>%
  mutate(
    pca = map(
      data,
      \(df) {
        mat <- df %>%
          column_to_rownames("gene_id") %>%
          as.matrix()
        prcomp(
          t(mat),
          center = TRUE,
          scale = FALSE
        )
      }
    )
  )

pca_x_long <- pca_combined %>%
  transmute(
    type,
    x = map(
      pca,
      \(pca) {
        as_tibble(
          pca$x,
          rownames = "sample_id_sequencing"
        )
      }
    )
  ) %>%
  unnest(x) %>%
  pivot_longer(
    -c(type, sample_id_sequencing),
    names_to = "component",
    values_to = "value"
  )

pca_x_vs <- combn(1:5, 2) %>%
  t() %>%
  magrittr::set_colnames(c("component_1", "component_2")) %>%
  as_tibble() %>%
  mutate(
    across(
      everything(),
      \(x) paste0("PC", x)
    )
  ) %>%
  left_join(
    pca_x_long,
    by = c("component_1" = "component")
  ) %>%
  left_join(
    pca_x_long,
    by = c("component_2" = "component", "sample_id_sequencing", "type"),
    suffix = c("_1", "_2")
  ) %>%
  left_join(
    de_meta,
    by = "sample_id_sequencing"
  )

pca_sd <- pca_combined %>%
  transmute(
    type,
    sd = map(
      pca,
      \(pca) {
        broom::tidy(pca, matrix = "eigenvalues")
      }
    )
  ) %>%
  unnest(sd) %>%
  mutate(
    across(PC, \(x) paste0("PC", x))
  )
```

```{r}
rescue_colors <- c(
  "none" = "grey70",
  "GNE-495" = "red",
  "JNK-inh" = "blue",
  "GNE-495 + JNK-inh" = "purple"
) %>%
  colorspace::lighten(amount = .4)

rescue_dose_colors <- c(
  c("none" = "grey70"),
  set_names(
    colorspace::lighten(c("red", "blue", "purple"), amount = .4),
    c("GNE-495 (1 μM)", "JNK-inh (1 μM)", "GNE-495 + JNK-inh (1 μM)")
  ),
  set_names(
    colorspace::lighten(c("red", "blue", "purple"), amount = .05),
    c("GNE-495 (3 μM)", "JNK-inh (3 μM)", "GNE-495 + JNK-inh (3 μM)")
  )
)

pca_vs_plots <- pca_x_vs %>%
  power_inner_join(
    pca_sd,
    by = c("type", "component_1" = "PC"),
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  power_inner_join(
    pca_sd,
    by = c("type", "component_2" = "PC"),
    suffix = c("_1", "_2"),
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  mutate(
    rescue = case_when(
      gne_495 == "yes" & jnk_in == "yes" ~ "GNE-495 + JNK-inh",
      gne_495 == "yes" ~ "GNE-495",
      jnk_in == "yes" ~ "JNK-inh",
      TRUE ~ "none"
    ),
    rescue_dose = if_else(
      rescue != "none",
      paste0(rescue, if_else(experiment == 1, " (3 μM)", " (1 μM)")),
      rescue
    ),
    component_1_sd = paste0(component_1, " (", signif(percent_1 * 100, 2), "%)"),
    component_2_sd = paste0(component_2, " (", signif(percent_2 * 100, 2), "%)")
  ) %>%
  group_nest(type) %>%
  rowwise() %>%
  mutate(
    p = list(
      ggplot(
        data,
        aes(
          x = value_1,
          y = value_2,
          color = ptx,
          fill = rescue_dose,
        )
      ) +
        geom_point(
          shape = 21,
          alpha = .7
        ) +
        scale_color_manual(
          values = c(
            "no" = "grey90",
            "yes" = "black"
          )
        ) +
        scale_fill_manual(
          values = rescue_dose_colors
        ) +
        guides(
          fill = guide_legend(override.aes = list(shape = 21, color = NA)),
          color = guide_legend(override.aes = list(shape = 21, fill = NA))
        ) +
        facet_wrap(
          vars(component_1_sd, component_2_sd),
          axes = "all",
          axis.labels = "margins"
        ) +
        envalysis::theme_publish() +
        coord_equal() +
        labs(
          x = "Component 1",
          y = "Component 2",
          color = "Paclitaxel",
          fill = "Rescue treatment",
          title = type
        )
    )
  )

pwalk(
  pca_vs_plots,
  \(type, p, ...) {
    ggsave(
      file.path("rnaseq", "plots", paste0("pca_vs_all_combined_", type, ".pdf")),
      p,
      width = 12,
      height = 10,
      device = cairo_pdf
    )
  }
)
```


```{r}


```

