---
title: "Peripheral neuropathy gene set analysis plotting"
author: "Clemens Hug"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(here)
library(tidyverse)
library(data.table)
library(synExtra)
library(qs)
library(powerjoin)
library(plotly)
library(ggrepel)

synapser::synLogin()
syn <- synDownloader(normalizePath("~/data"), .cache = TRUE)
```

## Reading GSE results

```{r}
topgo_res_pert <- syn("syn64461453") %>%
  qread()

fgsea_res_pert <- syn("syn64461445") %>%
  qread()

fgsea_res_rescue <- syn("syn64461444") %>%
  qread()

de_long_selected <- syn("syn64582734") %>%
  read_csv()

feasible_genes <- syn("syn64582742") %>%
  read_csv()

fgsea_overrep <- syn("syn64582959") %>%
  qread()

fgsea_intersection_overrep <- syn("syn64586672") %>%
  qread() %>%
  mutate(
    # avoid weird fucking bug when grouping that values are repeated
    across(estimate, unname)
  )

indra_res <- syn("syn64908970") %>%
  read_csv()

rag_ion_channel_genes <- syn("syn64745787") %>%
  readxl::read_excel() %>%
  pivot_longer(
    everything(),
    names_to = "gene_set",
    values_to = "gene_symbol_mouse"
  ) %>%
  drop_na()

ptx_rescue_list <- syn("syn64608137") %>%
  read_csv()

ptx_rescue_types <- syn("syn64608138") %>%
  read_csv()
```



## Plot Reactome PTX only and comparison with combination treatment


```{r}
fgsea_res_single_treatments <- fgsea_res_pert %>%
  filter(
    metric == "signed_padj"
  ) %>%
  mutate(
    database = str_extract(pathway, "^(REACTOME|HALLMARK|PID|KEGG_MEDICUS)"),
    pathway = str_remove(pathway, "^(REACTOME|HALLMARK|PID|KEGG_MEDICUS)_")
  ) %>%
  filter(
    database == "REACTOME",
    padj < 0.01
  ) %>%
  mutate(
    sign = if_else(NES > 0, "pos", "neg")
  )

ps <- fgsea_res_single_treatments %>%
  group_nest(
    batch, contrast
  ) %>%
  mutate(
    p = map2(
      contrast, data,
      \(x, y) {
        y %>%
          mutate(
            pathway = fct_reorder(pathway, NES)
          ) %>%
          ggplot(
            aes(
              x = NES,
              y = pathway,
              fill = sign
            )
          ) +
          geom_col() +
          geom_text(
            aes(
              x = if_else(sign == "pos", -.05, .05),
              hjust = if_else(sign == "pos", 1, 0),
              label = pathway
            ),
            size = 2
          ) +
          scale_fill_discrete(guide = "none", direction = -1) +
          theme(
            axis.text.y = element_blank(),
            panel.grid.major.y = element_blank(),
          ) +
          labs(
            x = "Normalized enrichment score",
            y = NULL,
            title = paste0("Reactome pathways enriched in ", x, " treatment"),
            subtitle = "FDR < 0.01"
          ) + {
            if (!any(y$sign == "neg"))
              lims(x = c(-max(abs(y$NES)), NA))
          }
      }
    )
  )

pwalk(
  ps,
  \(batch, contrast, p, ...) {
    ggsave(
      file.path("plots", paste0("reactome_", batch, "_", contrast, "_bars.pdf")),
      p, width = 8, height = 8
    )
  }
)
```

### Compare batch 1 vs batch 2

```{r}
fgsea_res_pert_vs <- fgsea_res_pert %>%
  mutate(
    database = str_extract(pathway, "^(REACTOME|HALLMARK|PID|KEGG_MEDICUS)"),
    pathway = str_remove(pathway, "^(REACTOME|HALLMARK|PID|KEGG_MEDICUS)_"),
    sign = if_else(NES > 0, "pos", "neg"),
    signed_padj = -sign(NES) * log10(padj)
  ) %>%
  pivot_wider(
    names_from = batch,
    values_from = c(pval, padj, log2err, ES, NES, leadingEdge, sign, size, signed_padj)
  )

p <- fgsea_res_pert_vs %>%
  ggplot(
    aes(
      x = signed_padj_batch1,
      y = signed_padj_batch2
    )
  ) +
  geom_point(
    shape = 16, alpha = .7
  ) +
  facet_grid(
    contrast ~ metric
  )
```

### Comparison with combination treatment

```{r}
all_contrasts <- fgsea_res_pert$contrast %>%
  unique()

contrast_plotting_pairs <- bind_rows(
  tibble(
    contrast_1 = "PTX",
    contrast_2 = all_contrasts[
      str_ends(all_contrasts, fixed("+ PTX")) |
      str_ends(all_contrasts, fixed("+ PTX vs PTX"))
    ]
  ),
  tibble(
    contrast_1 = "GNE-495 + PTX",
    contrast_2 = "JNK-inh + PTX"
  )
)

fgsea_res_ptx_vs_combo <- fgsea_res_pert %>%
  filter(
    metric == "signed_padj"
  ) %>%
  separate_wider_delim(
    pathway,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge"
  ) %>%
  mutate(
    signed_padj = -sign(NES) * log10(padj)
  ) %>%
  filter(
    database == "REACTOME"
  ) %>% {
    d <- .
    power_inner_join(
      contrast_plotting_pairs,
      d,
      by = join_by(contrast_1 == contrast),
      suffix = c("", "_1"),
      check = check_specs(
        unmatched_keys_left = "warn"
      )
    ) %>%
    power_inner_join(
      d,
      by = join_by(contrast_2 == contrast, metric, database, pathway),
      suffix = c("_1", "_2"),
      check = check_specs(
        unmatched_keys_left = "warn"
      )
    )
  } %>%
  mutate(
    sig_class = case_when(
      padj_1 < .05 & padj_2 < .05 ~ "both",
      padj_1 < .05 ~ "1",
      padj_2 < .05 ~ "2",
      TRUE ~ "none"
    )
  )

sig_class_colors <- c(
  "both" = "purple",
  "1" = "red",
  "2" = "blue",
  "none" = "grey"
)

ps <- fgsea_res_ptx_vs_combo %>%
  group_nest(contrast_1, contrast_2) %>%
  mutate(
    p = pmap(
      .,
      \(contrast_1, contrast_2, data, ...) {
        ggplot(
          data %>%
            arrange(match(sig_class, names(sig_class_colors))),
          aes(
            x = signed_padj_1,
            y = signed_padj_2
          )
        ) +
        geom_point(
          aes(
            color = sig_class,
            text = pathway
          ),
          alpha = .8
        ) +
        scale_color_manual(
          name = "Significant\nenrichment\nFDR < 0.05",
          values = sig_class_colors,
          breaks = names(sig_class_colors),
          labels = c(
            "Both",
            contrast_1,
            contrast_2,
            "None"
          )
        ) +
        theme_minimal() +
        labs(
          x = paste("Enrichment", contrast_1),
          y = paste("Enrichment", contrast_2)
        )
      }
    )
  )

pwalk(
  ps,
  \(contrast_1, contrast_2, p, ...) {
    ggsave(
      file.path("plots", paste0("reactome_", contrast_1, "_vs_", contrast_2, "_scatter_padj.pdf")),
      p, width = 7, height = 6
    )
    htmlwidgets::saveWidget(
      ggplotly(p),
      file.path("plots", paste0("reactome_", contrast_1, "_vs_", contrast_2, "_scatter_padj.html"))
    )
  }
)


ps <- fgsea_res_ptx_vs_combo %>%
  group_nest(contrast_ptx, contrast_combo) %>%
  mutate(
    p = map2(
      contrast_combo, data,
      \(x, y) {
        ggplot(
          y %>%
            arrange(match(sig_class, names(sig_class_colors))),
          aes(
            x = NES_ptx,
            y = NES_combo
          )
        ) +
        geom_point(
          aes(
            color = sig_class,
            text = pathway
          ),
          alpha = .8
        ) +
        scale_color_manual(
          values = sig_class_colors
        )
      }
    )
  )

pwalk(
  ps,
  \(contrast_ptx, contrast_combo, p, ...) {
    ggsave(
      file.path("plots", paste0("reactome_", contrast_ptx, "_vs_", contrast_combo, "_scatter.pdf")),
      p, width = 6, height = 6
    )
    htmlwidgets::saveWidget(
      ggplotly(p),
      file.path("plots", paste0("reactome_", contrast_ptx, "_vs_", contrast_combo, "_scatter.html"))
    )
  }
)
```

### Highlighting certain pathways

```{r}
highlighted_pathways <- tribble(
  ~contrast_1, ~contrast_2, ~pathways,
  "GNE-495 + PTX", "JNK-inh + PTX", c(
    "METABOLISM_OF_LIPIDS",
    "CHOLESTEROL_BIOSYNTHESIS",
    "ACTIVATION_OF_GENE_EXPRESSION_BY_SREBF_SREBP",
    "METABOLISM_OF_STEROIDS",
    "REGULATION_OF_CHOLESTEROL_BIOSYNTHESIS_BY_SREBP_SREBF",
    "FATTY_ACID_METABOLISM",
    "FATTY_ACYL_COA_BIOSYNTHESIS",
    "PHOSPHOLIPID_METABOLISM",
    "NEUTROPHIL_DEGRANULATION",
    "VESICLE_MEDIATED_TRANSPORT",
    "ER_TO_GOLGI_ANTEROGRADE_TRANSPORT",
    "CELL_CYCLE_MITOTIC",
    "RHO_GTPASE_EFFECTORS",
    "CYTOKINE_SIGNALING_IN_IMMUNE_SYSTEM",
    "EXTRACELLULAR_MATRIX_ORGANIZATION",
    "COPI_MEDIATED_ANTEROGRADE_TRANSPORT",
    "SYNDECAN_INTERACTIONS",
    ##
    "ACTIVATION_OF_GENE_EXPRESSION_BY_SREBF_SREBP",
    "FATTY_ACID_METABOLISM",
    "METABOLISM_OF_LIPIDS",
    "TRNA_AMINOACYLATION",
    "CELL_CYCLE",
    "REGULATION_OF_EXPRESSION_OF_SLITS_AND_ROBOS",
    "GAP_JUNCTION_TRAFFICKING_AND_REGULATION",
    "ER_TO_GOLGI_ANTEROGRADE_TRANSPORT",
    "COPI_MEDIATED_ANTEROGRADE_TRANSPORT"
  ) %>%
    unique()
)


ps <- fgsea_res_ptx_vs_combo %>%
  group_nest(contrast_1, contrast_2) %>%
  inner_join(
    highlighted_pathways,
    by = c("contrast_1", "contrast_2")
  ) %>%
  mutate(
    p = pmap(
      .,
      \(contrast_1, contrast_2, data, pathways, ...) {
        ggplot(
          data %>%
            arrange(match(sig_class, names(sig_class_colors))),
          aes(
            x = signed_padj_1,
            y = signed_padj_2
          )
        ) +
        geom_point(
          aes(
            color = sig_class,
            text = pathway
          ),
          alpha = .8
        ) +
        geom_text_repel(
          data = \(x) mutate(
              x,
              pathway_short = if_else(
                pathway %in% pathways,
                pathway,
                ""
              ) %>%
                str_replace_all(fixed("_"), " ") %>%
                str_wrap(20)
            ),
          aes(
            label = pathway_short
          ),
          size = 2,
          # force = 3,
          max.iter = 1e6,
          seed = 42,
          max.overlaps = Inf,
          min.segment.length = 0.3
        ) +
        scale_color_manual(
          name = "Significant\nenrichment\nFDR < 0.05",
          values = sig_class_colors,
          breaks = names(sig_class_colors),
          labels = c(
            "Both",
            contrast_1,
            contrast_2,
            "None"
          )
        ) +
        scale_x_continuous(
          expand = expansion(mult = c(.18, .05))
        ) +
        scale_y_continuous(
          expand = expansion(mult = c(.1, .05))
        ) +
        theme_minimal() +
        labs(
          x = paste("Enrichment", contrast_1),
          y = paste("Enrichment", contrast_2)
        )
      }
    )
  )

pwalk(
  ps,
  \(contrast_1, contrast_2, p, ...) {
    ggsave(
      file.path("plots", paste0("reactome_", contrast_1, "_vs_", contrast_2, "_scatter_padj_highlighted.pdf")),
      p, width = 7.5, height = 6
    )
    # htmlwidgets::saveWidget(
    #   ggplotly(p),
    #   file.path("plots", paste0("reactome_", contrast_ptx, "_vs_", contrast_combo, "_scatter_padj.html"))
    # )
  }
)
```


## Heatmap of Reactome PTX and combo


```{r}

library(seriation)
cluster_df <- function(df, row_var, col_var, value_var, values_fill = 0) {
  # browser()
  mat <- df %>%
    select({{row_var}}, {{col_var}}, {{value_var}}) %>%
    pivot_wider(names_from = {{col_var}}, values_from = {{value_var}}, values_fill = values_fill) %>%
    column_to_rownames(rlang::as_name(rlang::enquo(row_var)))
  # browser()
  if (rlang::is_bare_numeric(pull(df, {{value_var}}))) {
    dist_rows <- dist(mat, method = "euclidian")
    dist_cols <- dist(t(mat), method = "euclidian")
  } else {
    # browser()
    dist_rows <- cluster::daisy(mat, metric = "gower")
    dist_cols <- t(mat) %>%
      as.data.frame() %>%
      mutate(across(everything(), \(x) factor(x, levels = levels(pull(df, {{value_var}}))))) %>%
      cluster::daisy(metric = "gower")
  }
  clust_rows <- hclust(dist_rows, method = "average") %>%
      reorder(dist_rows, method = "olo")
  clust_cols <- hclust(dist_cols, method = "average") %>%
      reorder(dist_cols, method = "olo")
  df %>%
    mutate(
      "{{row_var}}" := factor({{row_var}}, levels = clust_rows$labels[clust_rows$order]),
      "{{col_var}}" := factor({{col_var}}, levels = clust_cols$labels[clust_cols$order])
    )
}


cluster_fun_eucl <- function(mat, sample_in_col = TRUE) {
  # if (!sample_in_col) {
  #   mat <- t(mat)
  # }
  # mat_imp <- impute.knn(
  #   mat, rng.seed = 42
  # )[["data"]]
  # if (!sample_in_col) {
  #   mat_imp <- t(mat_imp)
  #   mat <- t(mat)
  # }
  # browser()
  dist_mat <- dist(mat)
  # dist_mat <- as.dist(mat)
  clust <- hclust(dist_mat, method = "average")
  reorder(clust, dist_mat, method = "OLO")
}

```

```{r}
fgsea_hm_conditions <- c(
  "PTX", "GNE-495", "JNK-inh",
  "GNE-495 + PTX", "JNK-inh + PTX"
)

fgsea_res_pert_hm_data <- fgsea_res_pert %>%
  filter(
    metric == "signed_padj",
    contrast %in% fgsea_hm_conditions
  ) %>%
  separate_wider_delim(
    pathway,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge"
  ) %>%
  mutate(
    signed_padj = -sign(NES) * log10(padj)
  ) %>%
  filter(
    database == "REACTOME"
  ) %>%
  cluster_df(
    pathway, contrast, signed_padj, values_fill = 0
  )

fgsea_res_pert_mat <- fgsea_res_pert_hm_data %>%
  arrange(pathway, contrast) %>%
  select(pathway, contrast, signed_padj) %>%
  pivot_wider(names_from = contrast, values_from = signed_padj, values_fill = 0) %>%
  column_to_rownames("pathway") %>%
  as.matrix()

fgsea_res_pert_hm_pathways <- fgsea_res_pert_hm_data %>%
  filter(
    padj < 0.01
  ) %>%
  pull(pathway) %>%
  unique()
```


```{r}
row_clust <- cluster_fun_eucl(
  fgsea_res_pert_mat[as.character(fgsea_res_pert_hm_pathways), fgsea_hm_conditions]
)

row_clust_df_long <- tibble(k = 4:10) %>%
  mutate(
    res = map(k, \(x) enframe(cutree(row_clust, k = x), name = "pathway", value = "cluster"))
  ) %>%
  unnest(res)

row_clust_df <- row_clust_df_long %>%
  mutate(
    cluster = fct_inseq(as.character(cluster))
  ) %>%
  pivot_wider(
    names_from = k,
    values_from = cluster,
    names_prefix = "k_"
  )
```

k = 8 seems to work well


```{r}
dir.create(here("results"), showWarnings = FALSE)
write_csv(
  row_clust_df_long %>%
    filter(k == 8),
  here("results", "reactome_pert_clusters_k8.csv")
)

```

```{r}
library(ComplexHeatmap)
hm <- Heatmap(
  fgsea_res_pert_mat[as.character(fgsea_res_pert_hm_pathways), fgsea_hm_conditions],
  cluster_rows = cluster_fun_eucl,
  cluster_columns = FALSE,
  show_row_names = FALSE,
  left_annotation = HeatmapAnnotation(
    df = column_to_rownames(row_clust_df, "pathway")[as.character(fgsea_res_pert_hm_pathways), ],
    which = "row"
  )
)

fgsea_picked_to_highlight <- c(
  "METABOLISM_OF_LIPIDS",
  "METABOLISM_OF_STEROIDS",
  "CHOLESTEROL_BIOSYNTHESIS",
  "ACTIVATION_OF_GENE_EXPRESSION_BY_SREBF_SREBP",
  "REGULATION_OF_CHOLESTEROL_BIOSYNTHESIS_BY_SREBP_SREBF",
  "REGULATION_OF_EXPRESSION_OF_SLITS_AND_ROBOS",
  "SIGNALING_BY_ROBO_RECEPTORS",
  "FATTY_ACID_METABOLISM",
  "FATTY_ACYL_COA_BIOSYNTHESIS",
  "PHOSPHOLIPID_METABOLISM",
  "GLYCEROPHOSPHOLIPID_BIOSYNTHESIS",
  "GAP_JUNCTION_ASSEMBLY",
  "KINESINS",
  "GAP_JUNCTION_TRAFFICKING_AND_REGULATION",
  "ORGANELLE_BIOGENESIS_AND_MAINTENANCE",
  "RHO_GTPASES_ACTIVATE_IQGAPS",
  "COPI_INDEPENDENT_GOLGI_TO_ER_RETROGRADE_TRAFFIC",
  "RHO_GTPASES_ACTIVATE_FORMINS",
  "HDACS_DEACETYLATE_HISTONES",
  "COPI_DEPENDENT_GOLGI_TO_ER_RETROGRADE_TRAFFIC",
  "GOLGI_TO_ER_RETROGRADE_TRANSPORT",
  "INTRA_GOLGI_AND_RETROGRADE_GOLGI_TO_ER_TRAFFIC",
  "SIGNALING_BY_RHO_GTPASES_MIRO_GTPASES_AND_RHOBTB3",
  "TRANSPORT_TO_THE_GOLGI_AND_SUBSEQUENT_MODIFICATION",
  "RHO_GTPASE_EFFECTORS",
  "APOPTOSIS",
  "ER_TO_GOLGI_ANTEROGRADE_TRANSPORT",
  "VESICLE_MEDIATED_TRANSPORT",
  "MEMBRANE_TRAFFICKING"
)

mat <- fgsea_res_pert_mat[as.character(fgsea_res_pert_hm_pathways), fgsea_hm_conditions]
mat_max_abs <- max(abs(quantile(mat, c(.025, .975), na.rm = TRUE)))
hm <- Heatmap(
  mat,
  col = circlize::colorRamp2(seq(from = -mat_max_abs, to = mat_max_abs, length.out = 51), paletteer::paletteer_c("pals::ocean.balance", n = 51)),
  name = "Normalized enrichment score",
  cluster_rows = cluster_fun_eucl,
  cluster_columns = FALSE,
  show_row_names = FALSE,
  left_annotation = HeatmapAnnotation(
    df = row_clust_df %>%
        select(pathway, cluster = k_8) %>%
        slice(match(as.character(fgsea_res_pert_hm_pathways), pathway)) %>%
        column_to_rownames("pathway"),
    col = list(
      cluster = set_names(
        ggokabeito::palette_okabe_ito(1:8), 1:8
      )
    ),
    which = "row"
  ),
  right_annotation = HeatmapAnnotation(
    pathways = fgsea_picked_to_highlight %>%
      {
        anno_mark(
          at = match(., fgsea_res_pert_hm_pathways),
          labels = str_wrap(
            str_replace_all(., "_", " "),
            20
          ),
          labels_gp = gpar(fontsize = 6),
          padding = unit(2, "mm")
        )
      },
    which = "row"
  )
)

withr::with_pdf(
  here("plots", "reactome_pert_heatmap_k8.pdf"),
  draw(hm),
  width = 6, height = 8
)

InteractiveComplexHeatmap::htShiny(draw(hm), save = here("plots/reactome_pert_heatmap_k8"))

hm_with_names <- Heatmap(
  mat,
  col = circlize::colorRamp2(seq(from = -mat_max_abs, to = mat_max_abs, length.out = 51), paletteer::paletteer_c("pals::ocean.balance", n = 51)),
  name = "Normalized enrichment score",
  cluster_rows = cluster_fun_eucl,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  row_names_gp = gpar(fontsize = 8),
  left_annotation = HeatmapAnnotation(
    df = row_clust_df %>%
        select(pathway, cluster = k_8) %>%
        slice(match(as.character(fgsea_res_pert_hm_pathways), pathway)) %>%
        column_to_rownames("pathway"),
    col = list(
      cluster = set_names(
        ggokabeito::palette_okabe_ito(1:8), 1:8
      )
    ),
    which = "row"
  )
)

withr::with_pdf(
  here("plots", "reactome_pert_heatmap_k8_with_names.pdf"),
  draw(hm_with_names),
  width = 9, height = 16
)

```


## Plotting rescue vector FGSEA results

```{r}
fgsea_res_rescue_data <- fgsea_res_rescue %>%
  filter(metric == "signed_padj_rescue_clamped") %>%
  separate_wider_delim(
    pathway,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge"
  ) %>%
  mutate(
    sign = if_else(NES > 0, "pos", "neg"),
    pathway_db = paste(str_sub(database, 1, 1), pathway, sep = "_")
  )


ps <- fgsea_res_rescue_data %>%
  group_nest(
    compound
  ) %>%
  mutate(
    p = map2(
      compound, data,
      \(x, y) {
        y %>%
          filter(
            padj < .05
          ) %>%
          mutate(
            across(pathway_db, \(x) fct_reorder(x, NES))
          ) %>%
          ggplot(
            aes(
              x = NES,
              y = pathway_db,
              fill = sign
            )
          ) +
          geom_col() +
          geom_text(
            aes(
              x = if_else(sign == "pos", -.05, .05),
              hjust = if_else(sign == "pos", 1, 0),
              label = pathway_db
            ),
            size = 2
          ) +
          scale_fill_discrete(guide = "none", direction = -1) +
          theme(
            axis.text.y = element_blank(),
            panel.grid.major.y = element_blank(),
          ) +
          labs(
            x = "Normalized enrichment score",
            y = NULL,
            title = paste0("Pathways enriched in genes rescued by ", x),
            subtitle = "FDR < 0.05"
          ) + {
            if (!any(y$sign == "neg"))
              lims(x = c(-max(abs(y$NES)), NA))
          }
      }
    )
  )

pwalk(
  ps,
  \(compound, p, ...) {
    ggsave(
      file.path("plots", paste0("reactome_", compound, "_bars_rescue.pdf")),
      p, width = 8, height = 8
    )
  }
)
```


### Comparing rescue vector enrichment between combos

```{r}
fgsea_rescue_combo_vs_combo <- fgsea_res_rescue_data %>%
  mutate(
    signed_padj = -sign(NES) * log10(padj)
  ) %>%
  # filter(
  #   database == "REACTOME"
  # ) %>%
  pivot_wider(
    names_from = compound,
    values_from = -c(metric, compound, pathway, database)
  ) %>%
  mutate(
    sig_class = case_when(
      `padj_GNE-495` < .05 & `padj_JNK-inh` < .05 ~ "both",
      `padj_GNE-495` < .05 ~ "GNE-495",
      `padj_JNK-inh` < .05 ~ "JNK-inh",
      TRUE ~ "none"
    )
  )

sig_class_colors <- c(
  "both" = "purple",
  "GNE-495" = "red",
  "JNK-inh" = "blue",
  "none" = "grey"
)

p <- fgsea_rescue_combo_vs_combo %>%
  arrange(match(sig_class, names(sig_class_colors))) %>%
  ggplot(
    aes(
      x = `signed_padj_GNE-495`,
      y = `signed_padj_JNK-inh`
    )
  ) +
  geom_point(
    aes(
      color = sig_class,
      text = pathway
    ),
    alpha = .8
  ) +
  scale_color_manual(
    values = sig_class_colors
  ) +
  labs(
    x = paste("Enrichment GNE-495 rescue"),
    y = paste("Enrichment JNK-inh rescue")
  )

ggsave(
  file.path("plots", paste0("reactome_rescue_gne_vs_jnk_scatter_padj.pdf")),
  p, width = 8, height = 8
)
htmlwidgets::saveWidget(
  ggplotly(p),
  file.path("plots", paste0("reactome_rescue_gne_vs_jnk_scatter_padj.html"))
)


```

### Comparing rescue vector enrichment against PTX alone

```{r}
fgsea_rescue_combo_vs_tx <- fgsea_res_rescue_data %>%
  mutate(
    signed_padj = -sign(NES) * log10(padj)
  ) %>%
  power_inner_join(
    fgsea_res_pert %>%
      filter(
        contrast == "PTX",
        metric == "signed_padj"
      ) %>%
      mutate(
        signed_padj = -sign(NES) * log10(padj)
      ) %>%
      separate_wider_delim(
        pathway,
        delim = "_",
        names = c("database", "pathway"),
        too_many = "merge"
      ),
    by = c("database", "pathway"),
    suffix = c("_rescue", "_PTX"),
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  )

p <- fgsea_rescue_combo_vs_tx %>%
  ggplot(
    aes(
      x = signed_padj_PTX,
      y = signed_padj_rescue
    )
  ) +
  geom_point(
    aes(
      text = pathway
    )
  ) +
  facet_wrap(~compound)


htmlwidgets::saveWidget(
  ggplotly(p),
  file.path("plots", "reactome_rescue_vs_ptx_scatter_padj.html")
)
```


## Plot LFC heatmap of genes in enriched pathways PTX vs combo

```{r}
library(msigdbr)
all_msigdbr <- msigdbr()

msigdbr_of_interest <- all_msigdbr %>%
  filter(
    gs_cat == "H" |
    gs_subcat %in% c(
      "CP:PID",
      "CP:KEGG",
      "CP:REACTOME"
      # "GO:BP",
      # "GO:MF"
    ),
    ensembl_gene %in% feasible_genes$gene_id
  ) %>%
  separate_wider_delim(
    gs_name,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge",
    cols_remove = FALSE
  )

msigdbr_of_interest_mat <- msigdbr_of_interest %>%
  transmute(gs_name, gene_symbol, dummy = TRUE) %>%
  distinct() %>%
  pivot_wider(
    names_from = gs_name,
    values_from = dummy,
    values_fill = FALSE
  ) %>%
  column_to_rownames("gene_symbol") %>%
  as.matrix()

msigdbr_of_interest_sim <- msigdbr_of_interest_mat %>%
  proxy::dist(method = "Jaccard", by_rows = FALSE)
```


```{r}
fgsea_hm_conditions <- c(
  "PTX", "GNE-495", "JNK-inh",
  "GNE-495 + PTX", "JNK-inh + PTX"
)

de_pert_hm_data <- de_long_selected %>%
  filter(
    contrast %in% fgsea_hm_conditions
  ) %>%
  mutate(
    signed_padj = -sign(logFC) * log10(FDR_min_zero)
  )

gene_id_gene_symbol_map <- de_pert_hm_data %>%
  distinct(gene_id, gene_name) %>%
  {with(., set_names(gene_name, gene_id))}

plot_pathway_heatmap <- function(pathway) {
  gene_sets <- msigdbr_of_interest %>%
    filter(pathway == !!pathway) %>%
    pull(ensembl_gene) %>%
    unique()

  de_mat <- de_pert_hm_data %>%
    filter(gene_id %in% gene_sets) %>%
    select(gene_id, contrast, logFC) %>%
    pivot_wider(names_from = contrast, values_from = logFC, values_fill = 0) %>%
    column_to_rownames("gene_id") %>%
    as.matrix() %>% {
      .[, fgsea_hm_conditions]
    }

  row_meta <- fgsea_res_rescue_data %>%
    filter(pathway == !!pathway) %>%
    select(compound, gene_id = leadingEdge) %>%
    unchop(gene_id) %>%
    mutate(
      gene_id = factor(gene_id, levels = rownames(de_mat)),
      decoy = 1L
    ) %>%
    pivot_wider(
      names_from = compound, values_from = decoy,
      names_prefix = "leading_edge_",
      id_expand = TRUE, values_fill = 0L
    ) %>%
    column_to_rownames("gene_id") %>% {
      .[rownames(de_mat), ]
    }

  mat_max_abs <- max(abs(quantile(de_mat, c(.025, .975), na.rm = TRUE)))

  hm <- Heatmap(
    de_mat,
    name = "log2 fold change",
    col = circlize::colorRamp2(seq(from = -mat_max_abs, to = mat_max_abs, length.out = 51), paletteer::paletteer_c("pals::ocean.balance", n = 51)),
    cluster_rows = cluster_fun_eucl,
    cluster_columns = FALSE,
    row_labels = gene_id_gene_symbol_map[rownames(de_mat)],
    right_annotation = HeatmapAnnotation(
      df = row_meta,
      which = "row"
    )
  )

  return(hm)
}

plot_pathway_heatmap("METABOLISM_OF_LIPIDS")

library(ComplexHeatmap)
de_pert_rescue_hm_data <- tibble(
  pathway = c(
    "METABOLISM_OF_LIPIDS",
    "RECYCLING_PATHWAY_OF_L1",
    "REGULATION_OF_EXPRESSION_OF_SLITS_AND_ROBOS",
    "SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE",
    # "EPH_EPHRIN_SIGNALIN",
    "EPHB_MEDIATED_FORWARD_SIGNALING",
    "P38MAPK_EVENTS",
    "RHO_GTPASE_EFFECTORS"
  )
) %>%
  mutate(
    genes = map(
      pathway,
      \(x) msigdbr_of_interest %>%
        filter(
          pathway == x
        ) %>%
        pull(ensembl_gene) %>%
        unique()
    ),
    mat = map(
      genes,
      \(x) de_pert_hm_data %>%
        filter(
          gene_id %in% x
        ) %>%
        select(gene_id, contrast, logFC) %>%
        pivot_wider(names_from = contrast, values_from = logFC, values_fill = 0) %>%
        column_to_rownames("gene_id") %>%
        as.matrix() %>% {
          .[, fgsea_hm_conditions]
        }
    ),
    row_meta = map2(
      pathway, mat,
      \(x, y) {
        tibble(
          gene_id = rownames(y)
        ) %>%

        fgsea_res_rescue_data %>%
          filter(pathway == x) %>%
          select(compound, gene_id = leadingEdge) %>%
          unchop(gene_id) %>%
          mutate(
            gene_id = factor(gene_id, levels = rownames(y)),
            decoy = 1L
          ) %>%
          pivot_wider(
            names_from = compound, values_from = decoy,
            names_prefix = "leading_edgge_",
            id_expand = TRUE, values_fill = 0L
          ) %>%
          column_to_rownames("gene_id") %>% {
            .[rownames(y), ]
          }
      }
    ),
    hm = pmap(
      list(pathway, mat, row_meta),
      \(x, y, z) {
        mat_max_abs <- max(abs(quantile(y, c(.025, .975), na.rm = TRUE)))
        Heatmap(
          y,
          name = "log2 fold change",
          col = circlize::colorRamp2(seq(from = -mat_max_abs, to = mat_max_abs, length.out = 51), paletteer::paletteer_c("pals::ocean.balance", n = 51)),
          cluster_rows = cluster_fun_eucl,
          cluster_columns = FALSE,
          right_annotation = HeatmapAnnotation(
            df = z,
            which = "row"
          )
        )
      }
    )
  )

dir.create(
  here("plots", "de_heatmaps"),
  showWarnings = FALSE
)

pwalk(
  de_pert_rescue_hm_data,
  \(pathway, hm, ...) {
    withr::with_pdf(
      file.path("plots", "de_heatmaps", paste0("de_heatmap_", pathway, ".pdf")),
      draw(hm),
      width = 6, height = 10
    )
  }
)
```

## Overrepresentation analysis of rescue gene sets

### Plotting single conditions as bars

```{r}
bar_plot_colors <- c(
  up = "#d73027",
  down = "#4575b4"
)

fgsea_overrep_data <- fgsea_overrep %>%
  separate_wider_delim(
    pathway,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge"
  ) %>%
  mutate(
    # sign = if_else(NES > 0, "pos", "neg"),
    signed_padj = -if_else(perturbed == "up", 1, -1) * log10(padj_fgsea),
    pathway_db = paste(str_sub(database, 1, 1), pathway, sep = "_"),
    pathway_db_direction = paste(perturbed, pathway_db, sep = "_")
  )

ps <- fgsea_overrep_data %>%
  group_nest(
    compound, rescue_definition
  ) %>%
  mutate(
    p = pmap(
      list(compound, data, rescue_definition),
      \(x, y, z) {
        y %>%
          filter(
            padj_fgsea < .01
          ) %>%
          mutate(
            across(pathway_db_direction, \(x) fct_reorder(x, signed_padj))
          ) %>%
          ggplot(
            aes(
              x = signed_padj,
              y = pathway_db_direction,
              fill = perturbed
            )
          ) +
          geom_col() +
          geom_text(
            aes(
              x = if_else(perturbed == "up", -.05, .05),
              hjust = if_else(perturbed == "up", 1, 0),
              label = pathway_db
            ),
            size = 2
          ) +
          scale_fill_manual(
            values = alpha(bar_plot_colors, .7)
          ) +
          theme(
            axis.text.y = element_blank(),
            panel.grid.major.y = element_blank(),
          ) +
          labs(
            x = "Signed adjusted p-value",
            y = NULL,
            title = paste0("Pathways enriched in genes ", z, " with ", x),
            subtitle = "FDR < 0.01"
          )
      }
    )
  )

dir.create(
  here("plots", "overrep"),
  showWarnings = FALSE
)

pwalk(
  ps,
  \(compound, p, rescue_definition, ...) {
    n <- nrow(layer_data(p))
    h <- 2 + .1 * n
    ggsave(
      file.path("plots", "overrep", paste0("overrep_", rescue_definition, "_", compound, "_bars.pdf")),
      p, width = 8, height = h
    )
  }
)

```

## vs PTX

```{r}
fgsea_overrep_vs_ptx_pert <- fgsea_overrep %>%
  filter(
    rescue_definition != "perturbed"
  ) %>%
  separate_wider_delim(
    pathway,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge"
  ) %>%
  mutate(
    signed_padj = -if_else(perturbed == "up", 1, -1) * log10(padj)
  ) %>%
  power_inner_join(
    fgsea_res_single_treatments %>%
      filter(contrast == "PTX", metric == "signed_padj") %>%
      mutate(
        signed_padj = -if_else(NES > 0, 1, -1) * log10(padj)
      ),
    by = c("database", "pathway"),
    suffix = c("_overrep", "_PTX"),
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  )

p <- fgsea_overrep_vs_ptx_pert %>%
  ggplot(
    aes(
      x = signed_padj_PTX,
      y = signed_padj_overrep
    )
  ) +
  geom_point(
    aes(
      text = pathway
    )
  ) +
  facet_grid(
   vars(compound), vars(rescue_definition)
  )

htmlwidgets::saveWidget(
  ggplotly(p),
  file.path("plots", "overrep", "overrep_vs_fgsea_ptx_scatter_padj.html")
)
```

## Plotting intersections of rescue gene sets

Have to be careful about perturbed `both_up_down` class, messes up signed_padj

### intersections bar charts

```{r}
fgsea_intersection_overrep_data <- fgsea_intersection_overrep %>%
  separate_wider_delim(
    pathway,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge"
  ) %>%
  mutate(
    # sign = if_else(NES > 0, "pos", "neg"),
    signed_padj = -if_else(perturbed %in% c("both_up_down", "up"), 1, -1) * log10(padj_fgsea),
    pathway_db = paste(str_sub(database, 1, 1), pathway, sep = "_"),
    pathway_db_direction = paste(perturbed, pathway_db, sep = "_")
  )

fgsea_intersection_overrep_data %>%
  filter(
    rescue_definition == "significant_opposite",
    pathway == "SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE"
  ) %>%
  View()

ps <- fgsea_intersection_overrep_data %>%
  filter(perturbed %in% c("up", "down")) %>%
  group_nest(
    rescue_type, rescue_definition
  ) %>%
  mutate(
    p = pmap(
      list(rescue_type, rescue_definition, data),
      \(x, y, z) {
        z %>%
          filter(
            padj_fgsea < .01
          ) %>%
          mutate(
            across(pathway_db_direction, \(x) fct_reorder(x, signed_padj))
          ) %>%
          ggplot(
            aes(
              x = signed_padj,
              y = pathway_db_direction,
              fill = perturbed
            )
          ) +
          geom_col() +
          geom_text(
            aes(
              x = if_else(perturbed == "up", -.05, .05),
              hjust = if_else(perturbed == "up", 1, 0),
              label = pathway_db
            ),
            size = 2
          ) +
          scale_fill_manual(
            values = alpha(bar_plot_colors, .7)
          ) +
          theme(
            axis.text.y = element_blank(),
            panel.grid.major.y = element_blank(),
          ) +
          labs(
            x = "Signed adjusted p-value",
            y = NULL,
            title = paste0("Pathways enriched in genes rescued by ", x),
            subtitle = paste("FDR < 0.01", y)
          )
      }
    )
  )

dir.create(
  here("plots", "overrep"),
  showWarnings = FALSE
)

pwalk(
  ps,
  \(rescue_definition, rescue_type, p, ...) {
    ggsave(
      file.path("plots", "overrep", paste0("overrep_intersection_", rescue_definition, "_", rescue_type, "_bars.pdf")),
      p, width = 8, height = 8
    )
  }
)

```

### Dot plot


```{r}

fgsea_intersection_dotplot_data <- fgsea_intersection_overrep_data %>%
  mutate(
    padj_fgsea_bin = cut(
      padj_fgsea,
      breaks = c(0, .01, .05, 1),
      labels = c("FDR < 0.01", "0.01 < FDR < 0.05", "ns")
    )
  ) %>%
  filter(
    rescue_type %in% c("GNE-495 only", "JNK-inh only"),
    perturbed %in% c("up", "down")
  ) %>%
  # inner_join(
  #   fgsea_intersection_overrep_data_single,
  #   by = c("rescue_definition", "rescue_type", "pathway_db")
  # ) %>%
  # mutate(
  #   directed_estimate = if_else(direction == "up", estimate, -estimate)
  # ) %>%
  group_nest(
    rescue_definition
  ) %>%
  mutate(
    p = map(
      data,
      \(x) {
        d <- group_by(
          x,
          pathway_db
        ) %>%
          filter(
            any(padj_fgsea < .01)
          ) %>%
          ungroup()
        dmats <- d %>%
          select(rescue_type, pathway_db, perturbed, estimate) %>%
          arrange(pathway_db) %>%
          split(.$perturbed) %>%
          map(
            \(y) select(y, -perturbed) %>%
              pivot_wider(
                names_from = rescue_type,
                values_from = estimate
              ) %>%
              column_to_rownames("pathway_db") %>%
              as.matrix() %>%
              dist()
          )
        dmat <- reduce(dmats, `+`)
        clust <- hclust(dmat, method = "average") %>%
          reorder(dmat, method = "OLO")
        # browser()
        ggplot(
          d %>%
            mutate(
              pathway_db = factor(pathway_db, levels = clust$labels[clust$order])
            ),
          aes(
            x = rescue_type,
            y = pathway_db,
            color = estimate,
            size = padj_fgsea_bin
          )
        ) +
          geom_point(shape = 16) +
          scale_color_viridis_c(trans = "pseudo_log") +
          scale_size_manual(
            values = c("FDR < 0.01" = 4, "0.01 < FDR < 0.05" = 2, "ns" = 1)
          ) +
          # scale_color_distiller(palette = "RdBu") +
          facet_wrap(~perturbed)
      }
    )
  )
```

Check what genes are in JNK-inh rescued mitosis related pathways

```{r}
x <- fgsea_intersection_overrep_data %>%
  arrange(p.value_fgsea) %>%
  filter(
    rescue_definition == "significant_opposite",
    rescue_type == "JNK-inh only"
  )

x %>%
  filter(
    pathway == "METALLOPROTEASE_DUBS"
  ) %>%
  chuck("overlapGenes", 1) %>%
  {gene_id_gene_symbol_map[.]}

gene_id_gene_symbol_map[x$overlapGenes[[1]]]
```



```{r}

flag_similar_adjacent_leaves <- function(clust, sim_matrix, threshold) {
  # Get the dendrogram representation
  dend <- as.dendrogram(clust)

  # Function to find adjacent leaves and check their similarity
  find_adjacent_pairs <- function(node, parent_leaves = NULL) {
    if (is.leaf(node)) {
      # For leaf nodes, return the label
      return(list(leaves = labels(node), pairs = data.frame()))
    } else {
      # For internal nodes, process left and right children
      left_result <- find_adjacent_pairs(node[[1]], parent_leaves)
      right_result <- find_adjacent_pairs(node[[2]], left_result$leaves)

      # Combine leaves from both children
      all_leaves <- c(left_result$leaves, right_result$leaves)

      # Identify pairs of leaves that are adjacent across the left-right boundary
      # (the rightmost leaf of left subtree and leftmost leaf of right subtree)
      boundary_pairs <- data.frame(
        leaf1 = tail(left_result$leaves, 1),
        leaf2 = head(right_result$leaves, 1),
        stringsAsFactors = FALSE
      )

      # Calculate similarity for these boundary pairs
      if (nrow(boundary_pairs) > 0) {
        boundary_pairs$similarity <- sim_matrix[boundary_pairs$leaf1, boundary_pairs$leaf2]
        # Flag pairs below threshold
        boundary_pairs <- boundary_pairs[boundary_pairs$similarity <= threshold, ]
      }

      # Combine all pairs found
      all_pairs <- rbind(left_result$pairs, right_result$pairs, boundary_pairs)

      return(list(leaves = all_leaves, pairs = all_pairs))
    }
  }

  # Start the recursive search
  result <- find_adjacent_pairs(dend)

  return(result$pairs)
}


```

```{r}
msigdbr_of_interest_sim_trans <- msigdbr_of_interest_sim
attr(msigdbr_of_interest_sim_trans, "Labels") <- distinct(
  msigdbr_of_interest, gs_name, database, pathway
) %>%
  transmute(
    pathway_db = paste(str_sub(database, 1, 1), pathway, sep = "_"),
    gs_name
  ) %>%
  slice(match(labels(msigdbr_of_interest_sim_trans), gs_name)) %>%
  pull(pathway_db)
msigdbr_of_interest_sim_trans <- as.matrix(msigdbr_of_interest_sim_trans)

fgsea_intersection_overrep_data_single <- fgsea_intersection_overrep_data %>%
  filter(
    perturbed %in% c("up", "down")
  ) %>%
  group_by(
    rescue_definition, rescue_type, database, pathway_db, pathway
  ) %>%
  summarize(
    direction = case_when(
      all(padj_fgsea >= .01) ~ "none",
      padj_fgsea[1] < .2 * padj_fgsea[2] | padj_fgsea[2] < .2 * padj_fgsea[1] ~ perturbed[which.min(padj_fgsea)],
      TRUE ~ "both"
    ),
    estimate = estimate[order(padj_fgsea)[1]],
    padj_fgsea = min(padj_fgsea),
    # estimate = estimate[which.min(padj_fgsea)],
    .groups = "drop"
  )

fgsea_blacklist <- c(
  "HCMV EARLY EVENTS",
  "HCMV INFECTION",
  "INFECTIOUS DISEASE"
)

fgsea_intersection_dotplot_data_single <- fgsea_intersection_overrep_data_single %>%
  mutate(
    padj_fgsea_bin = cut(
      padj_fgsea,
      breaks = c(0, .001, .01, 1),
      labels = c("FDR < 0.001", "FDR < 0.01", "ns")
    ),
    # across(
    #   pathway_db,
    #   \(x) str_remove_all(
    #     x, coll("_")
    #   ) %>%
    #     str_sub(start = 3)
    # )
  ) %>%
  filter(
    !database %in% c("KEGG"),
    rescue_type %in% c("GNE-495 only", "JNK-inh only"),
  ) %>%
  mutate(
    directed_estimate = if_else(direction == "up", estimate, -estimate)
  ) %>%
  group_nest(
    rescue_definition
  ) %>%
  crossing(
    p_threshold = c(0.005, 0.001, .01)
  ) %>%
  mutate(
    res = map2(
      data, p_threshold,
      \(x, y) {
        # browser()
        d <- group_by(
          x,
          pathway_db
        ) %>%
          filter(
            any(padj_fgsea < y)
          ) %>%
          ungroup()
        dwide <- d %>%
          select(rescue_type, pathway_db, directed_estimate) %>%
          arrange(pathway_db) %>%
          pivot_wider(
            names_from = rescue_type,
            values_from = directed_estimate
          )
        dmat <- dwide %>%
          column_to_rownames("pathway_db") %>%
          as.matrix() %>%
          dist()
        # browser()
        clust <- hclust(dmat, method = "average") %>%
          reorder(dmat, method = "OLO")
        row_labels <- clust$labels[clust$order] %>% {
          set_names(
            str_replace_all(., coll("_"), " ") %>%
              str_sub(start = 2),
            .
          )
        }
        sim_hm <- msigdbr_of_interest_sim_trans[
          rev(clust$labels[clust$order]), rev(clust$labels[clust$order])
        ] %>%
          pheatmap(
            cluster_rows = FALSE,
            cluster_cols = FALSE
          )
        est_perc <- quantile(d$directed_estimate, c(.05, .95))
        abs_max_est_perc <- max(abs(est_perc))
        p <- ggplot(
          d %>%
            mutate(
              pathway_db = factor(pathway_db, levels = clust$labels[clust$order])
            ),
          aes(
            x = rescue_type,
            y = pathway_db,
            color = directed_estimate,
            size = padj_fgsea_bin
          )
        ) +
          geom_point(shape = 16) +
          # scale_color_viridis_c(trans = "pseudo_log") +
          # scale_color_distiller(
          #   limits = c(-abs_max_est_perc, abs_max_est_perc),
          #   palette = "RdYlBu", trans = "pseudo_log",
          #   oob = scales::squish
          # ) +
          paletteer::scale_color_paletteer_c(
            palette = "pals::ocean.balance",
            limits = c(-abs_max_est_perc, abs_max_est_perc),
            oob = scales::squish
          ) +
          scale_size_manual(
            values = rev(c("FDR < 0.001" = 6, "FDR < 0.01" = 4, "ns" = 2))
          ) +
          scale_x_discrete(position = "top") +
          scale_y_discrete(
            breaks = names(row_labels),
            labels = unname(row_labels)
          ) +
          theme(
            axis.text.x = element_text(angle = 45, hjust = 0)
          ) +
          labs(
            x = "Rescue by",
            y = NULL,
            color = "Rescue of\ngenes",
            size = NULL
          )
        list(
          p = p,
          sim_hm = sim_hm
        ) %>%
          map(list)
      }
    )
  ) %>%
  select(-data) %>%
  unnest_wider(
    res
  )

flag_similar_adjacent_leaves(clust, msigdbr_of_interest_sim_trans, .6)

pwalk(
  fgsea_intersection_dotplot_data_single,
  \(p, rescue_definition, p_threshold, sim_hm, ...) {
    ggsave(
      file.path("plots", "overrep", paste0("overrep_intersection_dotplot_", rescue_definition, "_", p_threshold, ".pdf")),
      p[[1]], width = 9.8, height = 9
    )
    withr::with_pdf(
      file.path("plots", "overrep", paste0("overrep_intersection_dotplot_", rescue_definition, "_", p_threshold, "_similarity.pdf")),
      draw(sim_hm[[1]]),
      width = 9.8, height = 9
    )
  }
)
 ```



### intersections versus each other and vs PTX


```{r}
fgsea_intersection_vs_ptx <- fgsea_intersection_overrep_data %>%
  power_inner_join(
    fgsea_res_pert %>%
      separate_wider_delim(
        pathway,
        delim = "_",
        names = c("database", "pathway"),
        too_many = "merge"
      ) %>%
      mutate(
        signed_padj = -sign(NES) * log10(padj)
      ) %>%
      filter(contrast == "PTX", metric == "signed_padj"),
    by = c("database", "pathway"),
    suffix = c("_rescue", "_ptx"),
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  )

ps <- fgsea_intersection_vs_ptx %>%
  filter(perturbed %in% c("up", "down")) %>%
  group_nest(
    rescue_definition
  ) %>%
  rowwise() %>%
  mutate(
    p = list(
    ggplot(
      data,
      aes(
        signed_padj_ptx,
        signed_padj_rescue,
        color = padj_fgsea < .05
      )
    ) +
    geom_point(
      aes(
        text = paste(
          pathway,
          database,
          "\n",
          paste(
            "padj rescue:", signif(padj_fgsea, 2),
            "overlap:", overlap,
            "gene set size:", size_ptx,
            "odds ratio rescue:", signif(estimate, 2),
            "padj ptx:", signif(padj, 2),
            "NES:", signif(NES, 2)
          )
        )
      ),
      alpha = .5,
      shape = 16
    ) +
    scale_color_manual(
      values = c(
        `TRUE` = "red",
        `FALSE` = "black"
        # up = "red",
        # down = "blue",
        # both_up_down = "black"
      )
    ) +
    ggh4x::facet_wrap2(~rescue_type, scales = "free", axes = "y")
    )
  ) %>%
  ungroup()

pwalk(
  ps,
  \(rescue_definition, p, ...) {
    ggsave(
      file.path("plots", "overrep", paste0("overrep_intersection_vs_ptx_", rescue_definition, ".pdf")),
      p, width = 8, height = 8
    )
    htmlwidgets::saveWidget(
      ggplotly(p),
      file.path("plots", "overrep", paste0("overrep_intersection_vs_ptx_", rescue_definition, ".html"))
    )
  }
)

fgsea_intersection_overrep_gne_vs_jnk <- fgsea_intersection_overrep_data %>%
  filter(
    rescue_definition == "significant_opposite",
    perturbed %in% c("up", "down"),
    rescue_type %in% c("GNE-495 only", "JNK-inh only")
  ) %>%
  select(
    perturbed, rescue_type, database, pathway, pathway_db, pathway_db_direction,
    signed_padj, estimate
  ) %>%
  pivot_wider(
    names_from = rescue_type,
    values_from = c(signed_padj, estimate)
  )

p <- fgsea_intersection_overrep_gne_vs_jnk %>%
  ggplot(
    aes(
      x = `signed_padj_JNK-inh only`,
      y = `signed_padj_GNE-495 only`,
      text = pathway_db
    )
  ) +
  geom_point()

p

plotly::ggplotly(p)
```


### Heatmap of intersection enrichments

```{r}
intersection_overrep_hm_rescure_type_order <- c(
  "both", "GNE-495 only", "JNK-inh only", "no rescue"
)

de_pert_hm_data <- de_long_selected %>%
  filter(
    contrast %in% fgsea_hm_conditions
  ) %>%
  mutate(
    signed_padj = -sign(logFC) * log10(FDR_min_zero)
  )

library(ComplexHeatmap)
de_pert_rescue_hm_data <- tibble(
  pathway = c(
    "RECYCLING_PATHWAY_OF_L1",
    "REGULATION_OF_EXPRESSION_OF_SLITS_AND_ROBOS",
    "SRP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE",
    # "EPH_EPHRIN_SIGNALIN",
    "EPHB_MEDIATED_FORWARD_SIGNALING",
    "P38MAPK_EVENTS",
    "RHO_GTPASE_EFFECTORS",
    "RAC1_PATHWAY",
    "RHO_GTPASES_ACTIVATE_ROCKS"
  )
) %>%
  mutate(
    genes = map(
      pathway,
      \(x) msigdbr_of_interest %>%
        filter(
          pathway == x
        ) %>%
        distinct(gene_symbol, ensembl_gene) %>%
        semi_join(
          de_pert_hm_data,
          by = c("ensembl_gene" = "gene_id")
        )
    ),
    mat = map(
      genes,
      \(x) de_pert_hm_data %>%
        semi_join(
          x,
          by = c("gene_id" = "ensembl_gene")
        ) %>%
        select(gene_id, contrast, logFC) %>%
        pivot_wider(names_from = contrast, values_from = logFC, values_fill = 0) %>%
        column_to_rownames("gene_id") %>%
        as.matrix() %>% {
          .[x$ensembl_gene, fgsea_hm_conditions]
        }
    ),
    row_meta = map2(
      pathway, genes,
      \(x, y) {
        power_left_join(
          y,
          fgsea_res_rescue_data %>%
            filter(pathway == x) %>%
            select(compound, gene_id = leadingEdge) %>%
            unchop(gene_id) %>%
            mutate(decoy = 1L) %>%
            pivot_wider(
              names_from = compound, values_from = decoy,
              names_prefix = "leading_edge_",
              values_fill = 0L
            ),
          by = c("ensembl_gene" = "gene_id"),
          check = check_specs(
            unmatched_keys_right = "warn",
            duplicate_keys_right = "warn",
            duplicate_keys_left = "warn"
          )
        ) %>%
        mutate(
          across(
            starts_with("leading_edge_"),
            \(x) if_else(is.na(x), 0L, x)
          )
        )
      }
    ),
    hm = pmap(
      list(pathway, mat, row_meta, genes),
      \(x, y, z, g) {
        # browser()
        mat_max_abs <- max(abs(quantile(y, c(.025, .975), na.rm = TRUE)))
        Heatmap(
          y,
          name = "log2 fold change",
          col = circlize::colorRamp2(seq(from = -mat_max_abs, to = mat_max_abs, length.out = 51), paletteer::paletteer_c("pals::ocean.balance", n = 51)),
          cluster_rows = cluster_fun_eucl,
          cluster_columns = FALSE,
          row_labels = g$gene_symbol,
          right_annotation = HeatmapAnnotation(
            df = z %>%
              select(ensembl_gene, starts_with("leading_edge_")) %>%
              column_to_rownames("ensembl_gene") %>%
              as.matrix(),
            which = "row"
          )
        )
      }
    )
  )

dir.create(
  here("plots", "de_heatmaps"),
  showWarnings = FALSE
)

pwalk(
  de_pert_rescue_hm_data,
  \(pathway, hm, ...) {
    withr::with_pdf(
      file.path("plots", "de_heatmaps", paste0("de_heatmap_", pathway, ".pdf")),
      draw(hm),
      width = 6, height = 10
    )
  }
)

```

## TopGO

check if sometimes the same pathway is enriched in both up and downregulated genes

```{r}
topgo_res_pert %>%
  pivot_wider(
    names_from = direction,
    values_from = -c(contrast, direction, ontology, GO.ID, Term, term_unique, Annotated)
  ) %>%
  filter(
    fisher_double_down < .05,
    fisher_double_up < .05
  ) %>%
  View()
```

Make directed topgo results object by taking the log10 p-value
and the sign corresponds to the direction of the enrichment

If the same pathway is enriched in both directions, the pathway
occurs twice with different signs

```{r}
topgo_res_pert_directed <- topgo_res_pert %>%
  mutate(
    log_odds_ratio = {Significant / Expected} %>%
      if_else(
        !is.finite(.) | . == 0,
        sign(.) * min(.[. > 0]) * .5,
        .
      ) %>%
      log10()
  ) %>%
  group_by(
    contrast, ontology, GO.ID, Term, term_unique, Annotated
  ) %>%
  slice(
    if (sum(fisher_double < .05) < 2) {
      order(fisher_double)[1]
    } else {
      order(fisher_double)
    }
  ) %>%
  mutate(
    picked_direction = if (n() < 2) direction else "both",
    signed_padj = -sign(log_odds_ratio) * log10(fisher_double)
  ) %>%
  ungroup()

```


```{r}
all_contrasts <- topgo_res_pert_directed$contrast %>%
  unique()

contrast_plotting_pairs <- bind_rows(
  tibble(
    contrast_1 = "PTX",
    contrast_2 = all_contrasts[
      str_ends(all_contrasts, fixed("+ PTX")) |
      str_ends(all_contrasts, fixed("+ PTX vs PTX"))
    ]
  ),
  tibble(
    contrast_1 = "GNE-495 + PTX",
    contrast_2 = "JNK-inh + PTX"
  )
)
```

```{r}
topgo_res_ptx_vs_combo <- topgo_res_pert_directed %>%
  filter(
    ontology == "BP"
  ) %>% {
    d <- .
    power_inner_join(
      contrast_plotting_pairs,
      d,
      by = join_by(contrast_1 == contrast),
      suffix = c("", "_1"),
      check = check_specs(
        unmatched_keys_left = "warn"
      )
    ) %>%
    power_inner_join(
      d,
      by = join_by(contrast_2 == contrast, ontology, GO.ID, Term, term_unique, Annotated),
      suffix = c("_1", "_2"),
      check = check_specs(
        unmatched_keys_left = "warn"
      )
    )
  } %>%
  mutate(
    sig_class = case_when(
      fisher_double_1 < .05 & fisher_double_2 < .05 ~ "both",
      fisher_double_1 < .05 ~ "1",
      fisher_double_2 < .05 ~ "2",
      TRUE ~ "none"
    )
  )

sig_class_colors <- c(
  "both" = "purple",
  "1" = "red",
  "2" = "blue",
  "none" = "grey"
)

ps <- topgo_res_ptx_vs_combo %>%
  group_nest(contrast_1, contrast_2) %>%
  mutate(
    p = pmap(
      .,
      \(contrast_1, contrast_2, data, ...) {
        ggplot(
          data %>%
            arrange(match(sig_class, names(sig_class_colors))),
          aes(
            x = signed_padj_1,
            y = signed_padj_2
          )
        ) +
        geom_point(
          aes(
            color = sig_class,
            text = Term
          ),
          alpha = .8
        ) +
        scale_color_manual(
          values = sig_class_colors
        ) +
        labs(
          x = paste("Enrichment", contrast_1),
          y = paste("Enrichment", contrast_2)
        )
      }
    )
  )

pwalk(
  ps,
  \(contrast_1, contrast_2, p, ...) {
    ggsave(
      file.path("plots", paste0("topgo_", contrast_1, "_vs_", contrast_2, "_scatter_padj.pdf")),
      p, width = 8, height = 8
    )
    htmlwidgets::saveWidget(
      ggplotly(p),
      file.path("plots", paste0("topgo_", contrast_1, "_vs_", contrast_2, "_scatter_padj.html"))
    )
  }
)


```

## Create Excel table

1. Differential expression
2. FGSEA on raw perturbations
3. FGSEA on rescue vectors
4. Overrepresentation on rescue gene sets
5. Overrepresentation on rescue gene set intersections

```{r}
gene_id_symbol_map <- de_long_selected %>%
  distinct(gene_id, gene_name)

gene_id_symbol_vec <- with(
  gene_id_symbol_map,
  set_names(gene_name, gene_id)
)

excel_de <- de_long_selected %>%
  select(-FDR_min_zero)

excel_fgsea_raw <- fgsea_res_pert %>%
  filter(metric == "signed_padj") %>%
  separate_wider_delim(
    pathway,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge"
  ) %>%
  select(-c(metric, log2err, ES)) %>%
  mutate(
    contrast = if_else(str_ends(contrast, fixed("vs PTX")), contrast, paste(contrast, "vs DMSO")),
    leadingEdge = map_chr(
      leadingEdge,
      \(x) str_flatten(gene_id_symbol_vec[x], collapse = ",")
    )
  )

excel_fgsea_rescue <- fgsea_res_rescue %>%
  filter(metric == "signed_padj_rescue_clamped") %>%
  separate_wider_delim(
    pathway,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge"
  ) %>%
  select(-c(metric, log2err, ES)) %>%
  mutate(
    leadingEdge = map_chr(
      leadingEdge,
      \(x) str_flatten(gene_id_symbol_vec[x], collapse = ",")
    )
  )

excel_overrep_rescue <- fgsea_overrep %>%
  filter(rescue_definition == "significant_opposite") %>%
  separate_wider_delim(
    pathway,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge"
  ) %>%
  transmute(
    compound, ptx_direction = perturbed,
    database, pathway,
    pval = p.value_fisher, padj = padj_fisher,
    odds_ratio = estimate,
    size, overlap,
    leadingEdge = map_chr(
      overlapGenes,
      \(x) str_flatten(gene_id_symbol_vec[x], collapse = ",")
    )
  )

excel_overrep_intersections <- fgsea_intersection_overrep %>%
  filter(rescue_definition == "significant_opposite") %>%
  separate_wider_delim(
    pathway,
    delim = "_",
    names = c("database", "pathway"),
    too_many = "merge"
  ) %>%
  transmute(
    rescue_type, ptx_direction = perturbed,
    database, pathway,
    pval = p.value_fisher, padj = padj_fisher,
    odds_ratio = estimate,
    size, overlap,
    leadingEdge = map_chr(
      overlapGenes,
      \(x) str_flatten(gene_id_symbol_vec[x], collapse = ",")
    )
  )
```

```{r}
library(openxlsx)

openxlsx::write.xlsx(
  list(
    "Differential expression" = excel_de,
    "FGSEA perturbations" = excel_fgsea_raw,
    "FGSEA rescue vectors" = excel_fgsea_rescue,
    "Overrep rescue gene sets" = excel_overrep_rescue,
    "Overrep rescue intersections" = excel_overrep_intersections
  ),
  file = here("results", "gene_enrichment_analysis_results.xlsx"),
  asTable = TRUE
)
```


## RAGs and ion channels

```{r}
library(biomaRt)

# mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")

# Install packages if you haven't already
# install.packages("httr")
# install.packages("jsonlite")

library(httr)
library(jsonlite)

# Define a function to query mygene.info for a given gene symbol (or alias)
query_mygene <- function(query_term, scopes = "symbol,alias", species = "mouse", fields = "ensembl.gene") {
  base_url <- "http://mygene.info/v3/query"

  # Build the query parameters
  params <- list(
    q = query_term,
    scopes = scopes,
    species = species,
    fields = fields
  )

  # Make the GET request
  res <- POST(url = base_url, body = params, encode = "json")

  # Check for a successful request (status code 200)
  if (status_code(res) == 200) {
    # Parse the JSON content
    data <- fromJSON(content(res, as = "text", encoding = "UTF-8"))
    return(data)
  } else {
    stop("Query failed with status: ", status_code(res))
  }
}

mouse_gene_id_mapping <- query_mygene(
  unique(rag_ion_channel_genes$gene_symbol_mouse)
) %>%
  transmute(
    gene_symbol_mouse = query,
    ensembl_gene_id_mouse = map(
      ensembl,
      \(x) if (class(x) == "list")
        tibble(gene = x[[1]])
      else
        x
    )
  ) %>%
  unnest(ensembl_gene_id_mouse)

mouse_homologue_mapping <- getHomologs(
  unique(mouse_gene_id_mapping$gene),
  "mus_musculus",
  "homo_sapiens"
)

human_gene_id_symbol_map <- getBM(
  attributes = c("ensembl_gene_id", "hgnc_symbol"),
  filters = "ensembl_gene_id",
  values = unique(na.omit(mouse_homologue_mapping$hsapiens_homolog_ensembl_gene)),
  mart = human
)

rag_ion_channel_genes_mapped <- rag_ion_channel_genes %>%
  left_join(
    mouse_gene_id_mapping %>%
      dplyr::distinct(gene_symbol_mouse, ensembl_gene_id_mouse = gene),
    by = "gene_symbol_mouse"
  ) %>%
  left_join(
    mouse_homologue_mapping %>%
      dplyr::select(
        ensembl_gene_id_mouse = ensembl_gene_id,
        ensembl_gene_id_human = hsapiens_homolog_ensembl_gene
      ),
    by = "ensembl_gene_id_mouse"
  ) %>%
  left_join(
    human_gene_id_symbol_map %>%
      dplyr::select(
        ensembl_gene_id_human = ensembl_gene_id,
        hgnc_symbol
      ),
    by = "ensembl_gene_id_human"
  ) %>%
  mutate(
    gene_display_name = coalesce(
      hgnc_symbol,
      gene_symbol_mouse,
      ensembl_gene_id_human,
      ensembl_gene_id_mouse
    )
  ) %>%
  group_by(gene_display_name) %>%
  mutate(
    unique_gene_id = if (n() > 1)
      paste0(gene_display_name, "-", seq_len(n()))
    else
      gene_display_name
  ) %>%
  ungroup()

```


```{r}
rag_hm_conditions <- c(
  "PTX", "GNE-495", "JNK-inh",
  "GNE-495 + PTX", "JNK-inh + PTX"
)

rag_ion_channel_genes_valid <- rag_ion_channel_genes_mapped %>%
  drop_na(
    ensembl_gene_id_human
  ) %>%
  filter(ensembl_gene_id_human %in% de_long_selected$gene_id)

de_rag_hm_data <- de_long_selected %>%
  filter(
    contrast %in% rag_hm_conditions
  ) %>%
  inner_join(
    rag_ion_channel_genes_valid,
    by = c("gene_id" = "ensembl_gene_id_human")
  ) %>%
  mutate(
    signed_padj = -sign(logFC) * log10(FDR_min_zero)
  )

de_rag_hm_mat <- de_rag_hm_data %>%
  dplyr::select(
    unique_gene_id, contrast, logFC
  ) %>%
  pivot_wider(names_from = contrast, values_from = logFC) %>%
  column_to_rownames("unique_gene_id") %>%
  as.matrix() %>% {
    .[
      rag_ion_channel_genes_valid$unique_gene_id,
      rag_hm_conditions
    ]
  }

library(ComplexHeatmap)
hm <- Heatmap(
  de_rag_hm_mat,
    # t() %>%
    # scale() %>%
    # t(),
  cluster_rows = cluster_fun_eucl,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  row_split = rag_ion_channel_genes_valid$gene_set
)

withr::with_pdf(
  file.path("plots", "de_heatmaps", "de_heatmap_rag_ion_channels_lfc.pdf"),
  draw(hm),
  width = 6, height = 10
)

de_rag_hm_mat <- de_rag_hm_data %>%
  dplyr::select(
    unique_gene_id, contrast, signed_padj
  ) %>%
  pivot_wider(names_from = contrast, values_from = signed_padj) %>%
  column_to_rownames("unique_gene_id") %>%
  as.matrix() %>% {
    .[
      rag_ion_channel_genes_valid$unique_gene_id,
      rag_hm_conditions
    ]
  }

library(ComplexHeatmap)
hm <- Heatmap(
  # de_rag_hm_mat,
  # Clamp to -10 to 10
  de_rag_hm_mat %>% {
    pmax(pmin(., 10), -10)
  },
  name = "-sign(log2FC) * log10(padj)",

    # t() %>%
    # scale() %>%
    # t(),
  col = circlize::colorRamp2(seq(-10, 10, length.out = 51), paletteer::paletteer_c("pals::ocean.balance", n = 51)),
  cluster_rows = cluster_fun_eucl,
  cluster_columns = FALSE,
  show_row_names = FALSE,
  row_split = rag_ion_channel_genes_valid$gene_set
)

withr::with_pdf(
  file.path("plots", "de_heatmaps", "de_heatmap_rag_ion_channels_padj.pdf"),
  draw(hm),
  width = 6, height = 10
)

hm <- Heatmap(
  de_rag_hm_mat,
  # Clamp to -10 to 10
  # de_rag_hm_mat %>% {
  #   pmax(pmin(., 10), -10)
  # },
  name = "-sign(log2FC) * log10(padj)",

    # t() %>%
    # scale() %>%
    # t(),
  col = circlize::colorRamp2(seq(-10, 10, length.out = 51), paletteer::paletteer_c("pals::ocean.balance", n = 51)),
  cluster_rows = cluster_fun_eucl,
  cluster_columns = FALSE,
  show_row_names = FALSE,
  row_split = rag_ion_channel_genes_valid$gene_set
)

withr::with_pdf(
  file.path("plots", "de_heatmaps", "de_heatmap_rag_ion_channels_padj2.pdf"),
  draw(hm),
  width = 6, height = 10
)
```

Overlap of RAGs and ion channels with GNE-495-only rescue genes


```{r}



```


### INDRA plots

```{r}
idnra_res_plot_data <- indra_res %>%
  filter(
    rescue_type %in% c(
      "GNE-495 only", "JNK-inh only"
    ),
    str_starts(query_type, fixed("indra"))
  ) %>%
  mutate(
    signed_q = if_else(
      perturbed == "down",
      1, -1
    ) * log10(q)
  ) %>%
  separate_wider_delim(
    curie,
    delim = ":",
    names = c("database", "id")
  ) %>%
  filter(
    database %in% c("fplx", "hgnc"),
    q < .01
  ) %>%
  mutate(
    sign = if_else(perturbed == "up", "pos", "neg")
  )

ps <- idnra_res_plot_data %>%
  group_nest(
    rescue_type, query_type
  ) %>%
  mutate(
    p = map2(
      rescue_type, data,
      \(x, y) {
        y %>%
          mutate(
            name = fct_reorder(name, signed_q)
          ) %>%
          ggplot(
            aes(
              x = signed_q,
              y = name,
              fill = sign
            )
          ) +
          geom_col() +
          geom_text(
            aes(
              x = if_else(sign == "pos", -.05, .05),
              hjust = if_else(sign == "pos", 1, 0),
              label = name
            ),
            size = 2
          ) +
          scale_fill_discrete(guide = "none", direction = -1) +
          theme(
            axis.text.y = element_blank(),
            panel.grid.major.y = element_blank(),
          ) +
          labs(
            x = "Signed log10 p-value",
            y = NULL,
            title = paste0(query_type, " analysis of genes rescued by ", x),
            subtitle = "FDR < 0.01"
          )
      }
    )
  )

pwalk(
  ps,
  \(rescue_type, query_type, p, ...) {
    ggsave(
      file.path("plots", paste0(query_type, "_", rescue_type, "_bars.pdf")),
      p, width = 8, height = 8
    )
  }
)

```


## Check if p38 target TFs are differentially expressed

```{r}
atf6_targets <- clipr::read_clip()

de_pert_hm_data_atf6 <- de_pert_hm_data %>%
  filter(
    gene_name %in% atf6_targets,
    gene_id %in% {
      de_pert_hm_data %>%
        filter(
          FDR < .05
        ) %>%
        pull(gene_id)
    }
  )

de_pert_hm_data_atf6_mat <- de_pert_hm_data_atf6 %>%
    dplyr::select(
      gene_id, contrast, logFC
    ) %>%
    pivot_wider(names_from = contrast, values_from = logFC, values_fill = 0) %>%
    column_to_rownames("gene_id") %>%
    as.matrix() %>%
    t() %>%
    scale() %>%
    t() %>% {
      .[, fgsea_hm_conditions]
    }

hm <- Heatmap(
  de_pert_hm_data_atf6_mat,
  cluster_rows = cluster_fun_eucl,
  cluster_columns = FALSE,
  show_row_names = TRUE
)
hm

pca_fun <- function(mat, meta) {
  pca <- prcomp(t(mat), center = TRUE, scale = FALSE)
  pca_x <- as_tibble(pca$x) %>%
    bind_cols(
      meta
    )
  list(
    x = pca_x,
    sd = broom::tidy(pca, matrix = "eigenvalues"),
    rotation = broom::tidy(pca, matrix = "rotation")
  )
}
de_pert_hm_data_atf6_pca <- pca_fun(
  de_pert_hm_data_atf6_mat,
  tibble(contrast = fgsea_hm_conditions)
)

p <- ggplot(
  de_pert_hm_data_atf6_pca$x,
  aes(
    x = PC1, y = PC3, color = contrast
  )
) +
  geom_point() +
  labs(
    x = paste0("PC1 (", round(de_pert_hm_data_atf6_pca$sd$percent[1] * 100, 2), "%)"),
    y = paste0("PC3 (", round(de_pert_hm_data_atf6_pca$sd$percent[3] * 100, 2), "%)")
  ) +
  theme_minimal()

hm <- Heatmap(
  de_pert_hm_data_atf6_pca$x %>%
    column_to_rownames("contrast") %>%
    # scale() %>%
    t(),
  cluster_rows = cluster_fun_eucl,
  cluster_columns = FALSE,
  show_row_names = TRUE
)
hm
```


VX-702
DORAMAPIMOD
LOSMAPIMOD
