---
title: "Peripheral neuropathy gene set analysis"
author: "Clemens Hug"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(data.table)
library(synExtra)
library(qs)
library(powerjoin)
library(openxlsx)

library(conflicted)
conflicts_prefer(dplyr::select)
conflicts_prefer(dplyr::filter)
conflicts_prefer(base::setdiff)
conflicts_prefer(dplyr::count)

synapser::synLogin()
syn <- synDownloader(normalizePath("~/data"), .cache = TRUE)
```

## Reading DEA results

DEA performed by Riki Kawaguchi from UCLA

```{r}
de_riki_raw <- syn("syn63854283") %>%
  readxl::read_excel()

de_riki_long <- de_riki_raw %>%
  select(
    -c(F, F.pValue, AvgExpr, matches("S[0-9]+_.+"))
  ) %>%
  pivot_longer(
    -c(gene_id, gene_name, gene_type, Description, chromosome, strand),
    names_to = c("metric", "contrast"),
    names_pattern = "(.*?)_(.*)",
    values_to = "value"
  ) %>%
  pivot_wider(
    names_from = metric,
    values_from = value
  ) %>%
  mutate(
    FDR_min_zero = if_else(FDR == 0, min(FDR[FDR > 0]) * .5, FDR),
    signed_padj = -sign(logFC) * log10(FDR_min_zero),
    across(
      gene_id,
      \(x) str_remove(x, "\\.[0-9]+")
    )
  )
  # separate_longer_delim(
  #   -c(gene_id, gene_name, gene_type, Description, chromosome, strand),
  #   delim = regex("(?<=^[^_]*)_")
  # )
```

And our own from combined analysis of the first and second sequencing run

```{r}
meta_combined <- syn("syn70780412") %>%
  read_csv()

de_long_combined <- syn("syn70777776") %>%
  read_csv() %>%
  replace_na(list(padj = 1)) %>%
  mutate(
    padj_min_zero = if_else(padj == 0, min(padj[padj > 0], na.rm = TRUE) * .5, padj),
    signed_padj = -sign(log2FoldChange) * log10(padj_min_zero)
  )
```

Using k = 0 (explicit batch effects only, no additional nuisance variables)
for this analysis

```{r}

contrast_map <- c(
  "PTX_vs_DMSO" = "PTX",
  "GNE.495.DMSO_vs_DMSO" = "GNE-495",
  "JNK.inh.DMSO_vs_DMSO" = "JNK-inh",
  "GNE.495.PTX_vs_DMSO" = "GNE-495 + PTX",
  "JNK.inh.PTX_vs_DMSO" = "JNK-inh + PTX",
  "GNE.495.PTX_vs_PTX" = "GNE-495 + PTX vs PTX",
  "JNK.inh.PTX_vs_PTX" = "JNK-inh + PTX vs PTX",
  "GNE-495 + JNK-inh",
  "GNE-495 + JNK-inh + PTX",
  "GNE-495 + JNK-inh + PTX vs PTX"
)

de_riki_long_selected <- de_riki_long %>%
  filter(contrast %in% names(contrast_map)) %>%
  mutate(
    contrast = contrast_map[contrast],
    perturbed = case_when(
      FDR > .05 ~ "not perturbed",
      logFC > 0 ~ "up",
      logFC < 0 ~ "down",
      TRUE ~ NA_character_
    )
  )

de_combined_selected <- de_long_combined %>%
  filter(
    name %in% contrast_map,
    k == 0
  ) %>%
  transmute(
    batch,
    contrast = name,
    gene_id, gene_name,
    logFC = log2FoldChange,
    PValue = pvalue,
    FDR = padj,
    FDR_min_zero = padj_min_zero,
    signed_padj,
    perturbed = case_when(
      FDR > .05 ~ "not perturbed",
      logFC > 0 ~ "up",
      logFC < 0 ~ "down",
      TRUE ~ NA_character_
    )
  )

de_long_selected_both <- bind_rows(
  de_riki_long_selected %>%
    mutate(batch = "batch_1_riki"),
  de_combined_selected
)

write_csv(
  de_long_selected_both,
  "de_long_selected_both.csv.gz"
)

feasible_genes <- de_long_selected_both %>%
  distinct(batch, gene_id)
```

## Reading counts

```{r}
counts_trans <- syn("syn70780244") %>%
  read_csv()

rlog_mat <- counts_trans %>%
  filter(type == "vst_batch_removed") %>%
  dplyr::select(-type) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()

rlog_scaled <- rlog_mat %>%
  t() %>%
  scale() %>%
  t()

dmso_means <- rlog_mat[
  ,
  meta_combined %>%
    filter(str_detect(condition, coll("dmso", ignore_case = TRUE))) %>%
    pull(sample_id_sequencing)
] %>%
  rowMeans()

# Scale so that mean is 0 at expression level of DMSO and scaled to SD of 1
rlog_scaled2 <- {
  rlog_mat - dmso_means
} %>%
  t() %>%
  scale(center = FALSE) %>%
  t()

```


## PTX rescue data

```{r}
clamp_symbols <- \(x, symbols, threshold) {
  case_when(
    abs(x) < threshold ~ "point",
    x > 0 ~ symbols[1],
    x < 0 ~ symbols[2],
    TRUE ~ NA_character_
  )
}

ptx_rescue_plot_data <- tribble(
  ~contrast_1, ~contrast_2,
  "PTX", "GNE-495 + PTX",
  "PTX", "JNK-inh + PTX"
) %>%
  power_inner_join(
    de_long_selected_both %>%
      select(batch, gene_id, logFC, PValue, FDR, FDR_min_zero, signed_padj, contrast),
    by = c("contrast_1" = "contrast"),
    check = check_specs(
      unmatched_keys_left = "warn"
    )
  ) %>%
  power_inner_join(
    de_long_selected_both %>%
      select(batch, gene_id, logFC, PValue, FDR, FDR_min_zero, signed_padj, contrast),
    by = c("contrast_2" = "contrast", "gene_id", "batch"),
    suffix = c("_1", "_2"),
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  mutate(
    across(
      starts_with("signed_padj"),
      list(
        clamped = \(x) if_else(
          abs(x) > if (cur_column() == "signed_padj_1") 25 else 100,
          sign(x) * if (cur_column() == "signed_padj_1") 25 else 100,
          x
        ),
        symbol = \(x) clamp_symbols(
          x,
          if (cur_column() == "signed_padj_1") c("right", "left") else c("up", "down"),
          if (cur_column() == "signed_padj_1") 25 else 100
        )
      )
    ),
    signed_padj_symbol = case_when(
      !xor(signed_padj_1_symbol == "point", signed_padj_2_symbol == "point") ~ "point",
      signed_padj_1_symbol != "point" ~ signed_padj_1_symbol,
      signed_padj_2_symbol != "point" ~ signed_padj_2_symbol,
      TRUE ~ NA_character_
    )
  )
```


```{r}
p <- ggplot(
  ptx_rescue_plot_data,
  aes(
    x = signed_padj_1_clamped,
    y = signed_padj_2_clamped,
    shape = signed_padj_symbol
  )
) +
  geom_abline() +
  geom_point(
    alpha = 0.5
  ) +
  scale_shape_manual(
    values = c(point = "\u25CF", right = "\u25BA", left = "\u25C4", down = "\u25BC", up = "\u25B2"),
    guide = "none"
  ) +
  ggh4x::facet_wrap2(
    ~contrast_2 + batch,
    scales = "free", axes = "all"
  ) +
  labs(
    x = "Signed log10(padj) Paclitaxel",
    y = "Signed log10(padj) Rescue"
  )

dir.create("plots", showWarnings = FALSE)
ggsave(
  file.path("plots", "ptx_rescue_scatter.pdf"),
  p, width = 7, height = 7, device = cairo_pdf
)


p <- ggplot(
  ptx_rescue_plot_data,
  aes(
    x = logFC_1,
    y = logFC_2
  )
) +
  geom_abline() +
  geom_point(
    alpha = 0.5,
    shape = "circle small"
  ) +
  ggh4x::facet_wrap2(
    ~contrast_2 + batch,
    scales = "free", axes = "all"
  ) +
  labs(
    x = "log2(FC) Paclitaxel",
    y = "log2(FC) Rescue"
  )

dir.create("plots", showWarnings = FALSE)
ggsave(
  file.path("plots", "ptx_rescue_logFC_scatter.pdf"),
  p, width = 7, height = 4, device = cairo_pdf
)
```

## Define rescue types

Make alluvial diagram of PTX rescue

First approach:

- Find genes up or down in PTX treatment, identify those that are rescued by at
lest 1 logs in GNE-495 + PTX or JNK-inh + PTX

Here `x + PTX` is `x + PTX vs DMSO`, bare values are `PTX vs DMSO`

```{r}
ptx_rescue_by_compound <- de_long_selected_both %>%
  filter(contrast == "PTX") %>%
  power_inner_join(
    de_long_selected_both %>%
      filter(
        str_detect(contrast, fixed("+ PTX"))
      ) %>%
      separate_wider_regex(
        contrast,
        patterns = c(compound = "(?:GNE-495|JNK-inh)", " ", contrast = ".*")
      ) %>%
      select(
        batch, gene_id, compound, contrast,
        logFC, PValue, FDR, perturbed, signed_padj
      ) %>%
      pivot_wider(
        names_from = contrast,
        values_from = c(logFC, PValue, FDR, perturbed, signed_padj),
        names_sep = " "
      ),
    by = c("batch", "gene_id")
  ) |>
  crossing(
    perturbed_threshold = c(.001, .01, .05)
  ) %>%
  mutate(
    perturbed = case_when(
      FDR < perturbed_threshold & logFC > 0 ~ "up",
      FDR < perturbed_threshold & logFC < 0 ~ "down",
      TRUE ~ "not perturbed"
    )
  )

ptx_rescue_list <- ptx_rescue_by_compound %>%
  mutate(
    rescue_significant_opposite = case_when(
      is.na(perturbed) | perturbed == "not perturbed" ~ "not perturbed",
      perturbed == "up" & `FDR + PTX vs PTX` < .05 & `logFC + PTX vs PTX` < 0 ~ "rescued",
      perturbed == "down" & `FDR + PTX vs PTX` < .05 & `logFC + PTX vs PTX` > 0 ~ "rescued",
      TRUE ~ "unrescued"
    ),
    rescue_return_baseline = case_when(
      is.na(perturbed) | perturbed == "not perturbed" ~ "not perturbed",
      `FDR + PTX` > .05 ~ "rescued",
      TRUE ~ "unrescued"
    ),
    rescue_return_baseline_plus = case_when(
      is.na(perturbed) | perturbed == "not perturbed" ~ "not perturbed",
      perturbed == "up" & (`FDR + PTX` > .05 | `logFC + PTX` < 0) ~ "rescued",
      perturbed == "down" & (`FDR + PTX` > .05 | `logFC + PTX` > 0) ~ "rescued",
      TRUE ~ "unrescued"
    )
  )

ptx_rescue_list_neat <- ptx_rescue_list %>%
  select(
    perturbed_threshold, batch, gene_id, compound, perturbed,
    starts_with("rescue")
  )

ptx_rescue_list_long <- ptx_rescue_list_neat %>%
  pivot_longer(
    starts_with("rescue"),
    names_pattern = "rescue_(.*)",
    names_to = "rescue_definition",
    values_to= "rescue"
  )

write_csv(
  ptx_rescue_list_long,
  "ptx_rescue_list_long.csv.gz"
)

ptx_rescue_type_list <- ptx_rescue_list_long %>%
  pivot_wider(
    names_from = compound,
    values_from = rescue
  ) %>%
  mutate(
    rescue_type = case_when(
      perturbed == "not perturbed" ~ "not perturbed",
      `GNE-495` == "rescued" & `JNK-inh` == "rescued" ~ "both",
      `GNE-495` == "rescued" ~ "GNE-495 only",
      `JNK-inh` == "rescued" ~ "JNK-inh only",
      TRUE ~ "no rescue"
    )
  )

write_csv(
  ptx_rescue_type_list,
  "ptx_rescue_type_list.csv.gz"
)
```

```{r}
rd_bu_colors <- RColorBrewer::brewer.pal(5, "RdBu")
rescue_type_colors <- c(
  # both = "#8856a7",
  both = "gray10",
  `JNK-inh only` = rd_bu_colors[1],
  `GNE-495 only` = tail(rd_bu_colors, 1),
  `no rescue` = "gray70"
)
```

## Create Excel sheets

```{r}
gtf_path <- file.path("rnaseq", "Homo_sapiens.GRCh38.114.gtf.gz")
if (!file.exists(gtf_path))
  download.file("ftp://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/Homo_sapiens.GRCh38.114.gtf.gz", gtf_path)

symbol_path <- file.path("rnaseq", "non_alt_loci_set.txt")
if (!file.exists(symbol_path))
  download.file("https://storage.googleapis.com/public-download-files/hgnc/tsv/tsv/non_alt_loci_set.txt", symbol_path)
symbol_info <- read_tsv(symbol_path)

gtf_raw <- rtracklayer::import(gtf_path)
gene_info <- gtf_raw %>%
  as_tibble() %>%
  filter(type == "gene") %>%
  distinct(
    gene_id, gene_name, gene_biotype, chromosome = seqnames
  )

recode_batch <- \(x) recode(x, batch_1 = "(3uM)", batch_2 = "(1uM)", batch_1_riki = "(3uM, Riki)")

de_sheet <- de_long_selected_both %>%
  mutate(
    contrast = factor(contrast, levels = intersect(contrast_map, contrast))
  ) %>%
  arrange(contrast) %>%
  # filter(!(contrast == "PTX" & batch == "batch_2")) %>%
  transmute(
    gene_id,
    contrast = paste(contrast, recode_batch(batch)),
    logFC, FDR
  ) %>%
  filter(
    contrast != "PTX (1uM)"
  ) %>%
  mutate(
    contrast = recode(contrast, "PTX (3uM)" = "PTX", "PTX (3uM, Riki)" = "PTX (Riki)")
  ) %>%
  pivot_wider(
    names_from = contrast,
    values_from = c(logFC, FDR),
    names_glue = "{contrast}_{.value}",
    names_vary = "slowest"
  ) %>%
  power_inner_join(
    gene_info,
    by = "gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  power_left_join(
    distinct(symbol_info, gene_name = symbol, gene_description = name),
    by = "gene_name",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  relocate(
    gene_id, gene_name, gene_description, gene_biotype, chromosome,
    .before = 1
  ) %>%
  power_left_join(
    ptx_rescue_type_list %>%
      transmute(
        batch = recode_batch(batch),
        rescue_definition = fct_relevel(rescue_definition, "significant_opposite"),
        gene_id, rescue_definition, rescue_type
      ) %>%
      arrange(rescue_definition) %>%
      pivot_wider(
        names_from = c(rescue_definition, batch),
        values_from = rescue_type,
        names_glue = "rescue_{rescue_definition}_{batch}",
        names_vary = "fastest"
      ),
    by = "gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

dir.create(file.path("tables"), showWarnings = FALSE)

rescue_type_styles <- rescue_type_colors %>%
  map(
    \(x) createStyle(fgFill = colorspace::lighten(x, amount = .5))
  )

significant_style <- createStyle(fontColour = "red")

de_table_wb <- createWorkbook()
addWorksheet(de_table_wb, "DGE results")
writeDataTable(de_table_wb, "DGE results", de_sheet, tableStyle = "TableStyleLight8")
walk(
  which(str_detect(names(de_sheet), coll("FDR"))),
  \(i) {
    data_rows <- nrow(de_sheet) + 1L  # +1 for header
    de_col_data <- de_sheet[[i]]
    significant_rows <- which(!is.na(de_col_data) & de_col_data < 0.05) + 1L  # +1 for header
    if (length(significant_rows) > 0) {
      addStyle(
        wb = de_table_wb,
        sheet = "DGE results",
        style = significant_style,
        rows = significant_rows,
        cols = i,
        gridExpand = FALSE
      )
    }
  }
)
walk(
  which(str_detect(names(de_sheet), coll("logFC"))),
  \(i) {
    data_rows <- nrow(de_sheet) + 1L  # +1 for header
    conditionalFormatting(
      wb = de_table_wb,
      sheet = "DGE results",
      cols = i,
      rows = 2:data_rows,
      type = "colourScale",
      style = c("#4575b4", "white", "#d73027"),
      rule = c(-3, 0, 3)
    )
  }
)
walk(
  which(str_detect(names(de_sheet), coll("rescue_"))),
  \(i) {
    iwalk(
      rescue_type_styles,
      \(style, rescue_type) {
        rows <- which(
          de_sheet[[i]] == rescue_type
        ) + 1L
        # browser()
        addStyle(
          wb = de_table_wb,
          sheet = "DGE results",
          style = style,
          rows = rows,
          cols = i,
          gridExpand = TRUE
        )
      }
    )
  }
)
setColWidths(de_table_wb, "DGE results", cols = seq_len(ncol(de_sheet)), widths = "auto")

saveWorkbook(de_table_wb, file.path("tables", "de_results.xlsx"), overwrite = TRUE)
```


## Plot rescue plots with rescue types


```{r}
ptx_rescue_plot_data_w_type <- ptx_rescue_plot_data %>%
  left_join(
    ptx_rescue_type_list,
    by = c("batch", "gene_id"),
    relationship = "many-to-many"
  ) %>%
  mutate(
    rescue_type = replace_na(rescue_type, "not perturbed") %>%
      factor(
        levels = c(
          "both",
          "JNK-inh only",
          "GNE-495 only",
          "no rescue",
          "not perturbed"
        )
      )
  )  %>%
  filter(rescue_type != "not perturbed") %>%
  arrange(
    if_else(
      rescue_type == "no rescue",
      -1,
      withr::with_seed(42, runif(n()))
    )
  )

p <- ggplot(
  ptx_rescue_plot_data_w_type,
  aes(
    x = signed_padj_1_clamped,
    y = signed_padj_2_clamped,
    shape = signed_padj_symbol,
    color = rescue_type
  )
) +
  geom_abline() +
  geom_point(
    alpha = 0.5
  ) +
  scale_shape_manual(
    values = c(point = "\u25CF", right = "\u25BA", left = "\u25C4", down = "\u25BC", up = "\u25B2"),
    guide = "none"
  ) +
  scale_color_manual(
    values = rescue_type_colors
  ) +
  ggh4x::facet_grid2(
    rescue_definition + batch ~ contrast_2,
    scales = "free", axes = "all"
  ) +
  labs(
    x = "Signed log10(padj) Paclitaxel",
    y = "Signed log10(padj) Rescue",
    color = "Rescue"
  )

dir.create("plots", showWarnings = FALSE)
ggsave(
  file.path("plots", "ptx_rescue_scatter_w_rescue_type_all.pdf"),
  p, width = 8, height = 15, device = cairo_pdf
)


p <- ggplot(
  ptx_rescue_plot_data_w_type,
  aes(
    x = logFC_1,
    y = logFC_2,
    # shape = signed_padj_symbol,
    color = rescue_type
  )
) +
  geom_abline() +
  geom_point(
    alpha = 0.5,
    shape = "circle small",
    size = 1
  ) +
  scale_color_manual(
    values = rescue_type_colors
  ) +
  facet_grid(
    rescue_definition + batch ~ contrast_2
  ) +
  coord_equal() +
  labs(
    x = "log2(fold-change) Paclitaxel",
    y = "log2(fold-change) Rescue",
    color = "Rescue"
  )

dir.create("plots", showWarnings = FALSE)
ggsave(
  file.path("plots", "ptx_rescue_scatter_logFC_w_rescue_type_all.pdf"),
  p, width = 5, height = 15, device = cairo_pdf
)
```

## Gene sets

```{r}
library(msigdbr)
all_msigdbr <- msigdbr()

msigdbr_of_interest <- all_msigdbr %>%
  filter(
    gs_collection %in% c("H") |
      gs_subcollection %in% c(
        "CP:PID",
        "CP:REACTOME",
        "CP:KEGG_MEDICUS",
        "GO:BP",
        "GO:MF",
        "GO:CC"
      ),
    !str_detect(gs_name, coll("MEDICUS_PATHOGEN")),
    !str_detect(gs_name, coll("MEDICUS_VARIANT"))
  )

msigdbr_of_interest_gs <- msigdbr_of_interest %>%
  group_by(gs_name) %>%
  summarize(
    gs = list(ensembl_gene),
    .groups = "drop"
  ) %>%
  with(
    set_names(
      gs,
      gs_name
    )
  )

```

## FGSEA

### Rescue vectors

Linear scale most rescued by agent to most not rescued. Only including
genes perturbed by PTX

Options

1. Order genes by magnitude of rescue in Agent + PTX vs PTX but clamp log2FC to
  magnitude of change in PTX vs DMSO
  i. By signed padj
  ii. By log2FC
2.


```{r}
library(fgsea)

feasible_gene_vec <- with(
  feasible_genes,
  split(gene_id, batch)
)

fgsea_rescue_vectors <- ptx_rescue_by_compound %>%
  semi_join(
    feasible_genes,
    by = c("batch", "gene_id")
  ) %>%
  mutate(
    signed_padj_rescue_raw = if_else(
      sign(signed_padj) == sign(`signed_padj + PTX vs PTX`),
      0,
      `signed_padj + PTX vs PTX`
    ),
    logFC_rescue_raw = if_else(
      sign(logFC) == sign(`logFC + PTX vs PTX`),
      0,
      `logFC + PTX vs PTX`
    ),
    signed_padj_rescue_clamped = if_else(
      abs(signed_padj_rescue_raw) > abs(signed_padj),
      sign(signed_padj_rescue_raw) * abs(signed_padj),
      signed_padj_rescue_raw
    ),
    logFC_rescue_clamped = if_else(
      abs(logFC_rescue_raw) > abs(logFC),
      sign(logFC_rescue_raw) * abs(logFC),
      logFC_rescue_raw
    ),
    signed_padj_raw = `signed_padj + PTX vs PTX`,
    logFC_raw = `logFC + PTX vs PTX`
  )

fgsea_rescue_vectors_long <- fgsea_rescue_vectors %>%
  dplyr::select(
    batch,
    gene_id,
    compound,
    contains("rescue"),
    ends_with("raw")
  ) %>%
  pivot_longer(
    -c(batch, gene_id, compound),
    names_to = "metric",
    values_to = "value"
  )
```

### Check rescue vectors

```{r}
p <- fgsea_rescue_vectors %>%
  filter(
    compound == "GNE-495"
  ) %>%
  ggplot(
    aes(
      x = signed_padj,
      y = signed_padj_rescue_clamped
    )
  ) +
  geom_point() +
  facet_wrap(~batch)
p
```

```{r}
fgsea_rescue_vectors_grouped <- fgsea_rescue_vectors_long %>%
  group_by(batch, metric, compound) %>%
  summarize(
    gene_stats = set_names(
      value, gene_id
    ) %>%
      list(),
    .groups = "drop"
  )

fgsea_rescue_vectors_res_raw <- fgsea_rescue_vectors_grouped %>%
  rowwise() %>%
  mutate(
    res = fgseaMultilevel(
      msigdbr_of_interest_gs,
      gene_stats,
      BPPARAM = BiocParallel::MulticoreParam(workers = 6)
    ) %>%
      list()
  ) %>%
  ungroup()

fgsea_rescue_vectors_res <- fgsea_rescue_vectors_res_raw %>%
  dplyr::select(-gene_stats) %>%
  unnest(res)

qsave(
  fgsea_rescue_vectors_res,
  "fgsea_res.qs"
)
fgsea_rescue_vectors_res <- qread("fgsea_res.qs")
```


```{r}
plotEnrichment(
  msigdbr_of_interest_gs[["REACTOME_RECYCLING_PATHWAY_OF_L1"]],
  fgsea_rescue_vectors_grouped %>%
    filter(compound == "GNE-495", metric == "signed_padj_rescue_clamped") %>%
    chuck("gene_stats", 1)
)

plotEnrichment(
  msigdbr_of_interest_gs[["REACTOME_RECYCLING_PATHWAY_OF_L1"]],
  fgsea_rescue_vectors_grouped %>%
    filter(compound == "JNK-inh", metric == "signed_padj_rescue_clamped") %>%
    chuck("gene_stats", 1)
)
```

### FGSEA on raw perturbations

```{r}
fgsea_pert_input <- de_long_selected_both %>%
  semi_join(
    feasible_genes,
    by = c("batch", "gene_id")
  ) %>%
  select(
    batch, gene_id, contrast, signed_padj, logFC
  ) %>%
  pivot_longer(
    -c(batch, gene_id, contrast),
    names_to = "metric",
    values_to = "value"
  ) %>%
  group_by(
    batch, contrast, metric
  ) %>%
  summarize(
    gene_stats = set_names(
      value, gene_id
    ) %>%
      list(),
    .groups = "drop"
  )

fgsea_pert_res_raw <- fgsea_pert_input %>%
  rowwise() %>%
  mutate(
    res = fgseaMultilevel(
      msigdbr_of_interest_gs,
      gene_stats,
      BPPARAM = BiocParallel::MulticoreParam(workers = 6)
    ) %>%
      list()
  ) %>%
  ungroup()

fgsea_pert_res <- fgsea_pert_res_raw %>%
  dplyr::select(-gene_stats) %>%
  unnest(res)

qsave(
  fgsea_pert_res,
  "fgsea_pert_res.qs"
)
# fgsea_pert_res <- qread("fgsea_pert_res.qs")
```

Briefly investigate P38MAPK_EVENTS specifically

```{r}
fgsea_pert_input %>%
  filter(contrast == "GNE-495 + PTX vs PTX", metric == "signed_padj") %>%
  rowwise() %>%
  mutate(
    p = list(
      plotEnrichment(
        msigdbr_of_interest_gs[["REACTOME_P38MAPK_EVENTS"]],
        gene_stats
      ) +
        labs(title = batch)
    )
  ) %>%
  with(set_names(p, batch)) %>%
  patchwork::wrap_plots()
```



```{r}
fgsea_pert_res %>%
  count(batch, contrast, metric, padj < .05) %>%
  View()

fgsea_pert_res %>%
  arrange(padj) %>%
  print(n = 100)
```

```{r}
library(seriation)
cluster_mat <- function(mat) {
  d <- dist(mat)
  set.seed(42)
  hclust(d, method = "average") %>%
    seriation:::reorder.hclust(dist = d, method = "olo")
}

cluster_df <- function(df, row_var, col_var, value_var, values_fill = 0) {
  # browser()
  mat <- df %>%
    dplyr::select({{row_var}}, {{col_var}}, {{value_var}}) %>%
    pivot_wider(names_from = {{col_var}}, values_from = {{value_var}}, values_fill = values_fill) %>%
    column_to_rownames(rlang::as_name(rlang::enquo(row_var)))
  # browser()
  if (rlang::is_bare_numeric(pull(df, {{value_var}}))) {
    dist_rows <- dist(mat, method = "euclidian")
    dist_cols <- dist(t(mat), method = "euclidian")
  } else {
    # browser()
    dist_rows <- cluster::daisy(mat, metric = "gower")
    dist_cols <- t(mat) %>%
      as.data.frame() %>%
      mutate(across(everything(), \(x) factor(x, levels = levels(pull(df, {{value_var}}))))) %>%
      cluster::daisy(metric = "gower")
  }
  clust_rows <- hclust(dist_rows, method = "average") %>%
      reorder(dist_rows, method = "olo")
  clust_cols <- hclust(dist_cols, method = "average") %>%
      reorder(dist_cols, method = "olo")
  df %>%
    mutate(
      "{{row_var}}" := factor({{row_var}}, levels = clust_rows$labels[clust_rows$order]),
      "{{col_var}}" := factor({{col_var}}, levels = clust_cols$labels[clust_cols$order])
    )
}
```


```{r}
fgsea_res_mat <- fgsea_pert_res %>%
  transmute(comparison_unique, pathway, signed_padj = -sign(NES)*log10(padj)) %>%
  pivot_wider(
    names_from = comparison_unique,
    values_from = signed_padj
  ) %>%
  column_to_rownames("pathway") %>%
  as.matrix() %>% {
    clust_rows <- cluster_mat(.)
    clust_cols <- cluster_mat(t(.))
    .[clust_rows$order, clust_cols$order]
  }

library(cmapR)
fgsea_res_gct <- new(
  "GCT",
  mat = fgsea_res_mat,
  rdesc = msigdbr_of_interest %>%
    distinct(
      id = gs_name,
      gs_cat, gs_subcat
    ) %>%
    dplyr::slice(match(rownames(fgsea_res_mat), id)) %>%
    as.data.frame(),
  cdesc = fgsea_gene_stats %>%
    distinct(
      id = comparison_unique,
      comparison_type
    ) %>%
    dplyr::slice(match(colnames(fgsea_res_mat), id)) %>%
    as.data.frame()
)

write_gct(
  fgsea_res_gct,
  "vemu/fgsea_res_h_pid_kegg.gct",
  appenddim = FALSE
)
```

## topGO


1. Use significant opposite rscue edfinition
2. Define PTX, GNE, JNK effects alone using GO enrichment
3. Make tables of GO enrichment of the GNE/JNK rescue gene sets,
4. Include GNE-only and JNK-only rescue gene sets
5. Make two big heatmaps of gene expression, group by PTX up/down, JNK/GNE rescue
6. Make TF enrichment for just raw JNK/GNE effects

### On PTX rescue gene sets

```{r}
ptx_rescue_gene_sets_list <- ptx_rescue_list_long %>%
  filter(rescue == "rescued") %>%
  group_by(
    perturbed_threshold, batch, compound, perturbed, rescue_definition
  ) %>%
  summarize(
    gene_vec = list(
      {
        feasible_gene_vec_cur <- feasible_gene_vec[[cur_group()$batch]]
        set_names(
          # factor(as.integer(feasible_gene_vec %in% gene_id)),
          if_else(feasible_gene_vec_cur %in% gene_id, .01, 1),
          feasible_gene_vec_cur
        )
      }
    ),
    .groups = "drop"
  )

library(topGO)

gene_sel_fun <- \(x) x < .05

topgo_objects <- ptx_rescue_gene_sets_list %>%
  group_by(batch) %>%
  slice_head(n = 1L) %>%
  ungroup() %>%
  crossing(
    ontology = c("BP", "MF", "CC")
  ) %>%
  transmute(
    batch, ontology,
    object = map2(
      ontology, gene_vec,
      \(o, g) new(
        "topGOdata",
        ontology = o,
        allGenes = g,
        geneSel = gene_sel_fun,
        nodeSize = 10,
        annot = annFUN.org,
        ID = "ensembl",
        mapping = "org.Hs.eg.db"
      )
    )
  )

qsave(
  topgo_objects,
  "topgo_objects.qs"
)
# topgo_objects <- qread("topgo_objects.qs")

# go_gene_map <- map(
#   topgo_objects,
#   \(x) x@graph@nodes %>%
#     set_names() %>%
#     map(\(y) genesInTerm(x, y)[[1]])
# ) %>%
#   # unlist(recursive = FALSE)
#   {do.call(c, .)}

# qsave(
#   go_gene_map,
#   "go_gene_map.qs"
# )
# go_gene_map <- qread("go_gene_map.qs")

topgo_res_raw <- ptx_rescue_gene_sets_list %>%
  power_inner_join(
    topgo_objects,
    by = "batch",
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn"
    )
  ) %>%
  rowwise() %>%
  mutate(
    res = {
      o <- object
      go <- updateGenes(
        o,
        geneList = gene_vec,
        geneSelFun = gene_sel_fun
      )
      # browser()
      raw <- runTest(
        go,
        algorithm = "weight01",
        statistic = "fisher"
      )
      res_table <- GenTable(
        go,
        fisher = raw,
        topNodes = length(usedGO(go)),
        numChar = 300
      )
      list(
        res_raw = raw,
        res = res_table
      ) %>%
        list()
    }
  ) %>%
  ungroup()

qsave(
  topgo_res_raw %>%
   select(-object),
  "topgo_res_raw.qs"
)
# topgo_res_raw <- qread("topgo_res_raw.qs")

topgo_res <- topgo_res_raw %>%
  rowwise() %>%
  mutate(
    res_table = list(res[["res"]])
  ) %>%
  ungroup() %>%
  dplyr::select(-c(res, gene_vec, object)) %>%
  unnest(res_table) %>%
  mutate(
    term_unique = paste(
      Term, GO.ID, sep = "_"
    ),
    fisher_double = as.double(str_remove(fisher, fixed("< ")))
  )

qsave(
  topgo_res,
  "topgo_res.qs"
)
# topgo_res <- qread("topgo_res.qs")

go_meta <- topgo_res %>%
  distinct(
    ontology, GO.ID, Term, term_unique
  )
```


### TopGO on raw perturbations

```{r}
topgo_up_down_funs <- list(
  up = \(x) x > -log10(.05),
  down = \(x) x < log10(.05)
)

topgo_pert_input <- fgsea_pert_input %>%
  filter(metric == "signed_padj") %>%
  dplyr::select(-metric) %>%
  mutate(
    gene_stats = map(
      gene_stats,
      \(x) {
        set_names(
          ifelse(feasible_gene_vec %in% names(x), x[feasible_gene_vec], 0),
          feasible_gene_vec
        )
      }
    )
  ) %>%
  crossing(
    ontology = names(topgo_objects),
    direction = c("up", "down")
  )

topgo_res_pert_raw <- topgo_pert_input %>%
  rowwise() %>%
  mutate(
    res = {
      # browser()
      go <- updateGenes(
        topgo_objects[[ontology]],
        geneList = gene_stats,
        geneSelFun = topgo_up_down_funs[[direction]]
      )
      # browser()
      raw <- runTest(
        go,
        algorithm = "weight01",
        statistic = "fisher"
      )
      res_table <- GenTable(
        go,
        fisher = raw,
        topNodes = length(usedGO(go)),
        numChar = 300
      )
      list(
        res_raw = raw,
        res = res_table
      ) %>%
        list()
    }
  ) %>%
  ungroup()

qsave(
  topgo_res_pert_raw,
  "topgo_res_pert_raw.qs"
)
topgo_res_pert_raw <- qread("topgo_res_pert_raw.qs")

topgo_res_pert <- topgo_res_pert_raw %>%
  rowwise() %>%
  mutate(
    res_table = list(res[["res"]])
  ) %>%
  ungroup() %>%
  dplyr::select(-c(res, gene_stats)) %>%
  unnest(res_table) %>%
  mutate(
    term_unique = paste(
      Term, GO.ID, sep = "_"
    ),
    fisher_double = as.double(str_remove(fisher, fixed("< ")))
  )

qsave(
  topgo_res_pert,
  "topgo_res_pert.qs"
)
# topgo_res_pert <- qread("topgo_res_pert.qs")

```

### TopGO on rescue intersection


```{r}
topgo_ptx_rescue_type_list <- ptx_rescue_type_list %>%
  group_by(
    perturbed_threshold, batch, perturbed, rescue_definition, rescue_type
  ) %>%
  summarize(
    gene_vec = list(
      set_names(
        # factor(as.integer(feasible_gene_vec %in% gene_id)),
        if_else(feasible_gene_vec[[cur_group()$batch]] %in% gene_id, .01, 1),
        feasible_gene_vec[[cur_group()$batch]]
      )
    ),
    .groups = "drop"
  )


topgo_ptx_rescue_type_list_raw <- topgo_ptx_rescue_type_list %>%
  power_inner_join(
    topgo_objects,
    by = "batch",
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn"
    )
  ) %>%
  rowwise() %>%
  mutate(
    res = {
      o <- object
      go <- updateGenes(
        o,
        geneList = gene_vec,
        geneSelFun = gene_sel_fun
      )
      # browser()
      raw <- runTest(
        go,
        algorithm = "weight01",
        statistic = "fisher"
      )
      res_table <- GenTable(
        go,
        fisher = raw,
        topNodes = length(usedGO(go)),
        numChar = 300
      )
      list(
        res_raw = raw,
        res = res_table
      ) %>%
        list()
    }
  ) %>%
  ungroup()

qsave(
  topgo_ptx_rescue_type_list_raw %>%
   select(-object),
  "topgo_ptx_rescue_type_list_raw.qs"
)
# topgo_ptx_rescue_type_list_raw <- qread("topgo_ptx_rescue_type_list_raw.qs")

topgo_ptx_rescue_type_list_res <- topgo_ptx_rescue_type_list_raw %>%
  rowwise() %>%
  mutate(
    res_table = list(res[["res"]])
  ) %>%
  ungroup() %>%
  dplyr::select(-c(res, gene_vec, object)) %>%
  unnest(res_table) %>%
  mutate(
    term_unique = paste(
      Term, GO.ID, sep = "_"
    ),
    fisher_double = as.double(str_remove(fisher, fixed("< ")))
  )

write_csv(
  topgo_ptx_rescue_type_list_res,
  "topgo_ptx_rescue_type_list_res.csv.gz"
)
# topgo_ptx_rescue_type_list_res <- read_csv("topgo_ptx_rescue_type_list_res.csv.gz")

topgo_ptx_rescue_type_list_res %>%
  filter(
    fisher_double < .05,
    rescue_definition == "significant_opposite",
    rescue_type == "GNE-495 only"
  ) %>%
  arrange(fisher_double) %>%
  View()
```


## FGSEA overrepresentation

### FGSEA overrepresentation on rescue gene sets

```{r}
ptx_rescue_overrepresentation_raw <- ptx_rescue_gene_sets_list %>%
  mutate(
    res = map2(
      gene_vec, batch,
      \(x, batch) {
        gs <- names(x)[x < .05]
        fora(
          msigdbr_of_interest_gs,
          gs,
          universe = feasible_gene_vec[[batch]]
        )
      }
    )
  )

ptx_rescue_overrepresentation <- ptx_rescue_overrepresentation_raw %>%
  transmute(
    perturbed_threshold, batch, rescue_definition, perturbed, compound,
    res
  ) %>%
  unnest(res)

qsave(
  ptx_rescue_overrepresentation,
  "ptx_rescue_overrepresentation.qs"
)
```

### FGSEA overrepresentation on rescue gene set intersections

Rescued by both, neither, or single drugs

```{r}
ptx_rescue_intersections <- ptx_rescue_type_list %>%
  filter(perturbed != "not perturbed") %>%
  group_nest(perturbed_threshold, batch, rescue_definition, perturbed, rescue_type) %>%
  bind_rows(
    ptx_rescue_type_list %>%
      filter(perturbed != "not perturbed") %>%
      group_nest(perturbed_threshold, batch, rescue_definition, rescue_type) %>%
      mutate(
        perturbed = "both_up_down"
      )
  )

ptx_rescue_intersections_overrepresentation_raw <- ptx_rescue_intersections %>%
  mutate(
    res = map2(
      data, batch,
      \(x, batch) {
        gs <- x$gene_id
        fora(
          msigdbr_of_interest_gs,
          gs,
          universe = feasible_gene_vec[[batch]]
        )
      }
    )
  )

ptx_rescue_intersections_overrepresentation <- ptx_rescue_intersections_overrepresentation_raw %>%
  transmute(
    perturbed_threshold, batch, rescue_definition, perturbed, rescue_type,
    res
  ) %>%
  unnest(res)

qsave(
  ptx_rescue_intersections_overrepresentation,
  "ptx_rescue_intersections_overrepresentation.qs"
)
```

```{r}
ptx_rescue_intersections_overrepresentation %>%
  filter(rescue_definition == "significant_opposite") %>%
  filter(rescue_type == "GNE-495 only", batch != "batch_1_riki") %>%
  arrange(pval) %>%
  View()

```

### FGSEA overrepresentation on raw perturbations


```{r}
library(furrr)
plan(multicore, workers = 6)
ptx_single_perturbation_overrep_raw <- de_long_selected_both %>%
  crossing(
    significant_threshold = c(.05)
  ) %>%
  mutate(
    direction = if_else(logFC > 0, "up", "down")
  ) %>%
  filter(
    FDR < significant_threshold
  ) %>% {
    bind_rows(
      mutate(
        ., direction = "both"
      ),
      .
    )
  } %>%
  group_nest(contrast, direction, significant_threshold, batch) %>%
  mutate(
    res = future_map2(
      data, batch,
      \(df, batch) {
        fora(
          msigdbr_of_interest_gs,
          df$gene_id,
          universe = feasible_gene_vec[[batch]]
        )
      },
      .progress = TRUE
    ),
    res_reduced = future_pmap(
      list(data, res, batch),
      \(df, res, batch) {
        collapsePathwaysORA(
          filter(res, padj < .01),
          msigdbr_of_interest_gs,
          df$gene_id,
          universe = feasible_gene_vec[[batch]]
        )
      },
      .progress = TRUE
    )
  )
plan(sequential)

qsave(
  ptx_single_perturbation_overrep_raw,
  file.path("results", "ptx_single_perturbation_overrepresentation_raw.qs")
)

ptx_single_perturbation_overrep <-ptx_single_perturbation_overrep_raw %>%
  transmute(
    contrast, direction, significant_threshold, batch,
    res = map2(
      res, res_reduced,
      \(full, reduced) {
        mutate(
          full,
          is_main = pathway %in% reduced$mainPathways,
          parent_pathway = reduced$parentPathways[pathway]
        )
      }
    )
  ) %>%
  unnest(res)

qsave(
  ptx_single_perturbation_overrep,
  file.path("results", "ptx_single_perturbation_overrepresentation.qs")
)

```

### FGSEA overrepresentation on LFC clusters

```{r}
lfc_clusters <- read_csv(
  file.path("results", "de_heatmap_all_combined_signed_p_scaled_smaller_hc_clusters.csv")
)

library(furrr)
plan(multicore, workers = 6)
lfc_clusters_overrep_raw <- lfc_clusters %>%
  group_nest(.cluster) %>%
  mutate(
    res = future_map(
      data,
      \(df) {
        fora(
          msigdbr_of_interest_gs,
          df$gene_id,
          universe = feasible_gene_vec[["batch_1"]]
        )
      },
      .progress = TRUE
    ),
    res_reduced = future_pmap(
      list(data, res),
      \(df, res) {
        collapsePathwaysORA(
          filter(res, padj < .01),
          msigdbr_of_interest_gs,
          df$gene_id,
          universe = feasible_gene_vec[["batch_1"]]
        )
      },
      .progress = TRUE
    )
  )
plan(sequential)

qsave(
  lfc_clusters_overrep_raw,
  file.path("results", "lfc_clusters_overrep_raw.qs")
)

lfc_clusters_overrep <- lfc_clusters_overrep_raw %>%
  transmute(
    .cluster,
    res = map2(
      res, res_reduced,
      \(full, reduced) {
        mutate(
          full,
          is_main = pathway %in% reduced$mainPathways,
          parent_pathway = reduced$parentPathways[pathway]
        )
      }
    )
  ) %>%
  unnest(res)

qsave(
  lfc_clusters_overrep,
  file.path("results", "lfc_clusters_overrepresentation.csv.gz")
)
```


#### Excel


```{r}
library(openxlsx)

significant_style <- createStyle(fontColour = "red")

lfc_clusters_overrep_excel <- createWorkbook()

lfc_clusters_overrep %>%
  select(-c(overlapGenes, is_main, parent_pathway)) %>%
  group_by(.cluster) %>%
  group_walk(
    \(df, g) {
      addWorksheet(
        lfc_clusters_overrep_excel,
        g$.cluster
      )
      writeDataTable(
        lfc_clusters_overrep_excel,
        sheet = g$.cluster,
        x = df,
        tableStyle = "TableStyleLight8"
      )
      setColWidths(
        lfc_clusters_overrep_excel, g$.cluster,
        cols = seq_len(ncol(df)),
        widths = "auto"
      )
      walk(
        which(names(df) == "padj"),
        \(i) {
          data_rows <- nrow(df) + 1L  # +1 for header
          de_col_data <- df[[i]]
          significant_rows <- which(!is.na(de_col_data) & de_col_data < 0.05) + 1L  # +1 for header
          if (length(significant_rows) > 0) {
            addStyle(
              wb = lfc_clusters_overrep_excel,
              sheet = g$.cluster,
              style = significant_style,
              rows = significant_rows,
              cols = i,
              gridExpand = FALSE
            )
          }
        }
      )
    }
  )

saveWorkbook(
  lfc_clusters_overrep_excel,
  file.path("tables", "lfc_clusters_overrepresentation.xlsx"),
  overwrite = TRUE
)

```



## Intersection of 3uM and 1uM PTX rescue gene sets

Checking PTX + rescue vs PTX at 1uM and 3uM, taking overlap

```{r}
ptx_rescue_both_doses <- de_long_combined %>%
  filter(
    name %in% c(
      "GNE-495 + PTX vs PTX",
      "JNK-inh + PTX vs PTX"
    ),
    k == 0,
    batch != "batch_1_riki"
  ) %>%
  pivot_wider(
    id_cols = c(name, gene_id, gene_name),
    names_from = batch,
    values_from = c(log2FoldChange, padj)
  ) %>%
  filter(
    padj_batch_1 < .05, padj_batch_2 < .05,
    sign(log2FoldChange_batch_1) == sign(log2FoldChange_batch_2)
  )

ptx_rescue_both_doses_raw <- ptx_rescue_both_doses %>%
  mutate(
    direction = if_else(
      log2FoldChange_batch_1 > 0,
      "up",
      "down"
    )
  ) %>% {
    bind_rows(
      mutate(
        ., direction = "both"
      ),
      .
    )
  } %>%
  group_by(name, direction) %>%
  summarize(
    res = fora(
      msigdbr_of_interest_gs,
      gene_id,
      universe = feasible_gene_vec[["batch_1"]]
    ) %>%
      list(),
    res_reduced = collapsePathwaysORA(
      filter(res[[1]], padj < .01),
      msigdbr_of_interest_gs,
      gene_id,
      universe = feasible_gene_vec[["batch_1"]]
    ) %>%
      list(),
    .groups = "drop"
  )

qsave(
  ptx_rescue_both_doses_raw,
  file.path("results", "ptx_rescue_both_doses_fgsea_overrepresentation_raw.qs")
)

ptx_rescue_both_doses_long <- ptx_rescue_both_doses_raw %>%
  transmute(
    name, direction,
    res = map2(
      res, res_reduced,
      \(full, reduced) {
        mutate(
          full,
          is_main = pathway %in% reduced$mainPathways,
          parent_pathway = reduced$parentPathways[pathway]
        )
      }
    )
  ) %>%
  unnest(res)

write_csv(
  ptx_rescue_both_doses_long,
  file.path("results", "ptx_rescue_both_doses_fgsea_overrepresentation.csv.gz")
)
```


### Alluvial diagram of doses combined rescue data

```{r}
library(ggalluvial)

rescue_both_doses_alluvial_data <-  de_long_combined %>%
  filter(
    name %in% c(
      "GNE-495 + PTX vs PTX",
      "JNK-inh + PTX vs PTX"
    ),
    k == 0,
    batch != "batch_1_riki"
  ) %>%
  pivot_wider(
    id_cols = c(name, gene_id, gene_name),
    names_from = batch,
    values_from = c(log2FoldChange, padj)
  )

ptx_rescue_type_list %>%
  count(perturbed)

ptx_rescue_alluvial_plots <- ptx_rescue_type_list %>%
  group_nest(perturbed_threshold, batch, rescue_definition) %>%
  mutate(
    alluvial_data = map(
      data,
      \(x) count(
        x,
        `PTX perturbed` = perturbed,
        `GNE-495 perturbed` = `GNE-495`,
        `JNK-inh perturbed` = `JNK-inh`,
        rescue_type
      ) %>%
        filter(`PTX perturbed` != "not perturbed")
    ),
    lode_data = map(
      alluvial_data,
      \(x) to_lodes_form(
        x,
        ends_with("perturbed"),
        key = "Contrast", value = "Gene set"
      )
    ),
    p = map(
      lode_data,
      \(x)
      ggplot(
        x,
        aes(
          x = Contrast, y = n, stratum = `Gene set`, alluvium = alluvium
        )
      ) +
        scale_x_discrete(expand = c(0.05, 0.05)) +
        geom_alluvium(
          aes(
            # color = `PTX perturbed 2`,
            fill = rescue_type
          ),
          width = 1/10
        ) +
        # ggokabeito::scale_color_okabe_ito(order = c(2, 1), guide = "none") +
        scale_fill_manual(
          values = rescue_type_colors
        ) +
        labs(x = NULL) +
        # scale_fill_manual(
        #     values = c(existing = "#a2a2a2", added = "#c81919"),
        #     guide = "none"
        # ) +
        geom_stratum(width = 1/6) +
        # geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
        geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
        geom_text(
          aes(
            x = stage(Contrast, after_scale = x + .12),
            label = n
          ),
          stat = StatAlluvium,
          hjust = 0,
          data = \(x) filter(x, as.integer(Contrast) == 1L)
        ) +
        theme_minimal() +
        # labs(y = "Data points", fill = NA) +
        theme(
            panel.grid.major.x = element_blank()
        ) +
        coord_cartesian(clip = "off")
    )
  )

pwalk(
  ptx_rescue_alluvial_plots,
  \(p, perturbed_threshold, batch, rescue_definition, ...) ggsave(
    file.path("plots", paste0("ptx_rescue_threshold_", perturbed_threshold, "_", rescue_definition, "_alluvium_", batch, ".pdf")),
    p, width = 6, height = 4
  )
)

ggsave(
  file.path("plots", "ptx_rescue_no_threshold_alluvium.pdf"),
  p, width = 6, height = 4
)
```


## Clue queries

### Clue queries on rescue gene sets

Requiring padj < .05, using 150 most significant genes for the query and only
query with BING genes

```{r}
library(cluequery)
library(org.Hs.eg.db)
library(AnnotationDbi)

entrez_gene_mapping <- select(
  org.Hs.eg.db,
  keys = unique(de_long_selected$gene_id),
  columns = c("ENSEMBL", "ENTREZID"),
  keytype = "ENSEMBL"
) %>%
  drop_na(ENTREZID) %>%
  distinct() %>%
  as_tibble()

clue_input_data <- de_long_selected %>%
  filter(
    FDR < .05
  ) %>%
  power_inner_join(
    entrez_gene_mapping,
    by = c("gene_id" = "ENSEMBL"),
    check = check_specs(
      duplicate_keys_right = "warn",
      unmatched_keys_left = "warn"
    )
  ) %>%
  filter(
    ENTREZID %in% cluequery::gene_space[["entrez_id"]][cluequery::gene_space[["bing"]]]
  ) %>%
  arrange(PValue) %>%
  group_by(contrast, perturbed) %>%
  slice_head(n = 150) %>%
  ungroup() %>%
  dplyr::select(
    gene_id = ENTREZID,
    gene_set = contrast,
    direction = perturbed
  )

clue_input_data %>%
  count(
    gene_set, direction
  )

clue_input_gmt <- clue_gmt_from_df(
  clue_input_data,
  drop_invalid = TRUE
)

clue_job <- clue_query_submit(
  clue_input_gmt[["up"]], clue_input_gmt[["down"]],
  "PTX rescue conditions BING"
)
clue_query_wait(clue_job)

clue_res_path <- clue_query_download(clue_job, file.path("results", "clue_res_rescue_gene_sets.tar.gz"))
clue_res_raw <- clue_parse_result(
  clue_res_path,
  score_level = "summary",
  score_type = "tau"
)

write_csv(
  clue_res_raw,
  file.path("results", "clue_res_rescue_gene_sets.csv.gz")
)
```

### Clue query heatmap

```{r}
hm_conditions <- c(
  "PTX", "GNE-495 + PTX", "JNK-inh + PTX", "GNE-495", "JNK-inh"
)

clue_res_kd_data <- clue_res_raw %>%
  filter(
    gene_set %in% hm_conditions
  ) %>%
  group_by(
    pert_id
  ) %>%
  filter(
    pert_type %in% c("trt_sh.cgs"),
    # any(abs(tau) > .9)
    abs(tau[gene_set == "PTX"]) > 90
  ) %>%
  ungroup()

clue_res_kd_mat <- clue_res_kd_data %>%
  dplyr::select(pert_iname, gene_set, tau) %>%
  pivot_wider(
    names_from = gene_set,
    values_from = tau
  ) %>%
  column_to_rownames("pert_iname") %>%
  as.matrix() %>% {
    .[, hm_conditions]
  }

hm <- Heatmap(
  clue_res_kd_mat,
  name = "Tau",
  col = circlize::colorRamp2(
    seq(from = -100, to = 100, length.out = 51),
    paletteer::paletteer_c("pals::ocean.balance", n = 51)
  ),
  cluster_rows = cluster_mat,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  use_raster = TRUE,
)

withr::with_pdf(
  file.path("plots", "clue_res_kd_heatmap.pdf"),
  draw(hm),
  width = 12, height = 12
)
```


## TopGO plots


```{r}
topgo_res_plotting <- topgo_res %>%
  group_by(
    ontology, term_unique
  ) %>%
  filter(any(fisher_double < 0.05)) %>%
  ungroup() %>%
  mutate(
    rescue_type = paste(perturbed, compound, rescue_definition, sep = "_"),
    odds_ratio = Significant / Expected
  )

topgo_res_plotting_clust <- topgo_res_plotting %>%
  dplyr::transmute(ontology, term_unique, rescue_type, odds_ratio, across(fisher_double, \(x) -log10(x))) %>%
  group_nest(ontology) %>%
  mutate(
    data = map(
      data,
      \(x) cluster_df(x, term_unique, rescue_type, odds_ratio)
    )
  ) %>%
  with(set_names(data, ontology))

plot_big_heatmap <- \(x, z, df) ggplot(
  df %>%
    mutate(
      odds_ratio = if_else(odds_ratio == 0, min(odds_ratio[odds_ratio > 0]) * .5, odds_ratio),
      perturbed = factor(perturbed, levels = c("up", "down")),
      compound = factor(compound, levels = c("PTX", "JNK-inh", "GNE-495")),
      rescue_definition = factor(rescue_definition, levels = c("return_baseline", "significant_opposite", "overshot")),
      across(term_unique, \(y) factor(y, levels = levels(topgo_res_plotting_clust[[x]]$term_unique)) %>% fct_relabel(\(z) str_sub(z, 1, 50))),
    ) %>%
    arrange(perturbed, compound, rescue_definition) %>%
    mutate(
      across(
        rescue_type, fct_inorder
      )
    ),
  aes(
    x = rescue_type,
    y = term_unique,
    fill = {{z}}
  )
) +
  geom_raster() +
  ggplot2::scale_fill_gradientn(
    colours = c("#2166ac", "white", "#b2182b"),
    # limits = c(quantile(pull(topgo_res_plotting, {{z}}), .01), 1),
    limits = {
      # browser()
      abs_max <- max(abs(quantile(pull(topgo_res_plotting, {{z}}), c(.01, .99))))
      # scales::rescale(c(-abs_max, 0, abs_max))
      c(1/abs_max, abs_max)
    },
    values = {
      abs_max <- max(abs(quantile(pull(topgo_res_plotting, {{z}}), c(.01, .99))))
      scales::rescale(log10(c(1/abs_max, 1, abs_max)))
    },
    oob = scales::squish,
    # labels = c("p > 0.05", "p = 0.05", "p <= 0.05"),
    trans = "log10"
  ) +
  # ggplot2::scale_fill_gradientn(
  #   colours = rev(c("#2166ac", "#d1e5f0", "#fddbc7", "#b2182b")),
  #   limits = c(quantile(topgo_res_plotting$fisher_double, .01), 1),
  #   values = scales::rescale(log10(c(quantile(topgo_res_plotting$fisher_double, .01), 0.04999, 0.05, 1))),
  #   oob = scales::squish,
  #   # labels = c("p > 0.05", "p = 0.05", "p <= 0.05"),
  #   trans = "log10"
  # ) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  )
  # scale_fill_viridis_c(direction = -1, trans = "log10")

ps <- map(
  names(topgo_res_plotting_clust) %>%
    set_names(),
  \(x, y) plot_big_heatmap(x, odds_ratio, topgo_res_plotting)
)

iwalk(
  ps,
  \(p, name) {
    ggsave(
      file.path("plots", paste0("topgo_big_heatmap_", name, "_or.pdf")),
      p, width = 12, height = 30
    )
  }
)

ggsave(
  "topgo_test.pdf", p,
  width = 12, height = 30
)

topgo_res_plotting_picks <- topgo_res_plotting %>%
  mutate(
    signif = fisher_double < .05
  ) %>%
  group_by(
    ontology, GO.ID
  ) %>%
  filter(
    is_empty(
      setdiff(c("GNE-495", "PTX"), compound[signif])
    ) | is_empty(
      setdiff(c("JNK-inh", "PTX"), compound[signif])
    )
  ) %>%
  ungroup()


ps <- map(
  names(topgo_res_plotting_clust) %>%
    set_names(),
  \(x, y) plot_big_heatmap(x, odds_ratio, topgo_res_plotting_picks)
)

iwalk(
  ps,
  \(p, name) {
    ggsave(
      file.path("plots", paste0("topgo_big_heatmap_picks_", name, "2.pdf")),
      p, width = 12, height = 35
    )
  }
)
```




```{r}
topgo_res_plotting_picks_mat <- topgo_res_plotting_picks %>%
  mutate(
    perturbed = factor(perturbed, levels = c("up", "down")),
    compound = factor(compound, levels = c("PTX", "JNK-inh", "GNE-495")),
    rescue_definition = factor(rescue_definition, levels = c("return_baseline", "significant_opposite", "overshot")),
    across(term_unique, \(x) factor(x, levels = levels(topgo_res_plotting_clust$term_unique))),
  ) %>%
  arrange(perturbed, compound, rescue_definition) %>%
  mutate(
    across(
      rescue_type, fct_inorder
    )
  ) %>%
  transmute(GO.ID, across(fisher_double, log10), rescue_type) %>%
  pivot_wider(
    names_from = rescue_type,
    values_from = fisher_double
  ) %>%
  column_to_rownames("GO.ID") %>%
  as.matrix()

topgo_res_plotting_picks_dist <- dist(topgo_res_plotting_picks_mat)
topgo_res_plotting_picks_clust <- hclust(topgo_res_plotting_picks_dist, method = "average") %>%
  reorder(topgo_res_plotting_picks_dist, method = "olo")

ss <- map_dbl(
  2:50,
  \(k) {
    x <- cutree(topgo_res_plotting_picks_clust, k = k) %>%
      silhouette(topgo_res_plotting_picks_dist)
    mean(x[, 3])
  }
)
```

```{r}
createAnnotation <- function(df, colors = list(), which = "column") {
  # Load necessary libraries
  color_maps <- list()

  # Iterate over each column in the dataframe
  for (col_name in names(df)) {
    # Check if the column is numeric
    if (is.numeric(df[[col_name]])) {
      # Create a color mapping function for numeric columns
      if (is.null(colors[[col_name]]))
        colors[[col_name]] <- c("white", "red")
      if (is.function(colors[[col_name]])) {
        col_fun <- colors[[col_name]]
      } else if (is.character(colors[[col_name]])) {
        n <- length(colors[[col_name]])
        if (n == 1) {
          col_fun <- circlize::colorRamp2(
            seq(from = min(df[[col_name]]), to = max(df[[col_name]]), length.out = 100),
            viridis(100, option = colors[[col_name]])
          )
        } else {
          col_fun <- circlize::colorRamp2(
            seq(from = min(df[[col_name]]), to = max(df[[col_name]]), length.out = n),
            colors[[col_name]]
          )
        }
      } else
        stop("Dont know how to handle colors for column ", col_name)
      color_maps[[col_name]] <- col_fun
    } else {
      if (is.character(colors[[col_name]])) {
        color_maps[[col_name]] <- colors[[col_name]]
        next
      }
      col_fun <- colors[[col_name]] %||% rainbow
      # Create a named vector of colors for categorical columns
      values <- df[[col_name]]
      unique_values <- if (is.factor(values)) levels(values) else unique(df[[col_name]])
      named_colors <- setNames(col_fun(length(unique_values)), unique_values)
      color_maps[[col_name]] <- named_colors
    }
  }

  # Combine all annotations
  combined_annotations <- HeatmapAnnotation(df = df, col = color_maps, which = which)

  return(combined_annotations)
}
```


```{r}
library(ComplexHeatmap)
library(viridisLite)

clust_cut <- cutree(topgo_res_plotting_picks_clust, k = 12)

hm_df <- enframe(clust_cut, name = "GO.ID", value = "cluster") %>%
  mutate(across(cluster, \(x) fct_inseq(as.character(x)))) %>%
  column_to_rownames("GO.ID")

hm <- Heatmap(
  topgo_res_plotting_picks_mat,
  name = "p-value",
  cluster_rows = topgo_res_plotting_picks_clust,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  show_column_names = TRUE,
  left_annotation = createAnnotation(hm_df, which = "row"),
  col = circlize::colorRamp2(
    log10(c(quantile(topgo_res_plotting$fisher_double, .01), 0.04999, 0.05, 1)),
    rev(c("#2166ac", "#d1e5f0", "#fddbc7", "#b2182b"))
  )
)

  ggplot2::scale_fill_gradientn(
    colours = rev(c("#2166ac", "#d1e5f0", "#fddbc7", "#b2182b")),
    limits = c(quantile(topgo_res_plotting$fisher_double, .01), 1),
    values = scales::rescale(log10(c(quantile(topgo_res_plotting$fisher_double, .01), 0.04999, 0.05, 1))),
    oob = scales::squish,
    # labels = c("p > 0.05", "p = 0.05", "p <= 0.05"),
    trans = "log10"
  ) +
```


```{r}
compound_specific_go_rescue <- topgo_res %>%
  group_by(
    GO.ID, Term, term_unique, perturbed
  ) %>%
  filter(
    fisher_double[compound == "PTX"] < .05
  ) %>%
  summarize(
    compound_rescues = case_when(
      any(fisher_double[compound == "GNE-495"] < .05) & any(fisher_double[compound == "JNK-inh"] < .05) ~ "both",
      any(fisher_double[compound == "GNE-495"] < .05) ~ "GNE-495",
      any(fisher_double[compound == "JNK-inh"] < .05) ~ "JNK-inh",
      TRUE ~ "none"
    ),
    rescue_definitions = paste(sort(rescue_definition[fisher_double < .05]), collapse = ", "),
    .groups = "drop"
  )

compound_specific_go_rescue %>%
  dplyr::count(perturbed, compound_rescues)
```



```{r}
picked_clusters <- "chondrocyte proliferation_GO:0035988
negative regulation of ubiquitinprotein transferase activity_GO:0051444
temperature homeostasis_GO:0001659
chondrocyte differentiation_GO:0002062
cardiac left ventricle morphogenesis_GO:0003214
negative regulation of ubiquitindependent protein catabolic process_GO:2000059
negative regulation of G1/S transition of mitotic cell cycle_GO:2000134
positive regulation of stem cell proliferation_GO:2000648
endoderm development_GO:0007492
negative regulation of cellmatrix adhesion_GO:0001953
neural crest cell differentiation_GO:0014033
presynaptic membrane organization_GO:0097090
negative regulation of wound healing_GO:0061045
ventricular septum morphogenesis_GO:0060412
axon extension involved in axon guidance_GO:0048846
aorta morphogenesis_GO:0035909
negative regulation of apoptotic process_GO:0043066
regulation of basement membrane organization_GO:0110011
hippo signaling_GO:0035329
formation of cytoplasmic translation initiation complex_GO:0001732" %>%
  str_match_all(
    "(GO:[0-9]+)"
  )


printGenes(
  topgo_objects[["BP"]],
  picked_clusters[[1]][, 1],

)
```


```{r}
# single type per gene
ptx_rescue_type_list_single <- ptx_rescue_type_list %>%
  filter(rescue_definition %in% c("significant_opposite", "return_baseline")) %>%
  group_by(gene_id) %>%
  arrange(
    gene_id,
    match(
      rescue_type, c("both", "JNK-inh only", "GNE-495 only", "no rescue", "not perturbed")
    )
  ) %>%
  slice_head(n = 1) %>%
  ungroup()

plot_gene_set <- function(selected_genes) {
  condition_order <- c(
    "DMSO" = "DMSO", "PTX" = "PTX",
    "JNKinhPTX" = "PTX + JNK-Inh", "GNE495PTX" = "PTX + GNE-495",
    "JNKinhDMSO" = "JNK-Inh", "GNE495DMSO" = "GNE-495"
  )
  df <- rlog_scaled2[intersect(selected_genes, rownames(rlog_scaled2)), ] %>%
    as_tibble(rownames = "gene_id") %>%
    pivot_longer(
      -gene_id,
      names_to = "sample_id",
      values_to = "value"
    ) %>%
    power_inner_join(
      de_meta,
      by = "sample_id",
      check = check_specs(
        unmatched_keys_left = "warn",
        duplicate_keys_right = "warn"
      )
    ) %>%
    power_inner_join(
      ptx_rescue_type_list_single,
      by = "gene_id",
      check = check_specs(
        unmatched_keys_left = "warn",
        duplicate_keys_right = "warn"
      )
    ) %>%
    power_inner_join(
      gene_id_gene_symbol_map,
      by = "gene_id",
      check = check_specs(
        unmatched_keys_left = "warn",
        duplicate_keys_right = "warn"
      )
    ) %>%
    mutate(
      condition = factor(recode(condition, !!!condition_order), levels = unname(condition_order)),
    )
  # browser()
  ggplot(
    df,
    aes(
      x = condition,
      y = value,
      color = rescue_type
    )
  ) +
    ggbeeswarm::geom_quasirandom(
      aes(
        group = rescue_type,
        text = paste0(
          gene_name, " (", gene_id, ")\n",
          "Rescue type: ", rescue_type, "\n",
          "Condition: ", condition, "\n",
          "Expression: ", signif(value, 2),
          " (", sample_id, ")"
        )
      ),
      shape = 16, dodge.width = .8
    ) +
    scale_color_manual(
      values = rescue_type_colors
    ) +
    geom_boxplot(
      fill = NA,
      outlier.shape = NA,
      position = position_dodge(width = .8)
    ) +
    labs(
      x = NULL, y = "Scaled expression",
      color = "Rescue type"
    )
    # scale_color_manual(
    #   values = c("highlighted" = "red", "background" = "gray50")
    # )
}


plot_go_set <- function(go_id) {
  gene_set <- go_gene_map[[paste0("BP.", go_id)]]
  plot_gene_set(gene_set) +
    labs(
      title = go_meta %>%
        filter(GO.ID == go_id) %>%
        pull(Term) %>%
        paste0(., " (", go_id, ")")
    )
}

p <- plot_go_set("GO:0097090")

ggsave(
  "plots/scaled_expression_presynaptic_membrane_organization.pdf",
  p, width = 8, height = 6
)

plotly::ggplotly(p, tooltip = "text") %>%
  plotly::layout(boxmode = "group") %>%
  htmlwidgets::saveWidget(
    "plots/scaled_expression_presynaptic_membrane_organization.html",
    selfcontained = TRUE
  )

example_go <- "GO:0097090"
example_background <- go_gene_map[[paste0("BP.", "GO:0097090")]]

p <- plot_gene_set(example_background) +
  labs(
    title = go_meta %>%
      filter(GO.ID == example_go) %>%
      pull(Term) %>%
      paste0(., " (", example_go, ")")
  )

plotly::ggplotly(p, tooltip = "text") %>%
  plotly::layout(boxmode = "group")

p <- plot_go_set("GO:0070527")

ggsave(
  "plots/scaled_expression_platelet_aggregation.pdf",
  p, width = 8, height = 6
)
plotly::ggplotly(p, tooltip = "text") %>%
  plotly::layout(boxmode = "group") %>%
  htmlwidgets::saveWidget(
    "plots/scaled_expression_platelet_aggregation.html",
    selfcontained = TRUE
  )
```


```{r}
library(ComplexHeatmap)
plot_gene_set_hm <- function(selected_genes) {
  condition_order <- c(
    "DMSO" = "DMSO", "PTX" = "PTX",
    "JNKinhPTX" = "PTX + JNK-Inh", "GNE495PTX" = "PTX + GNE-495",
    "JNKinhDMSO" = "JNK-Inh", "GNE495DMSO" = "GNE-495"
  )
  col_meta <- de_meta %>%
    mutate(
      condition = factor(recode(condition, !!!condition_order), levels = unname(condition_order))
    ) %>%
    arrange(condition, sample_id)

  selected_available_genes <- intersect(selected_genes, rownames(rlog_scaled2))
  row_meta <- ptx_rescue_type_list_single %>%
    filter(gene_id %in% selected_available_genes) %>%
    power_inner_join(
      gene_id_gene_symbol_map,
      by = "gene_id",
      check = check_specs(
        unmatched_keys_left = "warn",
        duplicate_keys_right = "warn"
      )
    ) %>%
    dplyr::select(
      gene_id, gene_name, rescue_type
    )

  mat <- rlog_scaled2[row_meta$gene_id, col_meta$sample_id]
  rownames(mat) <- row_meta$gene_name

  Heatmap(
    mat,
    name = "Scaled expression",
    cluster_rows = cluster_mat,
    cluster_columns = FALSE,
    show_row_names = TRUE,
    show_column_names = FALSE,
    column_split = col_meta$condition,
    left_annotation = HeatmapAnnotation(
      df = row_meta %>%
        dplyr::select(gene_name, rescue_type) %>%
        column_to_rownames("gene_name"),
      col = list(
        rescue_type = rescue_type_colors
      ),
      which = "row"
    )
  )
}


plot_go_set_hm <- function(go_id) {
  gene_set <- go_gene_map[[paste0("BP.", go_id)]]
  plot_gene_set_hm(gene_set)
}

hm <- plot_go_set_hm("GO:0097090")

hm <- plot_go_set_hm("GO:0070527")
```


## gProfiler2

```{r}
library(gprofiler2)

gprofiler_input <- ptx_rescue_gene_sets_short %>%
  mutate(
    gene_set_name = paste(
      rescue_definition, compound, perturbed, sep = "_"
    )
  ) %>%
  power_left_join(
    distinct(de_long, gene_id, gene_name),
    by = "gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  group_by(compound, perturbed, rescue_definition, gene_set_name) %>%
  summarize(
    gene_ids = list(gene_id),
    gene_names = list(gene_name),
    .groups = "drop"
  )

gprofiler_res_raw <- gost(
  with(
    gprofiler_input,
    set_names(
      gene_ids,
      gene_set_name
    )
  ),
  significant = FALSE
)

qsave(
  gprofiler_res_raw,
  "gprofiler_res_raw.qs"
)
gprofiler_res_raw <- qread("gprofiler_res_raw.qs")

# gostplot(gprofiler_res_raw)

gprofiler_res <- gprofiler_res_raw$result %>%
  as_tibble() %>%
  dplyr::rename(gene_set_name = query) %>%
  power_inner_join(
    gprofiler_input %>%
      dplyr::select(compound, perturbed, rescue_definition, gene_set_name),
    by = "gene_set_name",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

write_csv(
  gprofiler_res,
  "gprofiler_res.csv.gz"
)
```



```{r}
gprofiler_res %>%
  count(source)

gprofiler_plotting_res_tf <- gprofiler_res %>%
  filter(
    source == "TF"
  ) %>%
  group_by(
    term_id
  ) %>%
  filter(any(p_value < .05)) %>%
  ungroup() %>%
  mutate(
    rescue_type = paste(perturbed, compound, rescue_definition, sep = "_")
  )

plot_big_heatmap_gprofiler <- \(df, y_var, fill_var, log_trans = FALSE) {
  df_clust <- df %>%
    dplyr::transmute(rescue_type, {{y_var}}, if (log_trans) across({{fill_var}}, \(x) -log10(x)) else {{fill_var}}) %>%
    cluster_df({{y_var}}, rescue_type, {{fill_var}})
  df_trans <- df %>%
    mutate(
      perturbed = factor(perturbed, levels = c("up", "down")),
      compound = factor(compound, levels = c("PTX", "JNK-inh", "GNE-495")),
      rescue_definition = factor(rescue_definition, levels = c("return_baseline", "significant_opposite", "overshot")),
      across({{y_var}}, \(y) factor(y, levels = levels(pull(df_clust, {{y_var}}))) %>% fct_relabel(\(z) str_sub(z, 1, 50))),
    ) %>%
    arrange(perturbed, compound, rescue_definition) %>%
    mutate(
      across(
        rescue_type, fct_inorder
      )
    )
  ggplot(
    df_trans,
    aes(
      x = rescue_type,
      y = {{y_var}},
      fill = {{fill_var}}
    )
  ) +
    geom_raster() +
    scale_fill_viridis_c(direction = -1) +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
    )
}

plot_big_heatmap_gprofiler_pval <- \(df, y_var, fill_var) {
  p_val_scale <- ggplot2::scale_fill_gradientn(
    colours = rev(c("#2166ac", "#d1e5f0", "#fddbc7", "#b2182b")),
    limits = c(quantile(pull(df, {{fill_var}}), .01), 1),
    values = scales::rescale(log10(c(quantile(pull(df, {{fill_var}}), .01), 0.04999, 0.05, 1))),
    oob = scales::squish,
    # labels = c("p > 0.05", "p = 0.05", "p <= 0.05"),
    trans = "log10"
  )
  plot_big_heatmap_gprofiler(df, {{y_var}}, {{fill_var}}, log_trans = TRUE) + p_val_scale
}
  # scale_fill_viridis_c(direction = -1, trans = "log10")

p <- plot_big_heatmap_gprofiler_pval(gprofiler_plotting_res_tf, TF, p_value)

gprofiler_plotting_res_tf_combined <- gprofiler_plotting_res_tf %>%
  separate_wider_regex(
    term_name,
    c("Factor: ", TF = "[^;]+", "; motif: ", motif = "[^;]+", ";?.*")
  ) %>%
  group_by(
    TF, compound, perturbed, rescue_definition, rescue_type
  ) %>%
  summarize(
    p_value = psych::harmonic.mean(p_value),
    .groups = "drop"
  )


p <- plot_big_heatmap_gprofiler_pval(gprofiler_plotting_res_tf_combined, TF, p_value)

ggsave(
  "plots/gprofiler_hm_tf_combined.pdf",
  p, width = 8, height = 25
)
# SCRT1

chea3_res %>%
  filter(TF == "SCRT1") %>%
  View()
```

## Chea3

```{r}
chea3_submit <- function(N  , name = "myquery") {
  library(httr)
  library(jsonlite)

  url <- "https://maayanlab.cloud/chea3/api/enrich/"
  payload <- list(
    query_name = name,
    gene_set = genes
  )

  response <- POST(url = url, body = payload, encode = "json")
  json <- content(response, "text")

  fromJSON(json)
}

chea3_res_raw <- gprofiler_input %>%
  mutate(
    res = map(
      gene_names,
      \(x) {
        res <- chea3_submit(x)
        Sys.sleep(3)
        res
      }
    )
  )

qsave(
  chea3_res_raw,
  "chea3_res_raw.qs"
)

chea3_res <- chea3_res_raw %>%
  mutate(
    res = map(
      res,
      \(x) bind_rows(x, .id = "table")
    )
  ) %>%
  select(-gene_ids, -gene_names) %>%
  unnest(res)

write_csv(
  chea3_res,
  "chea3_res.csv.gz"
)
chea3_res <- read_csv("chea3_res.csv.gz")
```


```{r}
chea3_res %>%
  distinct(table)

chea3_res %>%
  filter(table == "Integrated--meanRank") %>%
  pull(Score) %>%
  hist()


chea3_res_plotting_integrated_mean <- chea3_res %>%
  filter(
    table == "Integrated--meanRank"
  ) %>%
  group_by(
    TF
  ) %>%
  filter(any(Score < 100)) %>%
  ungroup() %>%
  mutate(
    rescue_type = paste(perturbed, compound, rescue_definition, sep = "_")
  )

chea3_res_plotting_remap <- chea3_res %>%
  filter(
    table == "ReMap--ChIP-seq"
  ) %>%
  group_by(
    TF
  ) %>%
  filter(any(FDR < 0.05)) %>%
  ungroup() %>%
  mutate(
    rescue_type = paste(perturbed, compound, rescue_definition, sep = "_")
  )


chea3_res_plotting_encode <- chea3_res %>%
  filter(
    table == "ENCODE--ChIP-seq"
  ) %>%
  group_by(
    TF
  ) %>%
  filter(any(FDR < 0.05)) %>%
  ungroup() %>%
  mutate(
    rescue_type = paste(perturbed, compound, rescue_definition, sep = "_")
  )



plot_big_heatmap_chea3 <- \(df, fill_var) {
  df_clust <- df %>%
    dplyr::transmute(rescue_type, TF, {{fill_var}}) %>%
    cluster_df(TF, rescue_type, {{fill_var}})
  df_trans <- df %>%
    mutate(
      perturbed = factor(perturbed, levels = c("up", "down")),
      compound = factor(compound, levels = c("PTX", "JNK-inh", "GNE-495")),
      rescue_definition = factor(rescue_definition, levels = c("return_baseline", "significant_opposite", "overshot")),
      across(TF, \(y) factor(y, levels = levels(df_clust$TF)) %>% fct_relabel(\(z) str_sub(z, 1, 50))),
    ) %>%
    arrange(perturbed, compound, rescue_definition) %>%
    mutate(
      across(
        rescue_type, fct_inorder
      )
    )
  ggplot(
    df_trans,
    aes(
      x = rescue_type,
      y = TF,
      fill = {{fill_var}}
    )
  ) +
    geom_raster() +
    scale_fill_viridis_c(direction = -1) +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
    )
}

plot_big_heatmap_chea3_pval <- \(df, fill_var) {
  p_val_scale <- ggplot2::scale_fill_gradientn(
    colours = rev(c("#2166ac", "#d1e5f0", "#fddbc7", "#b2182b")),
    limits = c(quantile(pull(df, {{fill_var}}), .01), 1),
    values = scales::rescale(log10(c(quantile(pull(df, {{fill_var}}), .01), 0.04999, 0.05, 1))),
    oob = scales::squish,
    # labels = c("p > 0.05", "p = 0.05", "p <= 0.05"),
    trans = "log10"
  )
  plot_big_heatmap_chea3(df, {{fill_var}}) + p_val_scale
}
  # scale_fill_viridis_c(direction = -1, trans = "log10")

p <- plot_big_heatmap_chea3(chea3_res_plotting_integrated_mean, Score)
p

ggsave(
  "plots/chea3_hm_integrated_mean.pdf",
  p +
    scale_fill_viridis_c(direction = -1, trans = "log10"), width = 8, height = 15
)

p <- plot_big_heatmap_chea3_pval(chea3_res_plotting_remap, FDR)


ggsave(
  "plots/chea3_hm_remap.pdf",
  p,
  width = 8, height = 15
)

p <- plot_big_heatmap_chea3_pval(chea3_res_plotting_encode, FDR)

ggsave(
  "plots/chea3_hm_encode.pdf",
  p,
  width = 8, height = 15
)

# SCRT1

chea3_res %>%
  filter(TF == "SCRT1") %>%
  View()
```

## Plot selected gene counts

```{r}
fpkm_raw <- syn("syn63854603") %>%
  read_csv()

fpkm_long <- fpkm_raw %>%
  pivot_longer(
    -c(gene_id, gene_name, gene_type, Description, chromosome, strand),
    names_to = c("sample_id", "condition", "metric"),
    names_pattern = "(S[0-9]+?)_(.*?) ?\\.(.*)",
    values_to = "value"
  )

fpkm_long$condition %>%
  unique()
```

```{r}
selected_genes <- withr::with_seed(
  42,
  ptx_rescue_type_list %>%
    group_by(
      rescue_definition,
      rescue_type,
      perturbed
    ) %>%
    slice_sample(n = 3) %>%
    ungroup()
)


conditions_of_interest <- c(
  "DMSO",
  "PTX",
  "JNKinhPTX",
  "GNE495PTX"
)

library(ggbeeswarm)

selected_genes_fpkm <- selected_genes %>%
  inner_join(
    fpkm_long %>%
      filter(
        condition %in% conditions_of_interest
      ) %>%
      mutate(across(condition, \(x) factor(x, levels = conditions_of_interest))),
    by = "gene_id"
  )

ps <- selected_genes_fpkm %>%
  group_nest(
    rescue_definition
  ) %>%
  mutate(
    p = map(
      data,
      \(x) ggplot(
        x %>%
          arrange(rescue_type, perturbed) %>%
          mutate(
            gene_name = fct_inorder(gene_name)
          ),
        aes(
          x = condition,
          y = value,
        )
      ) +
        geom_rect(
          aes(fill = rescue_type),
          xmin = -Inf, xmax = Inf, ymin = -Inf, ymax = Inf, alpha = 0.3,
          inherit.aes = FALSE,
          data = \(x) x %>%
            distinct(gene_name, rescue_type)
        ) +
        geom_beeswarm() +
        geom_boxplot(
          outlier.shape = NA,
          fill = NA
        ) +
        scale_fill_manual(
          values = rescue_type_colors
        ) +
        ggh4x::facet_wrap2(~gene_name, scales = "free_y", axes = "y") +
        theme(
          axis.text.x = element_text(angle = 45, hjust = 1)
        ) +
        labs(y = "FPKM")
          )
  )

pwalk(
  ps,
  \(p, rescue_definition, ...) ggsave(
    file.path("plots", paste0("selected_genes_fpkm_", rescue_definition, ".pdf")),
    p, width = 8, height = 6, device = cairo_pdf
  )
)

ggsave(
  file.path("plots", "selected_genes_fpkm.pdf"),
  p, width = 8, height = 6, device = cairo_pdf
)
```


GNE-495 alone

## Alluvial diagram of rescued genes by definition

```{r}
library(ggalluvial)

ptx_rescue_type_list %>%
  count(perturbed)

ptx_rescue_alluvial_plots <- ptx_rescue_type_list %>%
  group_nest(perturbed_threshold, batch, rescue_definition) %>%
  mutate(
    alluvial_data = map(
      data,
      \(x) count(
        x,
        `PTX perturbed` = perturbed,
        `GNE-495 perturbed` = `GNE-495`,
        `JNK-inh perturbed` = `JNK-inh`,
        rescue_type
      ) %>%
        filter(`PTX perturbed` != "not perturbed")
    ),
    lode_data = map(
      alluvial_data,
      \(x) to_lodes_form(
        x,
        ends_with("perturbed"),
        key = "Contrast", value = "Gene set"
      )
    ),
    p = map(
      lode_data,
      \(x)
      ggplot(
        x,
        aes(
          x = Contrast, y = n, stratum = `Gene set`, alluvium = alluvium
        )
      ) +
        scale_x_discrete(expand = c(0.05, 0.05)) +
        geom_alluvium(
          aes(
            # color = `PTX perturbed 2`,
            fill = rescue_type
          ),
          width = 1/10
        ) +
        # ggokabeito::scale_color_okabe_ito(order = c(2, 1), guide = "none") +
        scale_fill_manual(
          values = rescue_type_colors
        ) +
        labs(x = NULL) +
        # scale_fill_manual(
        #     values = c(existing = "#a2a2a2", added = "#c81919"),
        #     guide = "none"
        # ) +
        geom_stratum(width = 1/6) +
        # geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
        geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
        geom_text(
          aes(
            x = stage(Contrast, after_scale = x + .12),
            label = n
          ),
          stat = StatAlluvium,
          hjust = 0,
          data = \(x) filter(x, as.integer(Contrast) == 1L)
        ) +
        theme_minimal() +
        # labs(y = "Data points", fill = NA) +
        theme(
            panel.grid.major.x = element_blank()
        ) +
        coord_cartesian(clip = "off")
    )
  )

pwalk(
  ptx_rescue_alluvial_plots,
  \(p, perturbed_threshold, batch, rescue_definition, ...) ggsave(
    file.path("plots", paste0("ptx_rescue_threshold_", perturbed_threshold, "_", rescue_definition, "_alluvium_", batch, ".pdf")),
    p, width = 6, height = 4
  )
)

ggsave(
  file.path("plots", "ptx_rescue_no_threshold_alluvium.pdf"),
  p, width = 6, height = 4
)
```


```{r}
library(ggalluvial)

p <- ggplot(
    pax_rescue_alluvials %>%
      filter(PTX_perturbed != "not perturbed"),
    aes(
      axis1 = `GNE-495 + PTX_perturbed`,
      axis2 = `JNK-inh + PTX_perturbed`,
      axis3 = `GNE-495 + PTX_rescue`,
      axis4 = `JNK-inh + PTX_rescue`,
      y = n
    )
  ) +
  # scale_x_discrete(limits = c("Class", "Sex", "Age"), expand = c(.2, .05)) +
  # xlab("Demographic") +
  # geom_alluvium(aes(fill = Survived)) +
  geom_alluvium() +
  geom_stratum() +
  geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  theme_minimal()


p <- ggplot(
    pax_rescue_alluvials %>%
      filter(PTX_perturbed != "not perturbed") %>%
      mutate(
        across(
          ends_with("_rescue"),
          \(x) str_replace(x, "_unrecovered", "") %>%
            str_replace("up_recovered", "recovered") %>%
            str_replace("down_recovered", "recovered")
        ),
      ),
    aes(
      axis1 = `PTX_perturbed`,
      axis2 = `GNE-495 + PTX_rescue`,
      axis3 = `JNK-inh + PTX_rescue`,
      y = n
    )
  ) +
  # scale_x_discrete(limits = c("Class", "Sex", "Age"), expand = c(.2, .05)) +
  # xlab("Demographic") +
  # geom_alluvium(aes(fill = Survived)) +
  geom_alluvium(
    aes(fill = PTX_perturbed)
  ) +
  ggokabeito::scale_fill_okabe_ito(order = c(2, 1)) +
  geom_stratum() +
  geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  theme_minimal()


p <- ggplot(
    pax_rescue_alluvials %>%
      filter(PTX_perturbed != "not perturbed"),
    aes(
      axis1 = `GNE-495 + PTX_perturbed`,
      axis2 = `JNK-inh + PTX_perturbed`,
      axis3 = `PTX_perturbed`,
      y = n
    )
  ) +
  # scale_x_discrete(limits = c("Class", "Sex", "Age"), expand = c(.2, .05)) +
  # xlab("Demographic") +
  # geom_alluvium(aes(fill = Survived)) +
  geom_alluvium() +
  geom_stratum() +
  geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  theme_minimal()

ggsave(

)
```


```{r}
pax_rescue_list_simple_dmso <- pax_rescue_list %>%
  filter(`PTX perturbed` != "not perturbed") %>%
  transmute(
    gene_id, gene_name, gene_type,
    across(ends_with(" perturbed")),
    across(
      ends_with(" rescue DMSO"),
      # \(x) str_replace(x, " unrescued", "") %>%
      \(x) str_replace(x, "up unrescued", "no rescue") %>%
        str_replace("down unrescued", "no rescue") %>%
        str_replace("up rescued", "rescued") %>%
        str_replace("down rescued", "rescued")
    ),
    `Rescue type` = case_when(
      `GNE-495 + PTX rescue DMSO` == "rescued" & `JNK-inh + PTX rescue DMSO` == "rescued" ~ "both",
      `GNE-495 + PTX rescue DMSO` == "rescued" ~ "GNE-495 only rescue",
      `JNK-inh + PTX rescue DMSO` == "rescued" ~ "JNK-inh only rescue",
      TRUE ~ "no rescue"
    )
  )

```


```{r}
ptx_rescue_plot_data_w_type <- ptx_rescue_plot_data %>%
  left_join(
    pax_rescue_list_simple_dmso %>%
      select(-gene_name, -gene_type),
    by = "gene_id"
  ) %>%
  mutate(
    `Rescue type` = replace_na(`Rescue type`, "not perturbed") %>%
      factor(
        levels = c(
          "both",
          "JNK-inh only rescue",
          "GNE-495 only rescue",
          "no rescue",
          "not perturbed"
        )
      )
  )

p <- ggplot(
  ptx_rescue_plot_data_w_type %>%
    filter(`Rescue type` != "not perturbed") %>%
    arrange(desc(`Rescue type`)),
  aes(
    x = signed_padj_1_clamped,
    y = signed_padj_2_clamped,
    shape = signed_padj_symbol,
    color = `Rescue type`
  )
) +
  geom_abline() +
  geom_point(
    alpha = 0.5
  ) +
  scale_shape_manual(
    values = c(point = "\u25CF", right = "\u25BA", left = "\u25C4", down = "\u25BC", up = "\u25B2"),
    guide = "none"
  ) +
  scale_color_manual(
    values = c(
      both = "black",
      `JNK-inh only rescue` = "purple",
      `GNE-495 only rescue` = "green",
      `no rescue` = "gray85"
    )
  ) +
  ggh4x::facet_wrap2(
    ~contrast_2,
    scales = "free", axes = "all"
  ) +
  labs(
    x = "Signed log10(padj) Paclitaxel",
    y = "Signed log10(padj) Rescue"
  )

dir.create("plots", showWarnings = FALSE)
ggsave(
  file.path("plots", "ptx_rescue_scatter_w_rescue_type_dmso.pdf"),
  p, width = 8, height = 4, device = cairo_pdf
)


p <- ggplot(
  ptx_rescue_plot_data_w_type %>%
    filter(`Rescue type` != "not perturbed") %>%
    arrange(desc(`Rescue type`)),
  aes(
    x = logFC_1,
    y = logFC_2,
    # shape = signed_padj_symbol,
    color = `Rescue type`
  )
) +
  geom_abline() +
  geom_point(
    alpha = 0.5,
    shape = "circle small"
  ) +
  scale_color_manual(
    values = c(
      both = "black",
      `JNK-inh only rescue` = "purple",
      `GNE-495 only rescue` = "green",
      `no rescue` = "gray85"
    )
  ) +
  ggh4x::facet_wrap2(
    ~contrast_2,
    scales = "free", axes = "all"
  ) +
  labs(
    x = "log2(fold-change) Paclitaxel",
    y = "log2(fold-change) Rescue"
  )

dir.create("plots", showWarnings = FALSE)
ggsave(
  file.path("plots", "ptx_rescue_scatter_logFC_w_rescue_type_dmso.pdf"),
  p, width = 8, height = 4, device = cairo_pdf
)
```



```{r}
pax_rescue_list_simple_dmso_plus <- pax_rescue_list %>%
  filter(`PTX perturbed` != "not perturbed") %>%
  transmute(
    gene_id, gene_name, gene_type,
    across(ends_with(" perturbed")),
    across(
      ends_with(" rescue DMSO plus"),
      # \(x) str_replace(x, " unrescued", "") %>%
      \(x) str_replace(x, "up unrescued", "no rescue") %>%
        str_replace("down unrescued", "no rescue") %>%
        str_replace("up rescued", "rescued") %>%
        str_replace("down rescued", "rescued")
    ),
    `Rescue type` = case_when(
      `GNE-495 + PTX rescue DMSO plus` == "rescued" & `JNK-inh + PTX rescue DMSO plus` == "rescued" ~ "both",
      `GNE-495 + PTX rescue DMSO plus` == "rescued" ~ "GNE-495 only rescue",
      `JNK-inh + PTX rescue DMSO plus` == "rescued" ~ "JNK-inh only rescue",
      TRUE ~ "no rescue"
    )
  )

```


```{r}
ptx_rescue_plot_data_w_type <- ptx_rescue_plot_data %>%
  left_join(
    pax_rescue_list_simple_dmso_plus %>%
      select(-gene_name, -gene_type),
    by = "gene_id"
  ) %>%
  mutate(
    `Rescue type` = replace_na(`Rescue type`, "not perturbed") %>%
      factor(
        levels = c(
          "both",
          "JNK-inh only rescue",
          "GNE-495 only rescue",
          "no rescue",
          "not perturbed"
        )
      )
  )

p <- ggplot(
  ptx_rescue_plot_data_w_type %>%
    filter(`Rescue type` != "not perturbed") %>%
    arrange(desc(`Rescue type`)),
  aes(
    x = signed_padj_1_clamped,
    y = signed_padj_2_clamped,
    shape = signed_padj_symbol,
    color = `Rescue type`
  )
) +
  geom_abline() +
  geom_point(
    alpha = 0.5
  ) +
  scale_shape_manual(
    values = c(point = "\u25CF", right = "\u25BA", left = "\u25C4", down = "\u25BC", up = "\u25B2"),
    guide = "none"
  ) +
  scale_color_manual(
    values = c(
      both = "black",
      `JNK-inh only rescue` = "purple",
      `GNE-495 only rescue` = "green",
      `no rescue` = "gray85"
    )
  ) +
  ggh4x::facet_wrap2(
    ~contrast_2,
    scales = "free", axes = "all"
  ) +
  labs(
    x = "Signed log10(padj) Paclitaxel",
    y = "Signed log10(padj) Rescue"
  )

dir.create("plots", showWarnings = FALSE)
ggsave(
  file.path("plots", "ptx_rescue_scatter_w_rescue_type_dmso_plus.pdf"),
  p, width = 8, height = 4, device = cairo_pdf
)


p <- ggplot(
  ptx_rescue_plot_data_w_type %>%
    filter(`Rescue type` != "not perturbed") %>%
    arrange(desc(`Rescue type`)),
  aes(
    x = logFC_1,
    y = logFC_2,
    # shape = signed_padj_symbol,
    color = `Rescue type`
  )
) +
  geom_abline() +
  geom_point(
    alpha = 0.5,
    shape = "circle small"
  ) +
  scale_color_manual(
    values = c(
      both = "black",
      `JNK-inh only rescue` = "purple",
      `GNE-495 only rescue` = "green",
      `no rescue` = "gray85"
    )
  ) +
  ggh4x::facet_wrap2(
    ~contrast_2,
    scales = "free", axes = "all"
  ) +
  labs(
    x = "log2(fold-change) Paclitaxel",
    y = "log2(fold-change) Rescue"
  )

dir.create("plots", showWarnings = FALSE)
ggsave(
  file.path("plots", "ptx_rescue_scatter_logFC_w_rescue_type_dmso_plus.pdf"),
  p, width = 8, height = 4, device = cairo_pdf
)
```

## INDRA queries

```{r}
library(indra.gsea)
library(biomaRt)

ensembl <- useEnsembl(
  biomart = "genes",
  dataset = "hsapiens_gene_ensembl"
)

hgnc_mapping <- getBM(
  attributes = c(
    "ensembl_gene_id",
    "hgnc_id",
    "hgnc_symbol"
  ),
  filters = "ensembl_gene_id",
  values = unique(ptx_rescue_type_list$gene_id),
  mart = ensembl
) %>%
  as_tibble() %>%
  mutate(
    across(hgnc_id, \(x) str_remove(x, "HGNC:"))
  ) %>%
  # Remove empty HGNC IDs
  filter(hgnc_id != "") %>%
  distinct()

ptx_rescue_type_indra_raw <- ptx_rescue_type_list %>%
  filter(rescue_definition == "significant_opposite") %>%
  # Join with HGNC mapping
  inner_join(
    hgnc_mapping,
    by = c("gene_id" = "ensembl_gene_id")
  ) %>%
  group_by(perturbed_threshold, perturbed, rescue_type) %>%
  summarize(
    res = list(
      safely(indra_discrete_gsea)(
        hgnc_id,
        minimum_evidence_count = 2,
        indra_path_analysis = TRUE
      )
    ),
    .groups = "drop"
  )

ptx_rescue_type_indra_raw %>%
  mutate(
    errored = map_lgl(res, \(x) !is.null(x$error))
  )

ptx_rescue_type_indra_res <- ptx_rescue_type_indra_raw %>%
  mutate(
    res = map(res, "result")
  ) %>%
  filter(map_lgl(res, \(x) !is.null(x))) %>%
  mutate(
    res = map(res, \(x) bind_rows(x, .id = "query_type"))
  ) %>%
  unnest(res)

write_csv(
  ptx_rescue_type_indra_res,
  "indra_res.csv.gz"
)

ptx_rescue_type_indra_res %>%
  filter(rescue_type == "GNE-495 only", perturbed == "down", query_type == "indra-upstream", str_starts(curie, fixed("hgnc"))) %>%
  arrange(q) %>%
  View()

ptx_rescue_type_list %>%
  filter(
    rescue_definition == "significant_opposite",
    perturbed == "down",
    rescue_type == "GNE-495 only"
  ) %>%
  inner_join(
    hgnc_mapping,
    by = c("gene_id" = "ensembl_gene_id")
  ) %>%
  pull("hgnc_symbol") %>%
  unique() %>%
  clipr::write_clip()
```

## Save to synapse

```{r}
cipn_syn <- "syn63837896"
analysis_syn <- synMkdir(cipn_syn, "analysis")

synStoreMany(
  c(
    "fgsea_res.qs",
    "fgsea_pert_res.qs",
    "topgo_res.qs",
    "topgo_res_raw.qs",
    "topgo_res_pert_raw.qs",
    "topgo_res_pert.qs",
    "gprofiler_res_raw.qs",
    "gprofiler_res.csv.gz",
    "chea3_res_raw.qs",
    "chea3_res.csv.gz",
    "de_long_selected.csv.gz",
    "feasible_genes.csv.gz",
    "ptx_rescue_overrepresentation.qs",
    "ptx_rescue_intersections_overrepresentation.qs",
    "ptx_rescue_list_long.csv.gz",
    "ptx_rescue_type_list.csv.gz",
    "indra_res.csv.gz",
    "topgo_ptx_rescue_type_list_res.csv.gz",
    "de_meta.csv.gz",
    "de_long_selected_both.csv.gz",
    file.path(
      "results",
      c(
        "clue_res_rescue_gene_sets.csv.gz",
        "clue_res_rescue_gene_sets.tar.gz",
        "ptx_rescue_both_doses_fgsea_overrepresentation.csv.gz",
        "ptx_single_perturbation_overrepresentation.qs",
        "lfc_clusters_overrepresentation.csv.gz"
      )
    )
  ),
  parentId = analysis_syn,
  forceVersion = FALSE
)
```


* Make sure that heatmap with fgsea terms includes *all* not just PTX perturbed
* Look at gene set interactions GNE rescue JNK rescue PTx perturbations
* Excel sheet of all the rescues!
* Update alluvial
* Interactive heatmap
* Massive heatmap with all genes
