---
title: "Deseq analysis of paclitaxel rescue RNA-seq"
author: "Clemens Hug"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(here)
library(tidyverse)
library(data.table)
library(synExtra)
library(qs)
library(powerjoin)
library(DESeq2)
library(conflicted)
library(readxl)
library(tximport)
library(emmeans)
library(ComplexHeatmap)

conflicts_prefer(base::setdiff)
conflicts_prefer(dplyr::filter)
conflicts_prefer(dplyr::select)

synapser::synLogin()
syn <- synDownloader(normalizePath("~/data"), .cache = TRUE)
```

```{r}
raw_counts <- syn("syn69961549") %>%
  read_rds()

meta_batch1 <- syn("syn70686477") %>%
  read_tsv() %>%
  mutate(
    salmon_syn = map_chr(
      run_accession,
      \(x) synPluck("syn69961369", paste0(x, ".tar.gz"))
    ),
    salmon_tar = map_chr(salmon_syn, syn),
    ptx = factor(if_else(str_detect(treatment, coll("PTX")), "yes", "no")),
    gne_495 = factor(if_else(str_detect(treatment, coll("GNE")), "yes", "no")),
    jnk_in = factor(if_else(str_detect(treatment, coll("JNK")), "yes", "no")),
    batch = factor(
      batch,
      levels = paste("Batch", 1:3),
      # Mapping confirmed
      labels = c("2022_12", "2023_10", "2023_07")
    )
  )

meta_raw <- syn("syn69962755") %>%
  read_excel() %>%
  mutate(
    ptx = factor(if_else(str_detect(condition, coll("PTX")), "yes", "no")),
    gne_495 = factor(if_else(str_detect(condition, coll("GNE")), "yes", "no")),
    jnk_in = factor(if_else(str_detect(condition, coll("JNK")), "yes", "no"))
  ) %>%
  nest_join(
    select(
      meta_batch1,
      run_accession, batch, ptx, gne_495, jnk_in
    ),
    by = c("batch", "ptx", "gne_495", "jnk_in"),
    name = "accessions"
  ) %>%
  group_by(batch, ptx, gne_495, jnk_in) %>%
  mutate(
    run_accession = if_else(
      sequencing_batch != 1,
      NA_character_,
      map2_chr(accessions, seq_len(n()), \(x, y) x$run_accession[y])
    ),
    sample_id_sequencing = coalesce(sample_id_sequencing, run_accession)
  ) %>%
  ungroup() %>%
  select(-accessions) %>%
  mutate(
    gne_495_3 = if_else(gne_495 == "yes" & experiment == 1, "yes", "no"),
    jnk_in_3 = if_else(jnk_in == "yes" & experiment == 1, "yes", "no"),
    gne_495_1 = if_else(gne_495 == "yes" & experiment == 2, "yes", "no"),
    jnk_in_1 = if_else(jnk_in == "yes" & experiment == 2, "yes", "no"),
    gne_combined = if_else(
      gne_495 == "no",
      "0",
      if_else(
        experiment == 1,
        "3",
        "1"
      )
    ),
    jnk_combined = if_else(
      jnk_in == "no",
      "0",
      if_else(
        experiment == 1,
        "3",
        "1"
      )
    ),
    across(
      c(experiment, sequencing_batch),
      factor
    ),
    across(
      c(ptx, gne_495, jnk_in, gne_495_3, jnk_in_3, gne_495_1, jnk_in_1),
      \(x) factor(x, levels = c("no", "yes"))
    ),
    condition_simple = paste(
      "PTX", ptx,
      "GNE", gne_combined,
      "JNK", jnk_combined,
      sep = "_"
    )
  )

write_csv(
  meta_raw,
  here("rnaseq", "meta_combined.csv.gz")
)

meta_batch2 <- meta_raw %>%
  filter(
    experiment == 2
  ) %>%
  mutate(
    salmon_syn = map_chr(
      sample_id_sequencing,
      \(x) synPluck("syn69961369", paste0(x, ".tar.gz"))
    ),
    salmon_tar = map_chr(salmon_syn, syn)
  )

tx2gene <- syn("syn69961460") %>%
  read_tsv()

raw_salmon_batch1 <- meta_batch1 %>%
  mutate(
    quant_path = map2_chr(
      salmon_tar, run_accession,
      \(x, y) {
        out_dir <- file.path(tempdir(), y)
        untar(x, files = "quant.sf", exdir = out_dir)
        file.path(out_dir, "quant.sf")
      }
    )
  ) %>%
  with(
    tximport(
      set_names(quant_path, run_accession),
      type = "salmon",
      txIn = TRUE, txOut = FALSE,
      tx2gene = select(tx2gene, transcript_id, gene_id)
    )
  )

raw_salmon_batch2 <- meta_batch2 %>%
  mutate(
    quant_path = map2_chr(
      salmon_tar, sample_id_sequencing,
      \(x, y) {
        out_dir <- file.path(tempdir(), y)
        untar(x, files = "quant.sf", exdir = out_dir)
        file.path(out_dir, "quant.sf")
      }
    )
  ) %>%
  with(
    tximport(
      set_names(quant_path, sample_id_sequencing),
      type = "salmon",
      txIn = TRUE, txOut = FALSE,
      tx2gene = select(tx2gene, transcript_id, gene_id)
    )
  )

feasible_genes_batch2 <- {raw_salmon_batch2$counts > 5} %>%
  rowSums() %>%
  { tibble(gene_id = names(.), feasible = . > 3) }

write_csv(
  feasible_genes_batch2,
  here("rnaseq", "feasible_genes_batch2.csv.gz")
)

raw_salmon_combined <- meta_raw  %>%
  mutate(
    salmon_syn = map_chr(
      sample_id_sequencing,
      \(x) synPluck("syn69961369", paste0(x, ".tar.gz"))
    ),
    salmon_tar = map_chr(salmon_syn, syn),
    quant_path = map2_chr(
      salmon_tar, sample_id_sequencing,
      \(x, y) {
        out_dir <- file.path(tempdir(), y)
        untar(x, files = "quant.sf", exdir = out_dir)
        file.path(out_dir, "quant.sf")
      }
    )
  ) %>%
  with(
    tximport(
      set_names(quant_path, sample_id_sequencing),
      type = "salmon",
      txIn = TRUE, txOut = FALSE,
      tx2gene = select(tx2gene, transcript_id, gene_id)
    )
  )


```


## RUV count normalization

https://bioconductor.org/packages/devel/bioc/vignettes/tximport/inst/doc/tximport.html#edgeR

```{r}
library(edgeR)
library(RUVSeq)

normMat <- raw_salmon_combined$length / exp(rowMeans(log(raw_salmon_combined$length)))
normCts <- raw_salmon_combined$counts / normMat

eff.lib <- calcNormFactors(normCts) * colSums(normCts)
normMat <- sweep(normMat, 2, eff.lib, "*")
normMat <- log10(normMat)

de_design_combined_no_batch <- ~ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 + ptx:gne_495_3 + ptx:jnk_in_3
de_design_combined <- ~batch + sequencing_batch + ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 + ptx:gne_495_3 + ptx:jnk_in_3
de_design_condition_simple_no_batch <- ~condition_simple
de_design_condition_simple <- ~batch + sequencing_batch + condition_simple

mm_no_batch <- model.matrix(de_design_combined_no_batch, data = meta_raw)
mm <- model.matrix(de_design_combined, data = meta_raw)
mm_condition_simple_no_batch <- model.matrix(de_design_condition_simple_no_batch, data = meta_raw)
mm_condition_simple <- model.matrix(de_design_condition_simple, data = meta_raw)

count_mat_edger <- DGEList(raw_salmon_combined$counts) |>
  scaleOffset(normMat)

keep <- filterByExpr(count_mat_edger, mm_no_batch)
yk <- count_mat_edger[keep, ]

yde <- yk |>
  estimateDisp(design = mm_no_batch, robust = TRUE)

yfit <- yde |>
  glmFit(design = mm_no_batch, robust = TRUE)

yset <- newSeqExpressionSet(
  round(raw_salmon_combined$counts[keep, ]),
  phenoData = data.frame(meta_raw) |>
    column_to_rownames("sample_id_sequencing")
)

ruvk5 <- RUVr(
  yset,
  rownames(yset),
  k = 5,
  residuals = residuals(yfit, type = "deviance")
)

ruvk5_meta <- pData(ruvk5) %>%
  as_tibble(rownames = "sample_id_sequencing")

write_csv(
  ruvk5_meta,
  here("rnaseq", "ruv_k5_meta_no_batch.csv.gz")
)

ruvk5_w_long <- ruvk5_meta |>
  select(sample_id_sequencing, starts_with("W_")) |>
  pivot_longer(
    starts_with("W_"),
    names_to = "W_component",
    values_to = "value"
  )

{ggplot(
  ruvk5_meta,
  aes(
    W_1, W_2,
    color = batch,
    shape = sequencing_batch,
    text = paste("Condition:", condition, "<br>Sample ID:", sample_id_sequencing)
  )
) +
  geom_point(alpha = .7)} |>
  plotly::ggplotly()

ruvk5_meta_vs <- combn(1:5, 2) |>
  t() %>%
  magrittr::set_colnames(c("component_1", "component_2")) %>%
  as_tibble() %>%
  mutate(
    across(
      everything(),
      \(x) paste0("W_", x)
    )
  ) %>%
  left_join(
    ruvk5_w_long,
    by = c("component_1" = "W_component")
  ) %>%
  left_join(
    ruvk5_w_long,
    by = c("component_2" = "W_component", "sample_id_sequencing"),
    suffix = c("_1", "_2")
  ) %>%
  left_join(
    meta_raw,
    by = "sample_id_sequencing"
  )

p <- ggplot(
  ruvk5_meta_vs,
  aes(
    x = value_1,
    y = value_2,
    shape = sequencing_batch,
    color = batch,
    text = paste("Condition:", condition, "<br>Sample ID:", sample_id_sequencing)
  )
) +
  geom_point(alpha = .7) +
  facet_wrap(~component_1 + component_2)

ggsave(
  here("rnaseq", "plots", "ruv_k5_components.pdf"),
  p, width = 8, height = 6
)
```

### Normalization considering batch effects explicitly

```{r}
yde_w_batch <- yk |>
  estimateDisp(design = mm, robust = TRUE)

yfit_w_batch <- yde_w_batch |>
  glmFit(design = mm, robust = TRUE)

ruv_with_batch <- map(
  1:5,
  \(k) RUVr(
    yset,
    rownames(yset),
    k = k,
    residuals = residuals(yfit_w_batch, type = "deviance")
  )
)

ruvk_with_batch_meta <- map(
  ruv_with_batch,
  \(x) pData(x) %>%
    as_tibble(rownames = "sample_id_sequencing")
) |>
  bind_rows(.id = "k") |>
  mutate(across(k, as.integer))

ruvk5_w_long <- ruvk_with_batch_meta |>
  select(k, sample_id_sequencing, starts_with("W_")) |>
  pivot_longer(
    starts_with("W_"),
    names_to = "W_component",
    values_to = "value"
  ) |>
  drop_na(value)

ruvk5_meta_vs <- combn(1:5, 2) |>
  t() %>%
  magrittr::set_colnames(c("component_1", "component_2")) %>%
  as_tibble() %>%
  mutate(
    across(
      everything(),
      \(x) paste0("W_", x)
    )
  ) %>%
  inner_join(
    ruvk5_w_long,
    by = c("component_1" = "W_component")
  ) %>%
  inner_join(
    ruvk5_w_long,
    by = c("component_2" = "W_component", "sample_id_sequencing", "k"),
    suffix = c("_1", "_2")
  ) %>%
  left_join(
    meta_raw,
    by = "sample_id_sequencing"
  )

p <- ggplot(
  ruvk5_meta_vs,
  aes(
    x = value_1,
    y = value_2,
    shape = sequencing_batch,
    color = batch,
    text = paste("Condition:", condition, "<br>Sample ID:", sample_id_sequencing)
  )
) +
  geom_point(alpha = .7) +
  facet_wrap(~k + component_1 + component_2)

ggsave(
  here("rnaseq", "plots", "ruv_k5_components_with_batch.pdf"),
  p, width = 8, height = 6
)
```


```{r}


```

## DESeq2 analysis


### Derived normalized and batch corrected counts

Here purposefully using all genes, not just feasible so they appear
in the VST count table if we need them

```{r}
feasible_genes_combined <- {raw_salmon_combined$counts > 5} %>%
  rowSums() %>%
  { tibble(gene_id = names(.), feasible = . > 3) }

raw_salmon_combined_feasible <- raw_salmon_combined %>%
  imap(
    \(x, nm) {
      if (nm %in% c("counts", "abundance", "length")) {
        x[feasible_genes_combined$feasible, ]
      } else {
        x
      }
    }
  )

de_combined <- DESeqDataSetFromTximport(
  raw_salmon_combined,
  meta_raw,
  design = ~batch + sequencing_batch + ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 + ptx:gne_495_3 + ptx:jnk_in_3,
) %>%
  DESeq()

```


```{r}
combined_counts <- list(
  raw = counts(de_combined, normalized = FALSE),
  normalized = counts(de_combined, normalized = TRUE),
  vst = assay(vst(de_combined, blind = FALSE)),
  vst_blind = assay(vst(de_combined, blind = TRUE)),
  vst_batch_removed = removeBatchEffect(
    assay(vst(de_combined, blind = FALSE)),
    batch = meta_raw$batch,
    batch2 = meta_raw$sequencing_batch,
    design = model.matrix(
      ~ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 + ptx:gne_495_3 + ptx:jnk_in_3,
      data = meta_raw
    )
  ),
  tpm = raw_salmon_combined$abundance,
  tpm_batch_removed = removeBatchEffect(
    log2(raw_salmon_combined$abundance + 1),
    batch = meta_raw$batch,
    batch2 = meta_raw$sequencing_batch,
    design = model.matrix(
      ~ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 + ptx:gne_495_3 + ptx:jnk_in_3,
      data = meta_raw
    )
  )
) %>%
  map(
    \(x) as_tibble(x, rownames = "gene_id")
  ) %>%
  bind_rows(.id = "type")

write_csv(
  combined_counts,
  here("rnaseq", "combined_counts_all.csv.gz")
)
```


```{r}
p <- batch2_counts %>%
  filter(type == "raw") %>%
  select(-type) %>%
  pivot_longer(
    -c(gene_id),
    names_to = "sample_id_sequencing",
    values_to = "count"
  ) %>%
  group_by(sample_id_sequencing) %>%
  summarize(
    total_count = sum(count),
    .groups = "drop"
  ) %>%
  mutate(
    sample_id_sequencing = fct_reorder(sample_id_sequencing, total_count)
  ) %>%
  ggplot(
    aes(x = sample_id_sequencing, y = total_count)
  ) +
  geom_col()

ggsave(
  here("rnaseq", "plots", "total_counts_per_sample_batch2.pdf"),
  p, width = 8, height = 4
)
```



### EMM analysis

```{r}
lrt_res <- results(de_batch2_reduced, tidy = TRUE) %>%
  as_tibble()

resultsNames(de_batch2)

extract_deseq_result <- function(de, contrast, name, res_args = list(), shrink_args = list()) {
  if (missing(contrast)) {
    res <- rlang::exec(results, de, name = name, !!!res_args)
  } else {
    res <- rlang::exec(results, de, contrast = contrast, !!!res_args)
  }
  if (!"type" %in% names(shrink_args)) {
    shrink_args$type <- "ashr"
  }
  shrunken <- rlang::exec(lfcShrink, de, res = res, !!!shrink_args)
  shrunken %>%
    as_tibble(rownames = "gene_id") %>%
    left_join(
      res %>%
        as_tibble(rownames = "gene_id") %>%
        select(gene_id, log2FoldChange, lfcSE),
      by = "gene_id", suffix = c("", "_MLE")
    )
}

res_long <- resultsNames(de_batch2) %>%
  setdiff("Intercept") %>%
  set_names() %>%
  map(
    \(x) extract_deseq_result(de_batch2, name = x)
  ) %>%
  bind_rows(.id = "term")
```



```{r}
single_gene_data <- meta_batch2 %>%
  mutate(
    expression = raw_salmon_batch2$counts["ENSG00000109084", ] %>%
      log10()
  )

mdl <- lm(
  expression ~ ptx*gne_495*jnk_in*batch,
  data = single_gene_data
)

desired_contrasts <- contrast_map <- tribble(
  ~name, ~numerator, ~denominator,
  "GNE-495", "no yes no", "no no no",
  "JNK-inh", "no no yes", "no no no",
  "GNE-495 + PTX", "yes yes no", "no no no",
  "JNK-inh + PTX", "yes no yes", "no no no",
  "PTX", "yes no no", "no no no",
  "GNE-495 + PTX vs PTX", "yes yes no", "yes no no",
  "JNK-inh + PTX vs PTX", "yes no yes", "yes no no",
  "GNE-495 + JNK-inh + PTX vs PTX", "yes yes yes", "yes no no",
  "GNE-495 + JNK-inh", "no yes yes", "no no no",
  "GNE-495 + JNK-inh + PTX vs JNK-inh + PTX", "yes yes yes", "yes no yes",
  "GNE-495 + JNK-inh + PTX vs GNE-495 + PTX", "yes yes yes", "yes yes no"
)

library(emmeans)
mdl_emm <- emmeans(
  mdl, ~ptx+gne_495+jnk_in
) %>%
  contrast(
    method = "revpairwise"
  )

res_emm <- list(
  mdl_emm
) %>%
  map(
    \(x) x@grid  %>%
      as_tibble() %>%
      mutate(
        contrast_vec = asplit(x@linfct, 1)
      )
  ) %>%
  bind_rows() %>%
  mutate(
    res = map(
      contrast_vec,
      \(x) {
        contrast_de <- set_names(
          x[colnames(model.matrix(mdl))],
          resultsNames(de_batch2)
        )
        extract_deseq_result(
          de_batch2,
          contrast = contrast_de
        )
      }
    )
  )

res_emm_long <- desired_contrasts %>%
  mutate(
    contrast = paste(numerator, "-", denominator)
  ) %>%
  power_inner_join(
    res_emm,
    by = "contrast",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_left = "warn"
    )
  ) %>%
  select(-c(numerator, denominator, contrast, contrast_vec)) %>%
  unnest(res) %>%
  power_inner_join(
    distinct(tx2gene, gene_id, gene_name),
    by = "gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

write_csv(
  res_emm_long,
  here("rnaseq", "batch2_deseq2_emm_results.csv.gz")
)
# res_emm_long <- read_csv(
#   here("rnaseq", "batch2_deseq2_emm_results.csv.gz")
# )
```

In the end using k = 0 for RUV as batch effects are modeled explicitly
Additional RUV components may just remove real biological variation here

## DESeq2 using RUVk normalization


```{r}

de_combined_ruv <- DESeqDataSetFromTximport(
  raw_salmon_combined,
  ruvk5_meta,
  design = ~ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 + ptx:gne_495_3 + ptx:jnk_in_3 + W_1 + W_2 + W_3 + W_4 + W_5,
) %>%
  DESeq()

colnames(mm_no_batch)

de_combined_ruv_simple <- DESeqDataSetFromTximport(
  raw_salmon_combined,
  ruvk5_meta %>%
    mutate(
      condition = paste(
        "PTX", ptx,
        "GNE", gne_combined,
        "JNK", jnk_combined,
        sep = "_"
      )
    ),
  design = ~condition + W_1 + W_2 + W_3 + W_4 + W_5,
) %>%
  DESeq()
```


### Extract results

```{r}

de_combined_ruv_simple_res_raw <- tribble(
  ~name, ~numerator, ~denominator,
  "GNE-495", "PTX_no_GNE_{conc}_JNK_0", "PTX_no_GNE_0_JNK_0",
  "JNK-inh", "PTX_no_GNE_0_JNK_{conc}", "PTX_no_GNE_0_JNK_0",
  "GNE-495 + PTX", "PTX_yes_GNE_{conc}_JNK_0", "PTX_no_GNE_0_JNK_0",
  "JNK-inh + PTX", "PTX_no_GNE_0_JNK_{conc}", "PTX_no_GNE_0_JNK_0",
  "PTX", "PTX_yes_GNE_0_JNK_0", "PTX_no_GNE_0_JNK_0",
  "GNE-495 + PTX vs PTX", "PTX_yes_GNE_{conc}_JNK_0", "PTX_yes_GNE_0_JNK_0",
  "JNK-inh + PTX vs PTX", "PTX_yes_GNE_0_JNK_{conc}", "PTX_yes_GNE_0_JNK_0",
  "GNE-495 + JNK-inh + PTX vs PTX", "PTX_yes_GNE_{conc}_JNK_{conc}", "PTX_yes_GNE_0_JNK_0",
  "GNE-495 + JNK-inh", "PTX_no_GNE_{conc}_JNK_{conc}", "PTX_no_GNE_0_JNK_0",
  "GNE-495 + JNK-inh + PTX vs JNK-inh + PTX", "PTX_yes_GNE_{conc}_JNK_{conc}", "PTX_yes_GNE_0_JNK_{conc}",
  "GNE-495 + JNK-inh + PTX vs GNE-495 + PTX", "PTX_yes_GNE_{conc}_JNK_{conc}", "PTX_yes_GNE_{conc}_JNK_0"
) %>%
  cross_join(
    tribble(
      ~batch, ~conc,
      "batch_1", "3",
      "batch_2", "1"
    )
  ) %>%
  mutate(
    across(
      c(numerator, denominator),
      \(z) {
        str_replace_all(z, fixed("{conc}"), conc)
      }
    ),
    res = pmap(
      pick(everything()),
      \(numerator, denominator, ...) {
        possibly(extract_deseq_result)(
          de_combined_ruv_simple,
          contrast = c("condition", numerator, denominator)
        )
      }
    )
  )

de_combined_ruv_simple_res <- de_combined_ruv_simple_res_raw %>%
  select(name, batch, res) %>%
  unnest(res) %>%
  power_left_join(
    distinct(tx2gene, gene_id, gene_name),
    by = "gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )


write_csv(
  de_combined_ruv_simple_res,
  here("rnaseq", "combined_deseq2_ruv_only_results.csv.gz")
)
```

### Compare RUVk vs EMM

```{r}
x_s <- extract_deseq_result(de_combined_ruv_simple, name = "condition_PTX_yes_GNE_0_JNK_0_vs_PTX_no_GNE_0_JNK_0")
x_c <- extract_deseq_result(de_combined_ruv, name = "ptx_yes_vs_no")

print(arrange(x_s, pvalue), n = 20)
print(arrange(x_c, pvalue), n = 20)

de_combined_edger_test_input <- DGEList(raw_salmon_combined$counts) |>
  scaleOffset(normMat)

mm_condition_simple_w_ruv <- model.matrix(
  ~condition_simple + W_1 + W_2 + W_3 + W_4 + W_5,
  data = ruvk5_meta
)

keep <- filterByExpr(de_combined_edger_test_input, mm_condition_simple_w_ruv)
de_combined_edger_test_kept <- de_combined_edger_test_input[keep, ]

de_combined_edger_test_de <- de_combined_edger_test_kept |>
  estimateDisp(design = mm_condition_simple_w_ruv, robust = TRUE) |>
  glmQLFit(design = mm_condition_simple_w_ruv, robust = TRUE, legacy = FALSE)

x_se <- glmQLFTest(
  de_combined_edger_test_de,
  coef = "condition_simplePTX_yes_GNE_0_JNK_0"
) %>%
  topTags(n = Inf) |>
  as.data.frame() %>%
  as_tibble(rownames = "gene_id")

print(res_emm_long %>% filter(name == "PTX") %>% arrange(pvalue), n = 20)


x_s_vs_x_emm <- power_inner_join(
  x_s %>%
    mutate(
      signed_p = -log10(padj) * sign(log2FoldChange_MLE)
    ),
  res_emm_long %>%
    filter(name == "PTX") %>%
    mutate(
      signed_p = -log10(padj) * sign(log2FoldChange_MLE)
    ),
  by = "gene_id",
  suffix = c("_ruv", "_emm"),
  check = check_specs(
    unmatched_keys_left = "warn",
    unmatched_keys_right = "warn",
    duplicate_keys_left = "warn",
    duplicate_keys_right = "warn"
  )
)

ggplot(
  x_s_vs_x_emm,
  aes(
    x = signed_p_ruv,
    y = signed_p_emm,
    text = paste("Gene:", gene_name)
  )
) +
  geom_point(alpha = .5) +
  geom_abline(slope = 1, intercept = 0, color = "red")
```

### Heatmap LFC

```{r}
res_emm_hm_data <- res_emm_long %>%
  semi_join(
    filter(., padj < .001),
    by = "gene_id"
  ) %>%
  replace_na(list(padj = 1)) %>%
  mutate(
    padj = if_else(padj == 0, .5 * min(padj[padj > 0], na.rm = TRUE), padj),
    signed_p = -log10(padj) * sign(log2FoldChange_MLE)
  ) %>%
  select(
    name, gene_id, signed_p
  ) %>%
  pivot_wider(
    names_from = name,
    values_from = signed_p,
    values_fill = 0
  )

res_emm_hm_mat <- res_emm_hm_data %>%
  column_to_rownames("gene_id") %>%
  as.matrix()

hm <- Heatmap(
  res_emm_hm_mat,
  name = "signed -log10(p)",
  cluster_rows = TRUE,
  cluster_columns = FALSE,
  show_row_names = FALSE,
  show_column_names = TRUE,
  col = circlize::colorRamp2(
    quantile(res_emm_hm_mat, c(0.05, 0.95), na.rm = TRUE) %>%
      abs() %>%
      max() %>% {
        .*c(-1, 0, 1)
      },
    c("blue", "white", "red")
  )
)

hm
```


## Deseq combined


```{r}

feasible_genes_combined <- {raw_salmon_combined$counts > 5} %>%
  rowSums() %>%
  { tibble(gene_id = names(.), feasible = . > 3) }

raw_salmon_combined_feasible <- raw_salmon_combined %>%
  imap(
    \(x, nm) {
      if (nm %in% c("counts", "abundance", "length")) {
        x[feasible_genes_combined$feasible, ]
      } else {
        x
      }
    }
  )

de_combined_ruv <- bind_rows(
  meta_raw |>
    mutate(k = 0L),
  ruvk_with_batch_meta
) |>
  group_nest(k) |>
  mutate(
    de = map2(
      data, k,
      \(x, k) {
        x <- x |>
          select(where(\(y) !all(is.na(y))))
        design_formula <- as.formula(
          paste(
            "~ batch + sequencing_batch +",
            "ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 +",
            "ptx:gne_495_3 + ptx:jnk_in_3",
            if (k > 0) paste0("+ W_", seq_len(k), collapse = " + ") else ""
          )
        )
        DESeqDataSetFromTximport(
          raw_salmon_combined_feasible,
          x,
          design = design_formula,
        ) %>%
          DESeq() %>%
          nbinomWaldTest(maxit = 1000)
      }
    )
  )

de_combined <- DESeqDataSetFromTximport(
  raw_salmon_combined_feasible,
  meta_raw,
  design = ~batch + sequencing_batch + ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 + ptx:gne_495_3 + ptx:jnk_in_3,
) %>%
  DESeq() %>%
  nbinomWaldTest(maxit = 1000)
#  + ptx:gne_495_3 + ptx:jnk_in_3,


combined_counts <- list(
  raw = counts(de_combined, normalized = FALSE),
  normalized = counts(de_combined, normalized = TRUE),
  vst = vst(de_combined) %>%
    assay()
) %>%
  map(
    \(x) as_tibble(x, rownames = "gene_id")
  ) %>%
  bind_rows(.id = "type")
```


## Variance partitioning


```{r}
library(variancePartition)

varPart <- fitExtractVarPartModel(
  combined_counts %>%
    filter(type == "vst") %>%
    select(-type) %>%
    column_to_rownames("gene_id") %>%
    as.matrix(),
  ~ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 + ptx:gne_495_3 + ptx:jnk_in_3 + batch + sequencing_batch,
  meta_raw |>
    column_to_rownames("sample_id_sequencing")
)
plotVarPart(varPart)
```


```{r}
pca_combined_raw <- combined_counts %>%
  filter(type == "vst") %>%
  select(-type) %>%
  column_to_rownames("gene_id") %>%
  t() %>%
  prcomp(center = TRUE, scale = FALSE)

pca_x_combined <- pca_combined_raw$x %>%
  as_tibble(rownames = "sample_id_sequencing") %>%
  power_inner_join(
    meta_raw,
    by = "sample_id_sequencing",
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  )

pca_sd_combined <- broom::tidy(pca_combined_raw, "eigenvalues")

```
## DGEseq2 DE RUV

```{r}
mdl_emm_ruv <- de_combined_ruv |>
  mutate(
    mdl = map(
      de,
      \(de) {
        single_gene_data <- colData(de) |>
          as_tibble() |>
          mutate(
            expression = combined_counts |>
              filter(type == "vst", gene_id == "ENSG00000109084") |>
              select(-type, -gene_id) |>
              unlist()
          )
        lm(
          paste(
            "expression",
            paste(as.character(design(de)), collapse = " ")
          ) |>
            as.formula(),
          data = single_gene_data
        )
      }
    ),
    emm = map2(
      mdl, k,
      \(mdl, k) {
        nuisance_values <- set_names(
          rep(0, k),
          paste0("W_", seq_len(k))
        )
        list(
          batch_1 = emmeans(
            mdl, ~ptx + gne_495_3 + jnk_in_3,
            at = c(
              list(gne_495_1 = "no", jnk_in_1 = "no"),
              nuisance_values
            )
          ),
          batch_2 = emmeans(
            mdl, ~ptx + gne_495_1 + jnk_in_1,
            at = c(
              list(gne_495_3 = "no", jnk_in_3 = "no"),
              nuisance_values
            )
          )
        )
      }
    )
  )


res_emm_ruv <- mdl_emm_ruv |>
  select(-data) |>
  unnest_longer(emm, indices_to = "batch", values_to = "emm") |>
  mutate(
    res = pmap(
      pick(everything()),
      \(emm, de, mdl, ...) {
        contr <- contrast(emm, method = "revpairwise")
        contr@grid %>%
          as_tibble() %>%
          mutate(
            contrast_vec = asplit(contr@linfct, 1)
          ) %>%
          mutate(
            res = map(
              contrast_vec,
              \(x) {
                contrast_de <- set_names(
                  x[colnames(model.matrix(mdl))],
                  resultsNames(de)
                )
                extract_deseq_result(
                  de,
                  contrast = contrast_de
                )
              }
            )
          ) %>%
          select(-contrast_vec) %>%
          unnest(res)
      }
    )
  )

qsave(
  res_emm_ruv,
  here("rnaseq", "combined_deseq2_emm_ruv_results.qs")
)


desired_contrasts <- tribble(
  ~name, ~numerator, ~denominator,
  "GNE-495", "no yes no", "no no no",
  "JNK-inh", "no no yes", "no no no",
  "GNE-495 + PTX", "yes yes no", "no no no",
  "JNK-inh + PTX", "yes no yes", "no no no",
  "PTX", "yes no no", "no no no",
  "GNE-495 + PTX vs PTX", "yes yes no", "yes no no",
  "JNK-inh + PTX vs PTX", "yes no yes", "yes no no",
  "GNE-495 + JNK-inh + PTX vs PTX", "yes yes yes", "yes no no",
  "GNE-495 + JNK-inh", "no yes yes", "no no no",
  "GNE-495 + JNK-inh + PTX vs JNK-inh + PTX", "yes yes yes", "yes no yes",
  "GNE-495 + JNK-inh + PTX vs GNE-495 + PTX", "yes yes yes", "yes yes no",
  "GNE-495 + JNK-inh + PTX", "yes yes yes", "no no no"
)

res_emm_ruv_long <- desired_contrasts %>%
  mutate(
    contrast = paste(numerator, "-", denominator)
  ) %>%
  power_inner_join(
    res_emm_ruv %>%
      select(k, batch, res) %>%
      unnest(res),
    by = "contrast",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_left = "warn"
    )
  ) %>%
  select(-c(numerator, denominator, contrast)) %>%
  power_inner_join(
    distinct(tx2gene, gene_id, gene_name),
    by = "gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

write_csv(
  res_emm_ruv_long,
  here("rnaseq", "combined_deseq2_emm_ruv_results.csv.gz")
)

single_gene_data <- meta_raw %>%
  mutate(
    expression = raw_salmon_combined$counts["ENSG00000109084", ] %>%
      log10()
  )

mdl <- lm(
  expression ~ batch + sequencing_batch + ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 + ptx:gne_495_3 + ptx:jnk_in_3,
  data = single_gene_data
)

library(emmeans)
mdl_emm_batch_2 <- emmeans(
  mdl, ~ptx + gne_495_1 + jnk_in_1,
  at = list(gne_495_3 = "no", jnk_in_3 = "no")
)

mdl_emm_batch_1 <- emmeans(
  mdl, ~ptx + gne_495_3 + jnk_in_3,
  at = list(gne_495_1 = "no", jnk_in_1 = "no")
)

res_emm <- list(
  batch_1 = mdl_emm_batch_1,
  batch_2 = mdl_emm_batch_2
) %>%
  map(
    \(x) {
      contr <- contrast(x, method = "revpairwise")
      contr@grid %>%
        as_tibble() %>%
        mutate(
          contrast_vec = asplit(contr@linfct, 1)
        )
    }
  ) %>%
  bind_rows(.id = "batch") %>%
  mutate(
    res = map(
      contrast_vec,
      \(x) {
        contrast_de <- set_names(
          x[colnames(model.matrix(mdl))],
          resultsNames(de_combined)
        )
        extract_deseq_result(
          de_combined,
          contrast = contrast_de
        )
      }
    )
  )

res_emm_long <- desired_contrasts %>%
  mutate(
    contrast = paste(numerator, "-", denominator)
  ) %>%
  power_inner_join(
    res_emm,
    by = "contrast",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_left = "warn"
    )
  ) %>%
  select(-c(numerator, denominator, contrast, contrast_vec)) %>%
  unnest(res) %>%
  power_inner_join(
    distinct(tx2gene, gene_id, gene_name),
    by = "gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

write_csv(
  res_emm_long,
  here("rnaseq", "combined_deseq2_emm_results.csv.gz")
)
# res_emm_long <- read_csv(
#   here("rnaseq", "combined_deseq2_emm_ruv_results.csv.gz")
# )
```


### Average effect doses

```{r}
mdl_emm_ruv_avg <- de_combined_ruv |>
  mutate(
    mdl = map(
      de,
      \(de) {
        single_gene_data <- colData(de) |>
          as_tibble() |>
          mutate(
            expression = combined_counts |>
              filter(type == "vst", gene_id == "ENSG00000109084") |>
              select(-type, -gene_id) |>
              unlist()
          )
        lm(
          paste(
            "expression",
            paste(as.character(design(de)), collapse = " ")
          ) |>
            as.formula(),
          data = single_gene_data
        )
      }
    ),
    emm = map2(
      mdl, k,
      \(mdl, k) {
        nuisance_values <- set_names(
          rep(0, k),
          paste0("W_", seq_len(k))
        )
        list(
          combined_batch = emmeans(
            mdl, ~ptx + gne_495_3 + jnk_in_3 + gne_495_1 + jnk_in_1,
            at = c(
              nuisance_values
            )
          )
        )
      }
    )
  )

res_emm_ruv_avg_contrasts <- tribble(
  ~name, ~numerator, ~denominator,
  "GNE-495", "no yes no", "no no no",
  "JNK-inh", "no no yes", "no no no",
  "GNE-495 + PTX", "yes yes no", "no no no",
  "JNK-inh + PTX", "yes no yes", "no no no",
  "PTX", "yes no no", "no no no",
  "GNE-495 + PTX vs PTX", "yes yes no", "yes no no",
  "JNK-inh + PTX vs PTX", "yes no yes", "yes no no",
  "GNE-495 + JNK-inh + PTX vs PTX", "yes yes yes", "yes no no",
  "GNE-495 + JNK-inh", "no yes yes", "no no no",
  "GNE-495 + JNK-inh + PTX vs JNK-inh + PTX", "yes yes yes", "yes no yes",
  "GNE-495 + JNK-inh + PTX vs GNE-495 + PTX", "yes yes yes", "yes yes no",
  "GNE-495 + JNK-inh + PTX", "yes yes yes", "no no no"
)

res_emm_ruv_avg <- mdl_emm_ruv_avg |>
  select(-data) |>
  unnest_longer(emm, indices_to = "batch", values_to = "emm") |>
  mutate(
    res = pmap(
      pick(everything()),
      \(emm, de, mdl, ...) {
        contr <- contrast(emm, method = "revpairwise")
        contr@grid %>%
          as_tibble() %>%
          mutate(
            contrast_vec = asplit(contr@linfct, 1)
          ) %>%
          mutate(
            res = map(
              contrast_vec,
              \(x) {
                contrast_de <- set_names(
                  x[colnames(model.matrix(mdl))],
                  resultsNames(de)
                )
                extract_deseq_result(
                  de,
                  contrast = contrast_de
                )
              }
            )
          ) %>%
          select(-contrast_vec) %>%
          unnest(res)
      }
    )
  )

qsave(
  res_emm_ruv,
  here("rnaseq", "combined_deseq2_emm_ruv_results.qs")
)

```

## EdgeR DE

```{r}

edger_combined_ruv <- bind_rows(
  meta_raw |>
    mutate(k = 0L),
  ruvk_with_batch_meta
) |>
  group_nest(k) |>
  mutate(
    de = map2(
      data, k,
      \(x, k) {
        x <- x |>
          select(where(\(y) !all(is.na(y))))
        design_formula <- as.formula(
          paste(
            "~ batch + sequencing_batch +",
            "ptx * gne_495_1 * jnk_in_1 + gne_495_3 + jnk_in_3 +",
            "ptx:gne_495_3 + ptx:jnk_in_3",
            if (k > 0) paste0("+ W_", seq_len(k), collapse = " + ") else ""
          )
        )
        mm <- model.matrix(design_formula, data = x)
        keep <- filterByExpr(count_mat_edger, mm)
        yk <- count_mat_edger[keep, ]

        yde <- yk |>
          estimateDisp(design = mm, robust = TRUE)

        yfit <- yde |>
          glmQLFit(design = mm, robust = TRUE)

        list(design = design_formula, fit = yfit)
      }
    )
  )

edger_mdl_emm_ruv <- edger_combined_ruv |>
  mutate(
    mdl = map2(
      de, data,
      \(de, m) {
        single_gene_data <- m |>
          as_tibble() |>
          mutate(
            expression = combined_counts |>
              filter(type == "vst", gene_id == "ENSG00000109084") |>
              select(-type, -gene_id) |>
              unlist()
          )
        lm(
          paste(
            "expression",
            paste(as.character(de$design), collapse = " ")
          ) |>
            as.formula(),
          data = single_gene_data
        )
      }
    ),
    emm = map2(
      mdl, k,
      \(mdl, k) {
        nuisance_values <- set_names(
          rep(0, k),
          paste0("W_", seq_len(k))
        )
        list(
          batch_1 = emmeans(
            mdl, ~ptx + gne_495_3 + jnk_in_3,
            at = c(
              list(gne_495_1 = "no", jnk_in_1 = "no"),
              nuisance_values
            )
          ),
          batch_2 = emmeans(
            mdl, ~ptx + gne_495_1 + jnk_in_1,
            at = c(
              list(gne_495_3 = "no", jnk_in_3 = "no"),
              nuisance_values
            )
          )
        )
      }
    )
  )


edger_res_emm_ruv <- edger_mdl_emm_ruv |>
  select(-data) |>
  unnest_longer(emm, indices_to = "batch", values_to = "emm") |>
  mutate(
    res = pmap(
      pick(everything()),
      \(emm, de, mdl, ...) {
        contr <- contrast(emm, method = "revpairwise")
        # browser()
        contr@grid %>%
          as_tibble() %>%
          mutate(
            contrast_vec = asplit(linfct(contr), 1)
          ) %>%
          mutate(
            res = map(
              contrast_vec,
              \(x) {
                glmQLFTest(de$fit, contrast = x) |>
                  as.data.frame() |>
                  as_tibble(rownames = "gene_id") |>
                  mutate(
                    FDR = p.adjust(PValue, method = "BH")
                  )
              }
            )
          ) %>%
          select(-contrast_vec) %>%
          unnest(res)
      }
    )
  )

qsave(
  edger_res_emm_ruv,
  here("rnaseq", "combined_edger_emm_ruv_results.qs")
)


edger_res_emm_ruv_long <- desired_contrasts %>%
  mutate(
    contrast = paste(numerator, "-", denominator)
  ) %>%
  power_inner_join(
    edger_res_emm_ruv %>%
      select(k, batch, res) %>%
      unnest(res),
    by = "contrast",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_left = "warn"
    )
  ) %>%
  select(-c(numerator, denominator, contrast)) %>%
  power_inner_join(
    distinct(tx2gene, gene_id, gene_name),
    by = "gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

write_csv(
  edger_res_emm_ruv_long,
  here("rnaseq", "combined_edger_emm_ruv_results.csv.gz")
)

```


## Select good k for RUVr

```{r}
de_stats_ruvk <- res_emm_ruv_long |>
  group_by(k, batch, name) %>%
  summarize(
    n_sig = sum(padj < .05, na.rm = TRUE),
    .groups = "drop"
  ) |>
  arrange(name, batch, as.integer(k)) |>
  mutate(
    batch = recode(
      batch,
      batch_1 = "Batch 1 (3uM)",
      batch_2 = "Batch 2 (1uM)"
    )
  )

p <- ggplot(
  de_stats_ruvk,
  aes(
    x = n_sig,
    y = name,
    fill = fct_inseq(as.character(k))
  )
) +
  geom_col(position = position_dodge(width = 0.7)) +
  facet_wrap(~batch) +
  labs(
    fill = "RUVr k",
    x = "Number of DE genes (FDR < 0.05)",
    y = NULL
  )
p
ggsave(
  here("rnaseq", "plots", "de_genes_by_k_batch_combined.pdf"),
  p, width = 8, height = 6
)
```


```{r}
edger_de_stats_ruvk <- edger_res_emm_ruv_long |>
  group_by(k, batch, name) %>%
  summarize(
    n_sig = sum(FDR < .05, na.rm = TRUE),
    .groups = "drop"
  ) |>
  arrange(name, batch, as.integer(k)) |>
  mutate(
    batch = recode(
      batch,
      batch_1 = "Batch 1 (3uM)",
      batch_2 = "Batch 2 (1uM)"
    )
  )

p <- ggplot(
  edger_de_stats_ruvk,
  aes(
    x = n_sig,
    y = name,
    fill = fct_inseq(as.character(k))
  )
) +
  geom_col(position = position_dodge(width = 0.7)) +
  facet_wrap(~batch) +
  labs(
    fill = "RUVr k",
    x = "Number of DE genes (FDR < 0.05)",
    y = NULL
  )
p
ggsave(
  here("rnaseq", "plots", "de_genes_by_k_batch_combined.pdf"),
  p, width = 8, height = 6
)
```

## Export for GEO

```{r}
geo_counts <- raw_salmon_combined[c("abundance", "counts", "length")] %>%
  map(
    \(x) as_tibble(x, rownames = "gene_id")
  ) %>%
  bind_rows(.id = "type")

write_csv(
  geo_counts,
  here("rnaseq", "combined_counts_salmon.csv.gz")
)
```


## Save to Synapse

```{r}
synStoreMany(
  c(
    here(
      "rnaseq",
      c(
        "batch2_deseq2_emm_results.csv.gz",
        "feasible_genes_batch2.csv.gz",
        "combined_deseq2_emm_results.csv.gz",
        "combined_deseq2_emm_ruv_results.csv.gz",
        "combined_edger_emm_ruv_results.csv.gz",
        "combined_counts_all.csv.gz",
        "meta_combined.csv.gz",
        "combined_deseq2_ruv_only_results.csv.gz"
      )
    )
  ),
  parentId = "syn69968963",
  forceVersion = FALSE
)
```
